---
title: "03_HDCA_stereoscope_preparation"
author: "Julia Foyer, Ludvig Larsson & RaphaÃ«l Mauron"
output: html_document
editor_options: 
  chunk_output_type: console
---

# HDCA heart - stereoscope preparation

In order to perform deconvolution using `stereoscope`, several files are required:
* single-cell annotation (HDCA_heart_sc_annotation.tsv)
* single-cell variable features (HDCA_heart_sc_var.features_5000.txt)
* single-cell count matrix (HDCA_heart_sc_cnt_data.tsv)
* spatial transcriptomics data (hdca_heart_ST_data.tsv)

Here we generate the 3 *single-cell* files required.

Note that it has been decided to perform deconvulution on the high level annotation AND on the deep level annotation. Therefore, a total of 6 files are generated here.

## Settings

### Set library path

```{r setup, include=FALSE}
.libPaths("/home/rstudio/project/renv/library/R-4.3/aarch64-unknown-linux-gnu/")
knitr::opts_chunk$set(echo = TRUE)
```


### Set environment variables

```{r}
# Set the directory 
myfolder <- "/home/rstudio/project/output/HDCA_heart_sc_analysis_docker/"
```


### Load Seurat object

```{r, message = FALSE}
out_path <- "/home/rstudio/project/output/HDCA_heart_sc_analysis_docker/stereoscope/"
data <- readRDS(paste0(myfolder, "rds_objects/HDCA_heart_sc_full_extended.rds"))
```


### Load libraries

```{r, message = FALSE}
library(Seurat)
library(Matrix)
library(dplyr)
library(data.table)
library(tibble)
```



## High level clusters
HL stereoscope: exclude 6, 19, 27, 34

```{r}
data_curated <- data[, data$high_level_clusters %in% c("1", "2", "3", "4_35", "5", "6", "7", "8", "9", "10", "11", "12", "13", "14", "15", "16", "17", "18_1", "18_2", "20", "21", "22", "23", "24", "25", "26", "28", "29", "30", "31", "32", "33", "34")]
```


### Get the 5000 most highly variable genes

Export top 5000 most variable genes for stereoscope.

```{r, message = FALSE, warning=FALSE}
data_curated <- FindVariableFeatures(data_curated, nfeatures = 5000)
var.features <- data_curated@assays$RNA@var.features

# Create path if not existing
if (!dir.exists(paste0(out_path, "/high_level/"))){
  dir.create(paste0(out_path, "/high_level/"))}

file <- paste0(out_path, "high_level/HDCA_heart_sc_var.features_5000.txt")
write.table(x = var.features, file = file, quote = FALSE, row.names = FALSE, col.names = FALSE)
```


### Down sample cells

```{r}
sampled_cells <- data_curated[[]] %>% 
  tibble::rownames_to_column(var = "barcode") %>% 
  group_by(high_level_clusters) %>% 
  arrange(-nFeature_RNA) %>% 
  slice_head(n = 500)
```


### Save the count matrix

Export count matrix for stereoscope. 

```{r, message = FALSE, warning=FALSE}
counts <- GetAssayData(data_curated, slot = "counts", assay = "RNA")
counts <- counts[var.features, sampled_cells$barcode]

# Sanity check
all(colnames(counts) == sampled_cells$barcode)

out_file <- paste0(out_path, "high_level/HDCA_heart_sc_cnt_data.tsv")
counts_df <- counts %>% t() %>% as.matrix() %>% as.data.frame()
data.table::fwrite(counts_df, file = out_file, sep = "\t", quote = FALSE, 
                   row.names = TRUE, col.names = TRUE)
```


### Export annotations for high level clusters

```{r, , message = FALSE}
annotation <- sampled_cells %>% 
  select(barcode, high_level_clusters) %>% 
  data.frame(row.names = 1) #data[["high_level_clusters"]]
colnames(annotation) <- "bio_celltype"

file <- paste0(out_path, "high_level/HDCA_heart_sc_annotation.tsv")
data.table::fwrite(annotation, row.names = TRUE, file = file, sep = "\t")
```

```{r}
exprMat <- data.table::fread(out_file, sep = "\t", header = TRUE) %>% 
  data.frame(row.names = 1) %>% 
  as.matrix()

annData <- read.table(file, header = TRUE, row.names = 1)

# Check that cells are identical in count matrix and annotation data.frame
all(rownames(exprMat) == rownames(annData))
```


## Deep clustering

### Keep selected cell types

Clusters to keep for stereoscope:
- HL: 6, 8, 10, 15, 18_1, 20, 22, 25, 26, 29, 34
to exclude since they are in deeplevel:
FB: 13, 30, 2, 23, 1, 14, 16, 31
EC: 32, 28, 12, 24, 3, 21, 33
CM: 4-35, 17, 9, 5, 11, 7, 18_2

deep level:
-CM: 1, 2, 3, 4-1, 4-2, 5, 6, 7, 8-1, 8-2, 8-3, 9, 10, 12, 14, 17, 18, 19-1, 19-2, 21 
-EN: 1, 2, 3, 5, 6, 7, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22
-FB: 1, 2, 3, 4, 5, 6, 7, 8, 9, 11, 12, 13, 14, 15, 17, 18, 19, 20, 21

cluster to exclude:
- high level: 6, 19, 27, 34
- CM 11, 13, 15, 16, 20, 22
- EN 4, 8, 9
- FB 10, 16, 22

```{r}
# Load annotations prepared earlier and filter
# CM
cardiomyocyte_clusters_keep <- c("1", "2", "3", "4-1", "4-2", "5", "6", "7", "8-1", "8-2", "8-3", "9", "10", "12", "14", "17", "18", "19-1", "19-2", "21")
cardiomyocyte_metadata <- readRDS(paste0(myfolder, "rds_objects/cardiomyocyte_metadata.rds")) %>% 
  mutate(clusters_subset = paste0("CM_", as.character(clusters_subset))) %>%  # rename cardiomyocyte clusters by adding a prefix CM_
  mutate_if(is.factor, as.character) %>% 
  filter(clusters_subset_split %in% cardiomyocyte_clusters_keep)

# EN
endothelial_clusters_keep <- c("1", "2", "3", "5", "6", "7", "10", "11", "12", "13", "14", "15", "16", "17", "18", "19", "20", "21", "22")
endothelial_metadata <- readRDS(paste0(myfolder, "rds_objects/endothelial_metadata.rds")) %>% 
  filter(clusters_subset %in% endothelial_clusters_keep) %>% 
  mutate(clusters_subset = paste0("EN_", as.character(clusters_subset))) # rename endothelial clusters by adding a prefix EN_

# FB
fibroblasts_clusters_keep <- c("1", "2", "3", "4", "5", "6", "7", "8", "9", "11", "12", "13", "14", "15", "17", "18", "19", "20", "21")
fibroblasts_metadata <- readRDS(paste0(myfolder, "rds_objects/fibroblasts_metadata.rds")) %>% 
  filter(clusters_subset %in% fibroblasts_clusters_keep) %>% 
  mutate(clusters_subset = paste0("FB_", as.character(clusters_subset))) # rename fibroblasts clusters by adding a prefix FB_

# Merge deep level clusters from the three different subsets
deep_clusters <- cardiomyocyte_metadata %>% 
  select(clusters_subset, clusters_subset_split) %>% # Only select clustering results
  bind_rows(endothelial_metadata %>% select(clusters_subset)) %>% # Add endothelial clusters
  bind_rows(fibroblasts_metadata %>% select(clusters_subset)) %>% # Add fibroblast clusters
  mutate(clusters_subset_split = case_when(is.na(clusters_subset_split) ~ clusters_subset, # rename cardiomyocyte clusters by adding a prefix CM_
                                           TRUE ~ paste0("CM_", clusters_subset_split)))

# Define clusters to keep
keep_clusters <- c(paste0("HL_", c("6", "8", "10", "15", "18_1", "20", "22", "25", "26", "29", "34")),
                   paste0("CM_", cardiomyocyte_clusters_keep),
                   paste0("EN_", endothelial_clusters_keep),
                   paste0("FB_", fibroblasts_clusters_keep))


# Create a new annotations tibble
annotation_split <- annotation 
annotation_split <- annotation_split %>% 
  rownames_to_column(var = "ID") %>% 
  left_join(y = deep_clusters %>% rownames_to_column(var = "ID"), by = "ID") %>% 
  mutate(clusters_subset_split = case_when(is.na(clusters_subset_split) ~ paste0("HL_", bio_celltype),
                                           TRUE ~ clusters_subset_split)) %>% 
  filter(clusters_subset_split %in% keep_clusters) # keep selected clusters

# Convert clusters_subset_split to a factor and set order of clusters
annotation_split <- annotation_split %>% 
  mutate(clusters_subset_split = factor(clusters_subset_split, keep_clusters))

# Check overlap (sanity check)
all(annotation$ID %in% colnames(data))
```


#### Export files for sampled cells

Keep a maximum of 250 cells per cell type to speed up computation. Select the top 250 cells with highest numbers of nFeature_RNA (unique genes).

```{r}
# Create path if not existing
if (!dir.exists(paste0(out_path, "/deep_level/"))){
  dir.create(paste0(out_path, "/deep_level/"))}

# Down sampling
clusters_split_cells <- annotation_split %>% 
  left_join(y = data[[]] %>% 
              rownames_to_column(var = "ID") %>% 
              select(ID, sampleID, nFeature_RNA), by = "ID") %>% 
  group_by(clusters_subset_split) %>% # Group by deep cluster
  arrange(-nFeature_RNA) %>% # Arrange by decreasing number of unique genes
  slice_head(n = 250) %>% # Select top 250 cells (highest quality cells based on number of unique genes)
  arrange(clusters_subset_split) # Arrange cells by cluster factor levels
  
# Subset Seurat object using the cell IDs created above
deep_level_subset_data <- data[, clusters_split_cells$ID]

# Find top variable features for selected cells
deep_level_subset_data <- FindVariableFeatures(deep_level_subset_data, nfeatures = 5000)

# Export variable genes
file <- paste0(out_path, "deep_level/HDCA_heart_sc_var_features_deep_level_5000.txt")
write.table(x = VariableFeatures(deep_level_subset_data), 
            file = file, quote = FALSE, row.names = FALSE, col.names = FALSE)

# Subset counts to include top 5000 most variable features (no need to export all genes)
counts <- GetAssayData(deep_level_subset_data, slot = "counts")[VariableFeatures(deep_level_subset_data), ]

# Export counts
out_file <- paste0(out_path, "deep_level/HDCA_heart_sc_cnt_data_deep_level.tsv")
counts_df <- counts %>% t() %>% as.matrix() %>% as.data.frame()
data.table::fwrite(counts_df, file = out_file, sep = "\t", quote = FALSE, 
                   row.names = TRUE, col.names = TRUE)

# Check that cell IDs match (sanity check)
all(clusters_split_cells$ID == colnames(counts))

# Export annotations for selected cells
file <- paste0(out_path, "deep_level/HDCA_heart_sc_annotation_deep_level.tsv")
data.table::fwrite(clusters_split_cells %>% 
                     column_to_rownames(var = "ID") %>% 
                     select(clusters_subset_split) %>% 
                     rename(bio_celltype = clusters_subset_split), 
       row.names = TRUE, file = file, sep = "\t")
```


#### Sanity check

```{r}
exprMat <- data.table::fread(out_file) %>% data.frame(row.names = 1, check.names = FALSE) %>% as.matrix()
var_genes <- readLines(paste0(out_path, "deep_level/HDCA_heart_sc_var_features_deep_level_5000.txt"))

# Check variable features
all(var_genes == colnames(exprMat))

# Check cell IDs
ann <- read.table(paste0(out_path, "deep_level/HDCA_heart_sc_annotation_deep_level.tsv"), header = T, row.names = 1)
all(rownames(ann) == rownames(exprMat))
```





###### Export files for all cells

Keep all cells for the selected cell types. Including all cells will make the stereoscope run take longer, but might be more sensitive.

```{r}
# # Subset deep level cells
# deep_level_subset_data <- data[, annotation_split$ID]
# 
# # Find top variable features
# deep_level_subset_data <- FindVariableFeatures(deep_level_subset_data, nfeatures = 5000)
# 
# # Export variable genes
# file <- paste0(out_path, "all_cells/HDCA_heart_sc_var_features_deep_level_5000.txt")
# write.table(x = VariableFeatures(deep_level_subset_data),
#             file = file, quote = FALSE, row.names = FALSE, col.names = FALSE)
# 
# # Subset counts
# counts <- GetAssayData(deep_level_subset_data, slot = "counts")[VariableFeatures(deep_level_subset_data), ]
# 
# # Export counts
# out_file <- paste0(out_path, "all_cells/HDCA_heart_sc_cnt_data_deep_level.tsv")
# counts_df <- counts %>% t() %>% as.matrix() %>% as.data.frame()
# data.table::fwrite(counts_df, file = out_file, sep = "\t", quote = FALSE,
#                    row.names = TRUE, col.names = TRUE)
# 
# # Check
# all(annotation_split$ID == colnames(counts))
# 
# # Export metadata
# file <- paste0(out_path, "all_cells/HDCA_heart_sc_annotation_deep_level.tsv")
# data.table::fwrite(annotation_split %>%
#                      column_to_rownames(var = "ID") %>%
#                      select(clusters_subset_split) %>%
#                      rename(bio_celltype = clusters_subset_split),
#        row.names = TRUE, file = file, sep = "\t")
```


###### Sanity check

```{r}
# exprMat <- data.table::fread(out_file) %>% data.frame(row.names = 1, check.names = FALSE) %>% as.matrix()
# var_genes <- readLines(paste0(out_path, "all_cells/HDCA_heart_sc_var_features_deep_level_5000.txt"))
# 
# # Check variable features
# all(var_genes == colnames(exprMat))
# 
# # Check cell IDs
# ann <- read.table(paste0(out_path, "all_cells/HDCA_heart_sc_annotation_deep_level.tsv"), header = T, row.names = 1)
# all(rownames(ann) == rownames(exprMat))
```


================================================================================
================================================================================
================================================================================
================================================================================
================================================================================









## Plot cluster layers / donut ??

Multiple layered "donut chart" to visualize the three layers of clustering

```{r, fig.asp=1}
library(stringr)

gg <- annotation_split %>% 
  mutate_if(is.factor, as.character) %>% 
  mutate(Level1 = case_when(str_detect(clusters_subset, "^CM") ~ "Cardiomyocytes",
                            str_detect(clusters_subset, "^EC") ~ "Endothelial",
                            str_detect(clusters_subset, "^FB") ~ "Fibroblasts",
                            TRUE ~ "Other")) %>% 
  select(Level1, bio_celltype,clusters_subset_split) %>% 
  setNames(nm = paste0("Level", 1:3)) %>% 
  arrange(Level1, Level2, Level3) %>% 
  mutate(Level1 = factor(Level1, levels = unique(Level1)),
         Level2 = factor(Level2, levels = unique(Level2)),
         Level3 = factor(Level3, levels = unique(Level3)))
gg1 <- gg %>% 
  group_by(Level1) %>% 
  summarize(n = n(), .groups = "drop") %>% 
  mutate(ymid = 1,
         ymax = ymid + 0.5, ymin = ymid - 0.4,
         xmin = c(0, head(cumsum(n), -1)),
         xmax = cumsum(n),
         xmid = (xmax + xmin) / 2)
gg2 <- gg %>% 
  group_by(Level2) %>% 
  summarize(n = n(), .groups = "drop") %>% 
  mutate(ymid = 2,
         ymax = ymid + 0.5, ymin = ymid - 0.4,
         xmin = c(0, head(cumsum(n), -1)),
         xmax = cumsum(n),
         xmid = (xmax + xmin) / 2)
gg3 <- gg %>% 
  group_by(Level3) %>% 
  summarize(n = n(), .groups = "drop") %>% 
  mutate(ymid = 3,
         ymax = ymid + 0.5, ymin = ymid - 0.4,
         xmin = c(0, head(cumsum(n), -1)),
         xmax = cumsum(n),
         xmid = (xmax + xmin) / 2)

# Combine results
ggc <- gg1 %>% bind_rows(gg2) %>% bind_rows(gg3) %>% 
  mutate(cols = case_when(is.na(as.character(Level1)) ~ as.character(Level2),
                          TRUE ~ as.character(Level1))) %>% 
  mutate(cols = case_when(is.na(as.character(cols)) ~ as.character(Level3),
                          TRUE ~ cols))

# Define colors
L1_lvls <- gg1 %>% pull(Level1) %>% as.character()
CM_lvls <- gg2 %>% filter(xmax <= 14590) %>% pull(Level2) %>% as.character()
EN_lvls <- gg2 %>% filter(between(x = xmax, left = 14591, right = 28721)) %>% pull(Level2) %>% as.character()
FB_lvls <- gg2 %>% filter(between(x = xmax, left = 28722, right = 47411)) %>% pull(Level2) %>% as.character()
HL_lvls <- gg2 %>% filter(xmax > 47411) %>% pull(Level2) %>% as.character()
CM3_lvls <- gg3 %>% filter(xmax <= 14590) %>% pull(Level3) %>% as.character()
EN3_lvls <- gg3 %>% filter(between(x = xmax, left = 14591, right = 28721)) %>% pull(Level3) %>% as.character()
FB3_lvls <- gg3 %>% filter(between(x = xmax, left = 28722, right = 47411)) %>% pull(Level3) %>% as.character()
HL3_lvls <- gg3 %>% filter(xmax > 47411) %>% pull(Level3) %>% as.character()

# Define color palettes
greens <- colorRampPalette(RColorBrewer::brewer.pal(n = 9, name = "Greens"))
blues <- colorRampPalette(RColorBrewer::brewer.pal(n = 9, name = "Blues"))
reds <- colorRampPalette(RColorBrewer::brewer.pal(n = 9, name = "Reds"))
oranges <- colorRampPalette(RColorBrewer::brewer.pal(n = 9, name = "Oranges"))

all_cols <- c(setNames(c(greens(6)[4], blues(6)[4], reds(5)[4], oranges(6)[4]), nm = L1_lvls),
              setNames(greens(length(CM_lvls)), nm = CM_lvls),
              setNames(blues(length(EN_lvls)), nm = EN_lvls),
              setNames(reds(length(FB_lvls)), nm = FB_lvls),
              setNames(oranges(length(HL_lvls)), nm = HL_lvls),
              setNames(greens(length(CM3_lvls)), nm = CM3_lvls),
              setNames(blues(length(EN3_lvls)), nm = EN3_lvls),
              setNames(reds(length(FB3_lvls)), nm = FB3_lvls),
              setNames(oranges(length(HL3_lvls)), nm = HL3_lvls))

# Cast to long format
ggl <- ggc %>% 
  pivot_longer(all_of(paste0("Level", 1:3))) %>% 
  na.omit() %>% 
  filter(!(name == "Level2" & xmax > 47411)) %>% 
  mutate(ymin = case_when(name == "Level3" & xmax > 47411 ~ 1.6, TRUE ~ ymin))

p <- ggplot(ggl, aes(xmid, ymid, fill = value)) +
  geom_rect(aes(xmin = xmin, xmax = xmax, ymin = ymin, ymax = ymax)) +
  geomtextpath::geom_textpath(data = ggl %>% filter(name == "Level1"), color = "white", 
                              aes(y = ymid, label = value, group = value), size = 5) +
  geomtextpath::geom_textpath(data = ggl %>% filter(name == "Level1"),
                              aes(y = ymid - 0.7, label = paste0("n=", n), group = value), 
                              size = 4.5) +
  scale_fill_manual(values = all_cols) +
  geomtextpath::geom_textsegment(data = ggl %>% filter(name == "Level3"), size = 3.5,
                                 aes(x = xmid, xend = xmid + 1.5, y = 3.5, yend = 4.5, 
                                     label = value, group = value)) +
  scale_y_continuous(limits = c(-2, 4.5)) +
  coord_polar() +
  theme_void() +
  theme(legend.position = "none", plot.margin = margin(t = 0, r = 0, b = 0, l = 0)) +
  geom_segment(data = ggl %>% filter(name == "Level1"), color = "black",
               aes(x = xmax, xend = xmax, y = 0.6, yend = 3.4)) 

ggsave(plot = p, filename = "../results/HDCA_heart_sc_analysis_it9/annotation_layers.jpg", width = 2000, height = 2000, dpi = 300, units = "px")

```


```{r, fig.asp=1}
# Cast to long format
ggl_small <- ggc %>% 
  pivot_longer(all_of(paste0("Level", c(1, 3)))) %>% 
  select(-Level2) %>% 
  na.omit() %>% 
  mutate(ymin = case_when(name == "Level3" ~ 1.6, TRUE ~ ymin),
         ymid = case_when(name == "Level3" ~ 2, TRUE ~ ymid),
         ymax = case_when(name == "Level3" ~ 2.5, TRUE ~ ymax))

p <- ggplot(ggl_small, aes(xmid, ymid, fill = value)) +
  geom_rect(aes(xmin = xmin, xmax = xmax, ymin = ymin, ymax = ymax)) +
  geomtextpath::geom_textpath(data = ggl %>% filter(name == "Level1"), color = "white", 
                              aes(y = ymid, label = value, group = value), size = 5) +
  geomtextpath::geom_textpath(data = ggl %>% filter(name == "Level1"),
                              aes(y = ymid - 0.7, label = paste0("n=", n), group = value), 
                              size = 4.5) +
  scale_fill_manual(values = all_cols) +
  geomtextpath::geom_textsegment(data = ggl %>% filter(name == "Level3"), size = 3.5,
                                 aes(x = xmid, xend = xmid + 1.5, y = 3, yend = 4, 
                                     label = value, group = value)) +
  scale_y_continuous(limits = c(-2, 4.5)) +
  coord_polar() +
  theme_void() +
  theme(legend.position = "none", plot.margin = margin(t = 0, r = 0, b = 0, l = 0)) +
  geom_segment(data = ggl %>% filter(name == "Level1"), color = "black",
               aes(x = xmax, xend = xmax, y = 0.6, yend = 2.5)) 

ggsave(plot = p, filename = "../results/HDCA_heart_sc_analysis_it9/annotation_layers_L1_L3.jpg", width = 2000, height = 2000, dpi = 300, units = "px")

```


