---
title: "03_1__02_HDCA_stereoscope_preparation"
author: "Julia Foyer, Ludvig Larsson & RaphaÃ«l Mauron"
output: html_document
editor_options: 
  chunk_output_type: console
---

# HDCA heart - stereoscope preparation

In order to perform deconvolution using `stereoscope`, several files are required:
* single-cell annotation (HDCA_heart_sc_annotation.tsv)
* single-cell variable features (HDCA_heart_sc_var.features_5000.txt)
* single-cell count matrix (HDCA_heart_sc_cnt_data.tsv)
* spatial transcriptomics data (hdca_heart_ST_data.tsv)

Here we generate the 3 *single-cell* files required.

Note that it has been decided to perform deconvolution on the high level annotation AND on the deep level annotation. Therefore, a total of 6 files are generated here.

Here, according to reviewers comment, we prepare temporal files (category:
w6-w7
w8-w9
w10-w12)

NOTE: The path to load the data and the path to save the figures should be changed at your convenience.

Environment: **Docker**


## Settings

### Set library path

```{r setup, include=FALSE}
.libPaths("/home/rstudio/project/renv/library/R-4.3/aarch64-unknown-linux-gnu/")
knitr::opts_chunk$set(echo = TRUE)
```


### Set environment variables

```{r}
# Set the directory 
myfolder <- "/home/rstudio/project/output/HDCA_heart_sc_analysis_docker/"
```


### Load Seurat object

```{r, message = FALSE}
out_path <- "/home/rstudio/project/output/HDCA_heart_sc_analysis_docker/stereoscope/"
data <- readRDS(paste0(myfolder, "rds_objects/HDCA_heart_sc_full_extended.rds"))
```


### Load libraries

```{r, message = FALSE}
library(Seurat)
library(Matrix)
library(dplyr)
library(data.table)
library(tibble)
```



## High level clusters
HL stereoscope: exclude 6, 19, 27, 34

```{r}
data_curated <- data[, data$high_level_clusters %in% c("1", "2", "3", "4_35", "5", "6", "7", "8", "9", "10", "11", "12", "13", "14", "15", "16", "17", "18_1", "18_2", "20", "21", "22", "23", "24", "25", "26", "28", "29", "30", "31", "32", "33", "34")]
```


### Add additional age groups to the metadata (to match age groups from ST data)
In order to be unbiased, take only the single-cell from the exact age groups that we have in the ST dataset.

```{r}
df <- data.frame(age = c(5.5, 6, 7, 8, 8.5, 9, 10, 11.5, 12, 13.25, 14), age_group_ST = c("5", "6-7", "6-7", "8-9", "8-9", "8-9", "10-12", "10-12", "10-12", "13-14", "13-14"))
age_to_age_group <- setNames(df$age_group_ST, nm = df$age)
data_curated$age_group_ST <- factor(age_to_age_group[as.character(data_curated$age)], levels = c("5", "6-7", "8-9", "10-12", "13-14"))
rm(df)
gc()
```

```{r}
table(data_curated@meta.data[["age_group_ST"]])
tb <- table(data_curated@meta.data[["age_group_ST"]], data_curated@meta.data[["cell.types"]]) %>% as.matrix()

#write table cell per cluster per age
write.csv(tb, file = paste0(out_path, "age_per_HL_types.csv"))
```


### Split by age groups

```{r}
data_curated_age1 <- data_curated[, data_curated$age_group_ST == "6-7"]
data_curated_age2 <- data_curated[, data_curated$age_group_ST == "8-9"]
data_curated_age3 <- data_curated[, data_curated$age_group_ST == "10-12"]
```


### Get the 5000 most highly variable genes

Export top 5000 most variable genes for stereoscope.

```{r, message = FALSE, warning=FALSE}
data_curated_age1 <- FindVariableFeatures(data_curated_age1, nfeatures = 5000)
var.features1 <- data_curated_age1@assays$RNA@var.features

data_curated_age2 <- FindVariableFeatures(data_curated_age2, nfeatures = 5000)
var.features2 <- data_curated_age2@assays$RNA@var.features

data_curated_age3 <- FindVariableFeatures(data_curated_age3, nfeatures = 5000)
var.features3 <- data_curated_age3@assays$RNA@var.features
```


### Save variable features per age groups

```{r, message = FALSE, warning=FALSE}
# Create path if not existing
if (!dir.exists(paste0(out_path, "/age_dep_HL/"))){
  dir.create(paste0(out_path, "/age_dep_HL/"))}

file <- paste0(out_path, "age_dep_HL/HDCA_heart_sc_var.features_5000_age1.txt")
write.table(x = var.features1, file = file, quote = FALSE, row.names = FALSE, col.names = FALSE)

file <- paste0(out_path, "age_dep_HL/HDCA_heart_sc_var.features_5000_age2.txt")
write.table(x = var.features2, file = file, quote = FALSE, row.names = FALSE, col.names = FALSE)

file <- paste0(out_path, "age_dep_HL/HDCA_heart_sc_var.features_5000_age3.txt")
write.table(x = var.features3, file = file, quote = FALSE, row.names = FALSE, col.names = FALSE)
```


### Down sample cells (for each age groups)

#### Age1

```{r}
sampled_cells1 <- data_curated_age1[[]] %>% 
  tibble::rownames_to_column(var = "barcode") %>% 
  group_by(high_level_clusters) %>% 
  arrange(-nFeature_RNA) %>% 
  slice_head(n = 500)
```

#### Save the count matrix

Export count matrix for stereoscope. 

```{r, message = FALSE, warning=FALSE}
counts <- GetAssayData(data_curated_age1, slot = "counts", assay = "RNA")
counts <- counts[var.features1, sampled_cells1$barcode]

# Sanity check
all(colnames(counts) == sampled_cells1$barcode)

out_file1 <- paste0(out_path, "age_dep_HL/HDCA_heart_sc_cnt_data1.tsv")
counts_df <- counts %>% t() %>% as.matrix() %>% as.data.frame()
data.table::fwrite(counts_df, file = out_file1, sep = "\t", quote = FALSE, 
                   row.names = TRUE, col.names = TRUE)
```


#### Age2

```{r}
sampled_cells2 <- data_curated_age2[[]] %>% 
  tibble::rownames_to_column(var = "barcode") %>% 
  group_by(high_level_clusters) %>% 
  arrange(-nFeature_RNA) %>% 
  slice_head(n = 500)
```

#### Save the count matrix

Export count matrix for stereoscope. 

```{r, message = FALSE, warning=FALSE}
counts <- GetAssayData(data_curated_age2, slot = "counts", assay = "RNA")
counts <- counts[var.features2, sampled_cells2$barcode]

# Sanity check
all(colnames(counts) == sampled_cells2$barcode)

out_file2 <- paste0(out_path, "age_dep_HL/HDCA_heart_sc_cnt_data2.tsv")
counts_df <- counts %>% t() %>% as.matrix() %>% as.data.frame()
data.table::fwrite(counts_df, file = out_file2, sep = "\t", quote = FALSE, 
                   row.names = TRUE, col.names = TRUE)
```

#### Age3

```{r}
sampled_cells3 <- data_curated_age3[[]] %>% 
  tibble::rownames_to_column(var = "barcode") %>% 
  group_by(high_level_clusters) %>% 
  arrange(-nFeature_RNA) %>% 
  slice_head(n = 500)
```

#### Save the count matrix

Export count matrix for stereoscope. 

```{r, message = FALSE, warning=FALSE}
counts <- GetAssayData(data_curated_age3, slot = "counts", assay = "RNA")
counts <- counts[var.features3, sampled_cells3$barcode]

# Sanity check
all(colnames(counts) == sampled_cells3$barcode)

out_file3 <- paste0(out_path, "age_dep_HL/HDCA_heart_sc_cnt_data3.tsv")
counts_df <- counts %>% t() %>% as.matrix() %>% as.data.frame()
data.table::fwrite(counts_df, file = out_file3, sep = "\t", quote = FALSE, 
                   row.names = TRUE, col.names = TRUE)
```


### Export annotations for high level clusters

#### Age1

```{r, , message = FALSE}
annotation <- sampled_cells1 %>% 
  select(barcode, high_level_clusters) %>% 
  data.frame(row.names = 1) #data[["high_level_clusters"]]
colnames(annotation) <- "bio_celltype"

file <- paste0(out_path, "age_dep_HL/HDCA_heart_sc_annotation1.tsv")
data.table::fwrite(annotation, row.names = TRUE, file = file, sep = "\t")
```

```{r}
exprMat <- data.table::fread(out_file1, sep = "\t", header = TRUE) %>% 
  data.frame(row.names = 1) %>% 
  as.matrix()

annData <- read.table(file, header = TRUE, row.names = 1)

# Check that cells are identical in count matrix and annotation data.frame
all(rownames(exprMat) == rownames(annData))
```


#### Age2

```{r, , message = FALSE}
annotation <- sampled_cells2 %>% 
  select(barcode, high_level_clusters) %>% 
  data.frame(row.names = 1) #data[["high_level_clusters"]]
colnames(annotation) <- "bio_celltype"

file <- paste0(out_path, "age_dep_HL/HDCA_heart_sc_annotation2.tsv")
data.table::fwrite(annotation, row.names = TRUE, file = file, sep = "\t")
```

```{r}
exprMat <- data.table::fread(out_file2, sep = "\t", header = TRUE) %>% 
  data.frame(row.names = 1) %>% 
  as.matrix()

annData <- read.table(file, header = TRUE, row.names = 1)

# Check that cells are identical in count matrix and annotation data.frame
all(rownames(exprMat) == rownames(annData))
```


#### Age3

```{r, , message = FALSE}
annotation <- sampled_cells3 %>% 
  select(barcode, high_level_clusters) %>% 
  data.frame(row.names = 1) #data[["high_level_clusters"]]
colnames(annotation) <- "bio_celltype"

file <- paste0(out_path, "age_dep_HL/HDCA_heart_sc_annotation3.tsv")
data.table::fwrite(annotation, row.names = TRUE, file = file, sep = "\t")
```

```{r}
exprMat <- data.table::fread(out_file3, sep = "\t", header = TRUE) %>% 
  data.frame(row.names = 1) %>% 
  as.matrix()

annData <- read.table(file, header = TRUE, row.names = 1)

# Check that cells are identical in count matrix and annotation data.frame
all(rownames(exprMat) == rownames(annData))
```















## Deep clustering

### Keep selected cell types

Clusters to keep for stereoscope:
- HL: 6, 8, 10, 15, 18_1, 20, 22, 29, 34
to exclude since they are in deeplevel:
FB: 13, 30, 2, 23, 1, 14, 16, 31
EC: 32, 28, 12, 24, 3, 21, 33
CM: 4-35, 17, 9, 5, 11, 7, 18_2
IN: 25, 26

deep level:
-CM: 1, 2, 3, 4-1, 4-2, 5, 6, 7, 8-1, 8-2, 8-3, 9, 10, 12, 14, 17, 18, 19-1, 19-2 
-EN: 1, 2, 3, 5, 6, 7, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21
-FB: 1, 2, 3, 4, 5, 6, 7, 8, 9, 11, 12, 13, 14, 15, 17, 18, 19, 20
-IN: 1, 3, 4, 7, 8, 11, 12, 13, 17, 19

cluster to exclude:
- CM 11, 13, 15, 16, 20, 21, 22
- EN 4, 8, 9, 22
- FB 10, 16, 21, 22
- IN 2, 5, 6, 9, 10, 14, 15, 16, 18 

### Get annotation

```{r}
cells <- data_curated[[]] %>% 
  tibble::rownames_to_column(var = "barcode") %>% 
  group_by(high_level_clusters) %>% 
  arrange(-nFeature_RNA)

annotation <- cells %>% 
  select(barcode, high_level_clusters) %>% 
  data.frame(row.names = 1)
colnames(annotation) <- "bio_celltype"
```


```{r}
# Load annotations prepared earlier and filter
# CM
cardiomyocyte_clusters_keep <- c("1", "2", "3", "4-1", "4-2", "5", "6", "7", "8-1", "8-2", "8-3", "9", "10", "12", "14", "17", "18", "19-1", "19-2")
cardiomyocyte_metadata <- readRDS(paste0(myfolder, "rds_objects/cardiomyocyte_metadata.rds")) %>% 
  mutate(clusters_subset = paste0("CM_", as.character(clusters_subset))) %>%  # rename cardiomyocyte clusters by adding a prefix CM_
  mutate_if(is.factor, as.character) %>% 
  filter(clusters_subset_split %in% cardiomyocyte_clusters_keep)

# EN
endothelial_clusters_keep <- c("1", "2", "3", "5", "6", "7", "10", "11", "12", "13", "14", "15", "16", "17", "18", "19", "20", "21")
endothelial_metadata <- readRDS(paste0(myfolder, "rds_objects/endothelial_metadata.rds")) %>% 
  filter(clusters_subset %in% endothelial_clusters_keep) %>% 
  mutate(clusters_subset = paste0("EN_", as.character(clusters_subset))) # rename endothelial clusters by adding a prefix EN_

# FB
fibroblasts_clusters_keep <- c("1", "2", "3", "4", "5", "6", "7", "8", "9", "11", "12", "13", "14", "15", "17", "18", "19", "20")
fibroblasts_metadata <- readRDS(paste0(myfolder, "rds_objects/fibroblasts_metadata.rds")) %>% 
  filter(clusters_subset %in% fibroblasts_clusters_keep) %>% 
  mutate(clusters_subset = paste0("FB_", as.character(clusters_subset))) # rename fibroblasts clusters by adding a prefix FB_

# IN
innervation_clusters_keep <- c("1", "3", "4", "7", "8", "11", "12", "13", "17", "19")
innervation_metadata <- readRDS(paste0(myfolder, "rds_objects/innervation_metadata.rds")) %>% 
  filter(clusters_subset %in% innervation_clusters_keep) %>% 
  mutate(clusters_subset = paste0("IN_", as.character(clusters_subset))) # rename innervation clusters by adding a prefix IN_

# Merge deep level clusters from the three different subsets
deep_clusters <- cardiomyocyte_metadata %>% 
  select(clusters_subset, clusters_subset_split) %>% # Only select clustering results
  bind_rows(endothelial_metadata %>% select(clusters_subset)) %>% # Add endothelial clusters
  bind_rows(fibroblasts_metadata %>% select(clusters_subset)) %>% # Add fibroblast clusters
  bind_rows(innervation_metadata %>% select(clusters_subset)) %>% # Add innervation clusters
  mutate(clusters_subset_split = case_when(is.na(clusters_subset_split) ~ clusters_subset, # rename cardiomyocyte clusters by adding a prefix CM_
                                           TRUE ~ paste0("CM_", clusters_subset_split)))

# Define clusters to keep
keep_clusters <- c(paste0("HL_", c("6", "8", "10", "15", "18_1", "20", "22", "29", "34")), 
                   paste0("CM_", cardiomyocyte_clusters_keep),
                   paste0("EN_", endothelial_clusters_keep),
                   paste0("FB_", fibroblasts_clusters_keep),
                   paste0("IN_", innervation_clusters_keep))


# Load annotation file
# file <- paste0(out_path, "high_level/HDCA_heart_sc_annotation.tsv")
# annotation <- read.table(file, header = TRUE, row.names = 1)

# Create a new annotations tibble
annotation_split <- annotation 
annotation_split <- annotation_split %>% 
  rownames_to_column(var = "ID") %>% 
  left_join(y = deep_clusters %>% rownames_to_column(var = "ID"), by = "ID") %>% 
  mutate(clusters_subset_split = case_when(is.na(clusters_subset_split) ~ paste0("HL_", bio_celltype),
                                           TRUE ~ clusters_subset_split)) %>% 
  filter(clusters_subset_split %in% keep_clusters) # keep selected clusters

# Convert clusters_subset_split to a factor and set order of clusters
annotation_split <- annotation_split %>% 
  mutate(clusters_subset_split = factor(clusters_subset_split, keep_clusters))

#add age group
age_group_ST <- as.data.frame(data_curated$age_group_ST)
names(age_group_ST) <- "age_group_ST"
annotation_split <- annotation_split %>% 
  left_join(y = age_group_ST %>% rownames_to_column(var = "ID"), by = "ID")

# Check overlap (sanity check)
all(annotation$ID %in% colnames(data))

# Check which cells we keep
table(annotation_split$clusters_subset_split)
```

```{r}
tb <- table(annotation_split$age_group_ST, annotation_split$clusters_subset_split) %>% as.matrix()

#write table cell per cluster per age
write.csv(tb, file = paste0(out_path, "age_per_DL_types.csv"))

#write table annotation_split
write.csv(annotation_split, file = paste0(out_path, "annotation_split.csv"))
```


#### Export files for sampled cells

Keep a maximum of 250 cells per cell type to speed up computation. Select the top 250 cells with highest numbers of nFeature_RNA (unique genes).

```{r}
# Create path if not existing
if (!dir.exists(paste0(out_path, "/deep_level_2/"))){
  dir.create(paste0(out_path, "/deep_level_2/"))}

# Down sampling
clusters_split_cells <- annotation_split %>% 
  left_join(y = data[[]] %>% 
              rownames_to_column(var = "ID") %>% 
              select(ID, sampleID, nFeature_RNA), by = "ID") %>% 
  group_by(clusters_subset_split) %>% # Group by deep cluster
  arrange(-nFeature_RNA) %>% # Arrange by decreasing number of unique genes
  slice_head(n = 250) %>% # Select top 250 cells (highest quality cells based on number of unique genes)
  arrange(clusters_subset_split) # Arrange cells by cluster factor levels
  
# Subset Seurat object using the cell IDs created above
deep_level_subset_data <- data[, clusters_split_cells$ID]

# Find top variable features for selected cells
deep_level_subset_data <- FindVariableFeatures(deep_level_subset_data, nfeatures = 5000)

# Export variable genes
file <- paste0(out_path, "deep_level_2/HDCA_heart_sc_var_features_deep_level_5000.txt")
write.table(x = VariableFeatures(deep_level_subset_data), 
            file = file, quote = FALSE, row.names = FALSE, col.names = FALSE)

# Subset counts to include top 5000 most variable features (no need to export all genes)
counts <- GetAssayData(deep_level_subset_data, slot = "counts")[VariableFeatures(deep_level_subset_data), ]

# Export counts
out_file <- paste0(out_path, "deep_level_2/HDCA_heart_sc_cnt_data_deep_level.tsv")
counts_df <- counts %>% t() %>% as.matrix() %>% as.data.frame()
data.table::fwrite(counts_df, file = out_file, sep = "\t", quote = FALSE, 
                   row.names = TRUE, col.names = TRUE)

# Check that cell IDs match (sanity check)
all(clusters_split_cells$ID == colnames(counts))

# Export annotations for selected cells
file <- paste0(out_path, "deep_level_2/HDCA_heart_sc_annotation_deep_level.tsv")
data.table::fwrite(clusters_split_cells %>%
                     column_to_rownames(var = "ID") %>%
                     select(clusters_subset_split) %>%
                     dplyr::rename(bio_celltype = clusters_subset_split),
       row.names = TRUE, file = file, sep = "\t")
```


#### Sanity check

```{r}
exprMat <- data.table::fread(out_file) %>% data.frame(row.names = 1, check.names = FALSE) %>% as.matrix()
var_genes <- readLines(paste0(out_path, "deep_level_2/HDCA_heart_sc_var_features_deep_level_5000.txt"))

# Check variable features
all(var_genes == colnames(exprMat))

# Check cell IDs
ann <- read.table(paste0(out_path, "deep_level_2/HDCA_heart_sc_annotation_deep_level.tsv"), header = T, row.names = 1)
all(rownames(ann) == rownames(exprMat))
```


================================================================================
================================================================================
================================================================================
================================================================================
================================================================================







