---
title: "02_1_HDCA_heart_cardiomyocyte_analysis"
author: "Julia Foyer, Ludvig Larsson & RaphaÃ«l Mauron"
output: html_document
editor_options: 
  chunk_output_type: console
---

# HDCA heart cardiomyocyte analysis

In this script, a subset of the *high level* data generated in the script `01_HDCA_heart_QC_and_high_level_analysis.Rmd` is selected and re-clustered and analyzed accordingly. It is part of the *deep level* analysis.

NOTE: The path to load the data and the path to save the figures should be changed at your convenience.

Environment: **Docker**


## Settings

### Set library path

```{r setup, include=FALSE}
.libPaths("/home/rstudio/project/renv/library/R-4.3/aarch64-unknown-linux-gnu/")
knitr::opts_chunk$set(echo = TRUE)
```


### Set environment variables

```{r}
# Set the directory 
myfolder <- "/home/rstudio/project/output/HDCA_heart_sc_analysis_docker/"
```


### Load Libraries

```{r, message = FALSE}
library(Seurat)
library(Matrix)
library(dplyr)
library(harmony)
library(ggplot2)
library(patchwork)
library(openxlsx)
library(niceRplots)
library(gtools)
library(patchwork)
library(plotly)
```


### Load global functions

```{r}
source("/home/rstudio/project/code/global_functions.R")
```


### Load filtered data

```{r}
data <- readRDS(paste0(myfolder, "rds_objects/HDCA_heart_sc_full_extended.rds"))
```


### Set seed

```{r}
set.seed(1)
```


## Subset, process, clustering

* Keep all cells annotated as cardiomyocyte
* Calculate variable features
* Scale data while regressing out "nCount_RNA", "nFeature_RNA", "percent_ribo", "percent_mito", "percent_hb", "percent_hsp", "S.Score" and "G2M.Score
* Run PCA on scaled data (50 PCs)
* Run harmony (batch correction, 50 PCs)
* Compute UMAP (50 PCs)
* Construct knn adjacency matrix with hnsw_knn
* Cluster cells using cluster_louvain

```{r}
# Subset the data to keep only CMs
CM <- data[, data$cell.types == "Ventricular CM_1" | #4_35
             data$cell.types == "Ventricular CM_3" | #5
             data$cell.types == "Proliferating CM" | #7
             data$cell.types == "Ventricular CM_2" | #9
             data$cell.types == "Atrial CM_2" |      #11
             data$cell.types == "Atrial CM_1" |      #17
             data$cell.types == "Immature CM"        #18_2
           ]

# Remove all genes with 0 counts
CM <- CM[rowSums(GetAssayData(CM, slot = "counts")) > 0, ]

# Process the data
CM <- CM %>% 
  NormalizeData() %>% 
  FindVariableFeatures(nfeatures = 3000) %>% 
  ScaleData(vars.to.regress = c("nCount_RNA", "nFeature_RNA", 
                                "percent_ribo", "percent_mito", 
                                "percent_hb", "percent_hsp", 
                                "S.Score", "G2M.Score"))

CM <- RunPCA(CM, npcs = 50, seed.use = 42)

set.seed(20)
CM <- RunHarmony(CM,
                 dims.use = 1:50,
                 group.by.vars = c("sampleID"),
                 reduction = "pca",
                 project.dim = FALSE,
                 reduction.save = "harmony_subset",
                 seed.use = 42)

CM <- RunUMAP(CM,
              dims = 1:50, # 30 
              reduction = "harmony_subset",
              n.neighbors = 30,
              n.epochs = 200, 
              min.dist = 0.5,
              n.threads = 7,
              seed.use = 42,
              reduction.key = "subset",
              reduction.name = "umap_subset")
gc()

# Compute k nearest neighbours
set.seed(1)
nn <- CreateNNAdjacencyMatrix(data = CM, reduction = "harmony_subset", dims = 1:50, k = 30)

# Cluster adjacency matrix
cl <- LouvainClustering(nn, seed = 1, resolution = 1.35)
CM$clusters_subset <- factor(cl)


pdf(paste0(myfolder, "cardiomyocytes/cardiomyocyte_subclusters.pdf"), width = 10, height = 10)
DimPlot(CM, reduction = "umap_subset", group.by = "clusters_subset", label = TRUE) + NoLegend() + labs(title = "Cardiomyocyte subclusters")
dev.off()

pdf(paste0(myfolder, "cardiomyocytes/cardiomyocyte_subclusters_cell_cycle.pdf"), width = 10, height = 10)
DimPlot(CM, reduction = "umap_subset", group.by = "Phase", label = F) + labs(title = "Cardiomyocyte subclusters")
dev.off()

gc()
```


#### DEG analysis prior subclustering some CM clusters

```{r, eval=FALSE}
CM <- SetIdent(CM, value = "clusters_subset")
de_markers <- FindAllMarkers(CM, 
                             logfc.threshold = 0.25,
                             only.pos = TRUE, 
                             min.pct = 0.05)

openxlsx::write.xlsx(de_markers %>% filter(p_val_adj < 0.01),
                     file = paste0(myfolder, "cardiomyocytes/detable_cardiomyocytes.xlsx"))
```


## Export cardiomyocyte object prior subclustering

```{r}
saveRDS(CM, file = paste0(myfolder, "rds_objects/cardiomyocytes01.rds"))
```

```{r}
CM <- readRDS(file = paste0(myfolder, "rds_objects/cardiomyocytes01.rds"))
```


### Subcluster (cluster 4)

```{r}
cl4 <- CM[, as.character(CM$clusters_subset) == "4"] %>% 
  RunUMAP(dims = 1:50,
          reduction = "harmony_subset",
          n.neighbors = 30,
          n.epochs = 100,
          n.threads = 7,
          reduction.key = "subset",
          reduction.name = "umap_subset_cl4",
          seed.use = 42)

# Compute k nearest neighbours
set.seed(1)
nn <- CreateNNAdjacencyMatrix(data = cl4, reduction = "harmony_subset", k = 30, dims = 1:50)

# Cluster adjacency matrix
cl <- LouvainClustering(nn = nn, resolution = 0.5)
cl4$clusters_subset_cl4 <- factor(cl)

pdf(paste0(myfolder, "cardiomyocytes/cardiomyocyte_subclusters_subset_cl4.pdf"), width = 10, height = 10)
DimPlot(cl4, reduction = "umap_subset_cl4", group.by = "clusters_subset_cl4", label = TRUE) + NoLegend()
dev.off()
```

```{r}
cl4 <- SetIdent(cl4, value = "clusters_subset_cl4")
de_markers_cl4 <- FindAllMarkers(cl4, only.pos = TRUE, logfc.threshold = 0.5) %>% 
  filter(p_val_adj < 0.01)

openxlsx::write.xlsx(de_markers_cl4 %>% filter(p_val_adj < 0.01),
                     file = paste0(myfolder, "cardiomyocytes/detable_cardiomyocytes_cl4.xlsx"))

p <- DotPlot(cl4, features = de_markers_cl4 %>% group_by(cluster) %>% 
          slice_head(n = 10) %>% pull(gene) %>% unique()) +
  coord_flip() +  # This flips the plot vertically
  theme(axis.text.y = element_text(angle = 0, hjust = 1, vjust = 0.5)) +
  scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 11, name = "RdBu") %>% rev()) +
  scale_x_discrete(limits = rev) +
  labs(x = "Genes", y = "Cluster")  # Flipped labels

pdf(paste0(myfolder, "cardiomyocytes/dotplot_de_genes_cl4.pdf"), width = 10, height = 7)
print(p)
dev.off()
```


### Subcluster 8 (cluster 8)

```{r}
cl8 <- CM[, as.character(CM$clusters_subset) == "8"] %>% 
  RunUMAP(dims = 1:50,
          reduction = "harmony_subset",
          n.neighbors = 30,
          n.epochs = 100,
          n.threads = 7,
          reduction.key = "subset",
          reduction.name = "umap_subset_cl8",
          seed.use = 42)

# Compute k nearest neighbours
set.seed(1)
nn <- CreateNNAdjacencyMatrix(data = cl8, reduction = "harmony_subset", k = 30, dims = 1:50)

# Cluster adjacency matrix
cl <- LouvainClustering(nn = nn, resolution = 0.5)
cl8$clusters_subset_cl8 <- factor(cl)

pdf(paste0(myfolder, "cardiomyocytes/cardiomyocyte_subclusters_subset_cl8.pdf"), width = 10, height = 10)
DimPlot(cl8, reduction = "umap_subset_cl8", group.by = "clusters_subset_cl8", label = TRUE) + NoLegend()
dev.off()
```

```{r}
cl8 <- SetIdent(cl8, value = "clusters_subset_cl8")
de_markers_cl8 <- FindAllMarkers(cl8, only.pos = TRUE, logfc.threshold = 0.5) %>% 
  filter(p_val_adj < 0.01)

openxlsx::write.xlsx(de_markers_cl8 %>% filter(p_val_adj < 0.01),
                     file = paste0(myfolder, "cardiomyocytes/detable_cardiomyocytes_cl8.xlsx"))

p <- DotPlot(cl8, features = de_markers_cl8 %>% group_by(cluster) %>% 
          slice_head(n = 10) %>% pull(gene) %>% unique()) +
  coord_flip() +  # This flips the plot vertically
  theme(axis.text.y = element_text(angle = 0, hjust = 1, vjust = 0.5)) +
  scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 11, name = "RdBu") %>% rev()) +
  scale_x_discrete(limits = rev) +
  labs(x = "Genes", y = "Cluster")  # Flipped labels

pdf(paste0(myfolder, "cardiomyocytes/dotplot_de_genes_cl8.pdf"), width = 10, height = 7)
print(p)
dev.off()
```


### Subcluster (cluster 19)

```{r}
cl19 <- CM[, as.character(CM$clusters_subset) == "19"] %>% 
  RunUMAP(dims = 1:50,
          reduction = "harmony_subset",
          n.neighbors = 30,
          n.epochs = 100,
          n.threads = 7,
          reduction.key = "subset",
          reduction.name = "umap_subset_cl19")

# Compute k nearest neighbours
nn <- CreateNNAdjacencyMatrix(data = cl19, reduction = "harmony_subset", k = 30, dims = 1:50)

# Create an undirected graph based on the neighbor matrix.
cl <- LouvainClustering(nn = nn, resolution = 0.5)
cl19$clusters_subset_cl19 <- factor(cl)

pdf(paste0(myfolder, "cardiomyocytes/cardiomyocyte_subclusters_subset_cl19.pdf"), width = 10, height = 10)
DimPlot(cl19, reduction = "umap_subset_cl19", group.by = "clusters_subset_cl19", label = TRUE) + NoLegend()
dev.off()
```

```{r}
cl19 <- SetIdent(cl19, value = "clusters_subset_cl19")
de_markers_cl19 <- FindAllMarkers(cl19, only.pos = TRUE, logfc.threshold = 0.5) %>% 
  filter(p_val_adj < 0.01)

openxlsx::write.xlsx(de_markers_cl19 %>% filter(p_val_adj < 0.01),
                     file = paste0(myfolder, "cardiomyocytes/detable_cardiomyocytes_cl19.xlsx"))

p <- DotPlot(cl19, features = de_markers_cl19 %>% group_by(cluster) %>% 
          slice_head(n = 10) %>% pull(gene) %>% unique()) +
  coord_flip() +  # This flips the plot vertically
  theme(axis.text.y = element_text(angle = 0, hjust = 1, vjust = 0.5)) +
  scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 11, name = "RdBu") %>% rev()) +
  scale_x_discrete(limits = rev) +
  labs(x = "Genes", y = "Cluster")  # Flipped labels

pdf(paste0(myfolder, "cardiomyocytes/dotplot_de_genes_cl19.pdf"), width = 10, height = 7)
print(p)
dev.off()
```


### Add cl4, cl8 and cl19 subclusters

```{r}
cl4_split <- cl4[[]] %>% 
  select(clusters_subset_cl4) %>% 
  rename(clusters_subset_split = clusters_subset_cl4)
cl8_split <- cl8[[]] %>% 
  select(clusters_subset_cl8) %>% 
  rename(clusters_subset_split = clusters_subset_cl8)
cl19_split <- cl19[[]] %>% 
  select(clusters_subset_cl19) %>% 
  rename(clusters_subset_split = clusters_subset_cl19)

all_split <- bind_rows(cl4_split, cl8_split) %>% 
  bind_rows(cl19_split)

new_ids <- CM[[]] %>% select(clusters_subset)
new_ids[, "clusters_subset_split"] <- all_split[colnames(CM), 
                                                "clusters_subset_split", drop = TRUE]
new_ids <- new_ids %>% 
  mutate_if(is.factor, as.character) %>% 
  mutate(clusters_subset_split = case_when(!is.na(clusters_subset_split) ~ paste0(clusters_subset, "-", clusters_subset_split),
                                           TRUE ~ clusters_subset))

CM$clusters_subset_split <- new_ids$clusters_subset_split


p <- DimPlot(CM, reduction = "umap_subset", group.by = "clusters_subset_split", label = TRUE) +
  NoLegend() +
  labs(title = "Cardiomyocyte subclusters")

pdf(file = paste0(myfolder, "cardiomyocytes/clusters_subsetted_and_split_cardiomyocytes_umap.pdf"), 
     width = 8, height = 8)
p
dev.off()
```


## Run DEG for all clusters

```{r}
CM <- SetIdent(CM, value = "clusters_subset_split")
de_markers_split <- FindAllMarkers(CM, logfc.threshold = 0.25,
                             only.pos = TRUE, 
                             min.pct = 0.05)

openxlsx::write.xlsx(de_markers_split %>% filter(p_val_adj < 0.01),
                     file = paste0(myfolder, "cardiomyocytes/detable_cardiomyocytes_split.xlsx"))
```


### Save images and data

```{r}
# CM subclusters projected on high level UMAP
p <- DimPlot(CM, reduction = "umap_oi", group.by = "clusters_subset_split", label = TRUE) + 
  labs(title = "Subclustered cardiomyocytes - original UMAP")
p$data$clusters_subset_split <- factor(p$data$clusters_subset_split, levels = mixedsort(unique(CM@meta.data[["clusters_subset_split"]])))

pdf(file = paste0(myfolder, "cardiomyocytes/clusters_cardiomyocytes_big_umap.pdf"), 
     width = 10, height = 10)
p
dev.off()

# High level CM clusters projected on the deep level CM UMAP or Rough clusters
pdf(file = paste0(myfolder, "cardiomyocytes/rough_clusters_cardiomyocytes.pdf"), 
     width = 10, height = 10)
DimPlot(CM, reduction = "umap_subset", group.by = "clusters_louvain_oi") + 
  labs(title = "Rough clusters - CM UMAP")
dev.off()

# nCounts
pdf(paste0(myfolder, "cardiomyocytes/nCounts_cardiomyocytes.pdf"), width = 10, height = 10)
FeaturePlot(CM, reduction = "umap_subset", features = "nCount_RNA", raster = FALSE)
dev.off()

# nFeatures
pdf(paste0(myfolder, "cardiomyocytes/nFeatures_cardiomyocytes.pdf"), width = 10, height = 10)
FeaturePlot(CM, reduction = "umap_subset", features = "nFeature_RNA", raster = FALSE)
dev.off()
```


### Cell type proportion by age group in entire CM object

```{r}
age1 <- CM[,CM$age <= 6]
age2 <- CM[,CM$age == 7 | CM$age == 8 | CM$age == 8.5]
age3 <- CM[,CM$age == 9 | CM$age == 10 | CM$age == 11.5]
age4 <- CM[,CM$age >= 12]

CM_age <- c(age1, age2, age3, age4)
rm(age1, age2, age3, age4)

# order the subclusters
CM_age[[1]]@meta.data$clusters_subset_split <- factor(CM_age[[1]]@meta.data$clusters_subset_split, 
                                                      levels = selected_cardiomyocyte_clusters)
CM_age[[2]]@meta.data$clusters_subset_split <- factor(CM_age[[2]]@meta.data$clusters_subset_split, 
                                                      levels = selected_cardiomyocyte_clusters)
CM_age[[3]]@meta.data$clusters_subset_split <- factor(CM_age[[3]]@meta.data$clusters_subset_split, 
                                                      levels = selected_cardiomyocyte_clusters)
CM_age[[4]]@meta.data$clusters_subset_split <- factor(CM_age[[4]]@meta.data$clusters_subset_split, 
                                                      levels = selected_cardiomyocyte_clusters)

# create df by age grouping the number of cell per cluster
cell_by_cluster_entire_1 <- as.data.frame(table(CM_age[[1]]@meta.data[["clusters_subset_split"]]))
cell_by_cluster_entire_2 <- as.data.frame(table(CM_age[[2]]@meta.data[["clusters_subset_split"]]))
cell_by_cluster_entire_3 <- as.data.frame(table(CM_age[[3]]@meta.data[["clusters_subset_split"]]))
cell_by_cluster_entire_4 <- as.data.frame(table(CM_age[[4]]@meta.data[["clusters_subset_split"]]))

# add the proportion of each cell type
cell_by_cluster_entire_1$proportion <- cell_by_cluster_entire_1$Freq/sum(cell_by_cluster_entire_1$Freq)
cell_by_cluster_entire_2$proportion <- cell_by_cluster_entire_2$Freq/sum(cell_by_cluster_entire_2$Freq)
cell_by_cluster_entire_3$proportion <- cell_by_cluster_entire_3$Freq/sum(cell_by_cluster_entire_3$Freq)
cell_by_cluster_entire_4$proportion <- cell_by_cluster_entire_4$Freq/sum(cell_by_cluster_entire_4$Freq)

# export the df as .csv
write.csv2(cell_by_cluster_entire_1, paste0(myfolder, "cardiomyocytes/cell_by_cluster_entire_1.csv"))
write.csv2(cell_by_cluster_entire_2, paste0(myfolder, "cardiomyocytes/cell_by_cluster_entire_2.csv"))
write.csv2(cell_by_cluster_entire_3, paste0(myfolder, "cardiomyocytes/cell_by_cluster_entire_3.csv"))
write.csv2(cell_by_cluster_entire_4, paste0(myfolder, "cardiomyocytes/cell_by_cluster_entire_4.csv"))
```


### Cluster remaining clusters

```{r}
values_to_keep <- c("6",  "1", "10", "8-1", "8-3", "17", "8-2", "18", "4-2",  "4-1", "3", "14",  "9",  "12", "19-2",  "19-1", "7",  "2", "5")

data_CM <- CM[, CM$clusters_subset_split %in% values_to_keep]

pdf(file = paste0(myfolder, "cardiomyocytes/cardiomyocytes_clusters_remaining.pdf"), width = 10, height = 10)
DimPlot(data_CM, reduction = "umap_subset", group.by = "clusters_subset_split", label = T) + NoLegend() + labs(title = "Cardiomyocyte subclusters")
dev.off()
```


### DEG CM - top5 ordered

```{r}
values_to_keep <- c("19-2", "19-1", "12", "9", "14", "4-2", "4-1", "18", "8-2", "8-1", "8-3", "17", "6", "1", "10", "3", "7", "2", "5")

data_CM <- CM[, CM$clusters_subset_split %in% values_to_keep]

# Select a new samples size for each cluster, to get groups that are equally (or almost equally) large.
sample_size <- table(data_CM$clusters_subset_split)
sample_size[sample_size > 150] <- 150
sample_size

# Randomly sample the cells that will be included when calculating DGE (same number of cells as selected above).
# Samples without replacement, so each cell only can be included once.
DGE_cells <- lapply(names(sample_size), function(x){ 
  set.seed(1)
  sample(colnames(data_CM)[data_CM$clusters_subset_split == x], size = sample_size[x])
  })
DGE_cells <- unlist(DGE_cells)

# Select all data that relates to the sampled DGE cells.
DGE_DATA <- data_CM[, DGE_cells]

# reorder the cell types desired
DGE_DATA$clusters_subset_split <- factor(DGE_DATA$clusters_subset_split, 
                                              levels = c("19-2", "19-1", "12", "9", "14", "4-2", "4-1", "18", "8-2", "8-1", "8-3", "17", "6", "1", "10", "3", "7", "2", "5"))

#Find differential expressed genes based on groups defined by "clusters_subset_split".
DGE_DATA <- SetIdent(DGE_DATA, value = "clusters_subset_split")

# The logfc threshold is set to 0.1 as compared to the default 0.25, which makes the detection of weaker signals possible (but it takes more computational power).
detable <- FindAllMarkers(DGE_DATA,
                          only.pos = TRUE,
                          max.cells.per.ident = 300,
                          logfc.threshold = .1,
                          assay = "RNA",
                          min.pct = 0.05)

# Only keep DEG that have a p-value below 0.05.
detable <- detable[detable$p_val < 0.05, ]

# Adding two extra columns giving the difference (and log difference) between pct1 and pct2.
# pct1 is the percentage of cells in the specific cluster that express this gene.
# pct2 is the percentage of cells in all other clusters that express this gene.
detable$pct.diff <- detable$pct.1 - detable$pct.2
detable$log.pct.diff <- log2((detable$pct.1*99+1) / (detable$pct.2*99+1))


# Pick the 40 genes with the lowest p value for each cluster, then 20 highest based on pct.diff, then 5 highest based on avg_log2FC.
detable %>%
  group_by(cluster) %>%
  top_n(-40, p_val) %>%
  top_n(20, pct.diff) %>%
  top_n(5, avg_log2FC) ->
  top5

pdf(paste0(myfolder, "cardiomyocytes/DEG_top5_CM_ordered_new.pdf"), width = 10, height = 17)
par(mfrow=c(1,1), mar=c(3,6,2,5))
DotPlot(DGE_DATA, features = top5 %>% pull(gene) %>% unique()) +
  coord_flip() +
  theme(axis.text.y = element_text(angle = 0, hjust = 1, vjust = 0.5)) +
  scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 11, name = "RdBu") %>% rev()) +
  scale_x_discrete(limits = rev) +
  scale_y_discrete() +
  labs(x = "", y = "") +
  theme(axis.text.x = element_text(angle = 30, hjust = 0.5))
dev.off()

gc()
rm(detable, DGE_DATA, top5, DGE_cells, sample_size)
```


### DEG CM selection of markers (not split & all)

```{r}
# specific list of marker to project the dotplot on
my_markers <- c("MYH7", "MYH6", "MYBPC3", "TNNI1", "MYOZ2", "FABP3", "TMSB10", "S100A11", "GNG11", "FABP4", "VWF", "CDH5", "PECAM1", "LYVE1", "NPR3", "BMPER", "POSTN", "APCDD1", "TWIST1", "DDR2", "TCF21", "COL1A1", "MYH11", "ENPEP", "ITLN1", "PCSK1N", "MPZ", "C1QC", "JCHAIN", "CD3E", "KLRB1", "UBE2C", "H1-1")

CM <- SetIdent(CM, value = "clusters_subset")

p <- DotPlot(CM, features = my_markers) +
  theme(axis.text.y = element_text(angle = 0, hjust = 1, vjust = 0.5)) +
  scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 11, name = "RdBu") %>% rev()) +
  scale_x_discrete() +
  scale_y_discrete(limits = rev) +
  labs(x = "", y = "") + 
  theme(axis.text.x = element_text(angle = 30, hjust = 1, size = 8))

pdf(paste0(myfolder, "cardiomyocytes/dotplot_markers_CM_subset_not_split.pdf"), width = 10, height = 6)
print(p)
dev.off()

rm(p)
gc()
```


### DEG CM selection of VCS markers

```{r}
values_to_keep <- c("18", "4-2", "4-1", "6")

data_CM <- CM[, CM$clusters_subset_split %in% values_to_keep]

# specific list of marker to project the dotplot on
my_markers <- c("RCAN1", "IRX1", "IRX2", "IRX5", "BRINP3", "SEMA3A", "IRX3","CSMD1", "PLXNA4", "GJA1")

data_CM$clusters_subset_split <- factor(data_CM$clusters_subset_split, 
                                              levels = c("18", "4-2", "4-1", "6"))

data_CM <- SetIdent(data_CM, value = "clusters_subset_split")

p <- DotPlot(data_CM, features = my_markers) +
  theme(axis.text.y = element_text(angle = 0, hjust = 1, vjust = 0.5)) +
  scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 11, name = "RdBu") %>% rev()) +
  scale_x_discrete() +
  scale_y_discrete(limits = rev) +
  labs(x = "", y = "") + 
  theme(axis.text.x = element_text(angle = 30, hjust = 1, size = 8))

pdf(paste0(myfolder, "cardiomyocytes/dotplot_markers_CM_subset_selection_18_4-2_4-1_6.pdf"), width = 5, height = 3)
print(p)
dev.off()

rm(p)
gc()
```


#### DEG table

```{r}
values_to_keep <- c("18", "4-2", "4-1", "6")

data_CM <- CM[, CM$clusters_subset_split %in% values_to_keep]

# Select a new samples size for each cluster, to get groups that are equally (or almost equally) large.
sample_size <- table(data_CM$clusters_subset_split)
sample_size[sample_size > 150] <- 150
sample_size

# Randomly sample the cells that will be included when calculating DGE (same number of cells as selected above).
# Samples without replacement, so each cell only can be included once.
DGE_cells <- lapply(names(sample_size), function(x){ 
  set.seed(1)
  sample(colnames(data_CM)[data_CM$clusters_subset_split == x], size = sample_size[x])
  })
DGE_cells <- unlist(DGE_cells)

# Select all data that relates to the sampled DGE cells.
DGE_DATA <- data_CM[, DGE_cells]

# reorder the cell types desired
DGE_DATA$clusters_subset_split <- factor(DGE_DATA$clusters_subset_split, 
                                              levels = c("18", "4-2", "4-1", "6"))

#Find differential expressed genes based on groups defined by "clusters_subset_split".
DGE_DATA <- SetIdent(DGE_DATA, value = "clusters_subset_split")

# The logfc threshold is set to 0.1 as compared to the default 0.25, which makes the detection of weaker signals possible (but it takes more computational power).
detable <- FindAllMarkers(DGE_DATA,
                          only.pos = TRUE,
                          max.cells.per.ident = 300,
                          logfc.threshold = .1,
                          assay = "RNA",
                          min.pct = 0.05)

# Only keep DEG that have a p-value below 0.05.
detable <- detable[detable$p_val < 0.05, ]

# Adding two extra columns giving the difference (and log difference) between pct1 and pct2.
# pct1 is the percentage of cells in the specific cluster that express this gene.
# pct2 is the percentage of cells in all other clusters that express this gene.
detable$pct.diff <- detable$pct.1 - detable$pct.2
detable$log.pct.diff <- log2((detable$pct.1*99+1) / (detable$pct.2*99+1))

# Save the detable as a csv file.
write.csv2(detable, paste0(myfolder, "cardiomyocytes/detable_18_4-2_4-1_6.csv"))

rm(detable, DGE_DATA, DGE_cells, sample_size)
```


### DEG CM selection of ACS markers

```{r}
values_to_keep <- c("19-2", "19-1", "12", "9", "14")

data_CM <- CM[, CM$clusters_subset_split %in% values_to_keep]

# specific list of marker to project the dotplot on
my_markers <- c("SHOX2", "PRRX1", "TBX18",  "ZNF385B", "TENM2", "TENM3", "TENM4", "GDF10", "SLIT2", "ADAMTS19", "NRXN3")

data_CM$clusters_subset_split <- factor(data_CM$clusters_subset_split, 
                                              levels = c("19-2", "19-1", "12", "9", "14"))

data_CM <- SetIdent(data_CM, value = "clusters_subset_split")

p <- DotPlot(data_CM, features = my_markers) +
  theme(axis.text.y = element_text(angle = 0, hjust = 1, vjust = 0.5)) +
  scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 11, name = "RdBu") %>% rev()) +
  scale_x_discrete() +
  scale_y_discrete(limits = rev) +
  labs(x = "", y = "") + 
  theme(axis.text.x = element_text(angle = 30, hjust = 1, size = 8))

pdf(paste0(myfolder, "cardiomyocytes/dotplot_markers_CM_subset_selection_19-2_19-1_12_9_14.pdf"), width = 7, height = 3.5)
print(p)
dev.off()

rm(p)
gc()
```


#### DEG table

```{r}
values_to_keep <- c("19-2", "19-1", "12", "9", "14")

data_CM <- CM[, CM$clusters_subset_split %in% values_to_keep]

# Select a new samples size for each cluster, to get groups that are equally (or almost equally) large.
sample_size <- table(data_CM$clusters_subset_split)
sample_size[sample_size > 150] <- 150
sample_size

# Randomly sample the cells that will be included when calculating DGE (same number of cells as selected above).
# Samples without replacement, so each cell only can be included once.
DGE_cells <- lapply(names(sample_size), function(x){ 
  set.seed(1)
  sample(colnames(data_CM)[data_CM$clusters_subset_split == x], size = sample_size[x])
  })
DGE_cells <- unlist(DGE_cells)

# Select all data that relates to the sampled DGE cells.
DGE_DATA <- data_CM[, DGE_cells]

# reorder the cell types desired
DGE_DATA$clusters_subset_split <- factor(DGE_DATA$clusters_subset_split, 
                                              levels = c("19-2", "19-1", "12", "9", "14"))

#Find differential expressed genes based on groups defined by "clusters_subset_split".
DGE_DATA <- SetIdent(DGE_DATA, value = "clusters_subset_split")

# The logfc threshold is set to 0.1 as compared to the default 0.25, which makes the detection of weaker signals possible (but it takes more computational power).
detable <- FindAllMarkers(DGE_DATA,
                          only.pos = TRUE,
                          max.cells.per.ident = 300,
                          logfc.threshold = .1,
                          assay = "RNA",
                          min.pct = 0.05)

# Only keep DEG that have a p-value below 0.05.
detable <- detable[detable$p_val < 0.05, ]

# Adding two extra columns giving the difference (and log difference) between pct1 and pct2.
# pct1 is the percentage of cells in the specific cluster that express this gene.
# pct2 is the percentage of cells in all other clusters that express this gene.
detable$pct.diff <- detable$pct.1 - detable$pct.2
detable$log.pct.diff <- log2((detable$pct.1*99+1) / (detable$pct.2*99+1))

# Save the detable as a csv file.
write.csv2(detable, paste0(myfolder, "cardiomyocytes/detable_19-2_19-1_12_9_14.csv"))

rm(detable, DGE_DATA, DGE_cells, sample_size)
```


### DEG CM ADRB1, CHRM2 (innvervation figure)

```{r}
values_to_keep <- c("19-2", "19-1", "12", "9", "14", "4-2", "4-1", "18", "6")

data_CM <- CM[, CM$clusters_subset_split %in% values_to_keep]

# specific list of marker to project the dotplot on
my_markers <- c("ADRB1", "CHRM2")

data_CM$clusters_subset_split <- factor(data_CM$clusters_subset_split, 
                                              levels = c("19-2", "19-1", "12", "9", "14", "4-2", "4-1", "18", "6"))

data_CM <- SetIdent(data_CM, value = "clusters_subset_split")

p <- DotPlot(data_CM, features = my_markers) +
  coord_flip() +
  theme(axis.text.y = element_text(angle = 0, hjust = 1, vjust = 0.5)) +
  scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 11, name = "RdBu") %>% rev()) +
  scale_x_discrete(limits = rev) +
  scale_y_discrete() +
  labs(x = "", y = "") + 
  theme(axis.text.x = element_text(angle = 30, hjust = 1, size = 8))

pdf(paste0(myfolder, "cardiomyocytes/dotplot_markers_CM_subset_selection_19-2_19-1_12_9_14_4-2_4-1_18_6.pdf"), width = 7, height = 3.5)
print(p)
dev.off()

rm(p)
gc()
```


### DEG CM selection of markers vCM_1, AVB-BB_CM, AVN_CM comparison (split & selection)

```{r}
values_to_keep <- c("8-2", "18", "19-1")

data_CM <- CM[, CM$clusters_subset_split %in% values_to_keep]

# specific list of marker to project the dotplot on
my_markers <- c("TNFRSF19", "XPO4", "TENM3",   "SPON1", "VCAM1", "IRX1", "IRX2", "RCAN1", "MYOM2", "ATP1A1", "BMP2", "HAMP", "FZD7", "TBX3", "ZFP36L1", "CNN1")

data_CM$clusters_subset_split <- factor(data_CM$clusters_subset_split, 
                                              levels = c("8-2", "18", "19-1"))

data_CM <- SetIdent(data_CM, value = "clusters_subset_split")

p <- DotPlot(data_CM, features = my_markers) +
  coord_flip() +
  theme(axis.text.y = element_text(angle = 0, hjust = 1, vjust = 0.5)) +
  scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 11, name = "RdBu") %>% rev()) +
  scale_x_discrete(limits = rev) +
  scale_y_discrete() +
  labs(x = "", y = "") + 
  theme(axis.text.x = element_text(angle = 30, hjust = 1, size = 8))

pdf(paste0(myfolder, "cardiomyocytes/dotplot_markers_CM_subset_selection_8-2_18_19-1.pdf"), width = 4, height = 4)
print(p)
dev.off()

rm(p)
gc()
```


## Gene Set Enrichment Analysis (GSEA) between ST-DEG and CM-subclusters

### Load ST-DEG

```{r}
# DEG calculated only between ST clusters in the heart subset 
DEG_ST_heart_subset <- read.xlsx("/home/rstudio/project/data/ST/230904_DE_heart_subset_two_versions.xlsx",
                                 sheet='230904_DE_heart_subset')

# DEG calculated between all 23 ST clusters
DEG_ST_all_clusters <- read.xlsx("/home/rstudio/project/data/ST/230904_DE_heart_subset_two_versions.xlsx",
                                 sheet='230904_DE_all_clusters')
```


### Gene sets and Expression Matrix

```{r}
# Get the gene set from single ST cluster
cluster_list <- c("2", "7", "9", "15", "17")
```


#### DEG_ST_heart_subset

```{r}
table(DEG_ST_heart_subset$cluster)

DEG_ST_heart_subset_new <- DEG_ST_heart_subset[DEG_ST_heart_subset$cluster %in% cluster_list, ]
table(DEG_ST_heart_subset_new$cluster)

# Create list of DEGs
GeneSets <- split(DEG_ST_heart_subset_new$gene, DEG_ST_heart_subset_new$cluster)

# Subset CM clusters of interest
subset_values <- c("19-2", "19-1", "12", "9", "14", "4-2", "4-1", "18", "8-2", "8-1", "8-3", "17", "6", "1", "10", "3", "7", "2", "5")
CM_subset <- CM[, CM$clusters_subset_split %in% subset_values]

CM_subset$clusters_subset_split <- factor(CM_subset$clusters_subset_split, levels = subset_values)

# Compute module score on single-cell data
CM_subset <- AddModuleScore(CM_subset, features = GeneSets, name = "Enrichment_Score")

# Rename the Enrichment_Score meta.data
# Create new_column_names using the list of numbers
new_column_names <- paste("Enrichment_Score_ST_cluster_", cluster_list, sep = "")

# Use a loop to rename the columns based on the list of indices and names
i_start <- as.numeric(dim(CM_subset@meta.data)[2]-length(GeneSets)+1)
i_end <- as.numeric(dim(CM_subset@meta.data)[2])
i_column <- as.numeric(dim(CM_subset@meta.data)[2]-length(GeneSets))

#dim(data@meta.data)[2] + length(GeneSets)
for (i in i_start:i_end) {
  colnames(CM_subset@meta.data)[i] <- new_column_names[i - i_column]
}

# Plot violin plots
pdf(file = paste0(myfolder, "cardiomyocytes/Violin_enrichment_ST.pdf"), 
     width = 20, height = 20)
VlnPlot(CM_subset, features = new_column_names, group.by = "clusters_subset_split", pt.size = 0) + 
  plot_layout(ncol = 2)
dev.off()
```


## Export cardiomyocyte annotations and object

```{r}
saveRDS(CM, file = paste0(myfolder, "rds_objects/cardiomyocytes02.rds"))
saveRDS(CM[[]], file = paste0(myfolder, "rds_objects/cardiomyocyte_metadata.rds"))
```

```{r}
CM <- readRDS(file = paste0(myfolder, "rds_objects/cardiomyocytes02.rds"))
```


================================================================================
================================================================================
================================================================================
================================================================================
================================================================================







