---
title: "09_HDCA_heart_complementary_QC_figures"
author: "RaphaÃ«l Mauron"
output: html_document
editor_options: 
  chunk_output_type: console
---

# HDCA heart Quality Controle figures

In this script, additional QC figures are plotted.

NOTE: The path to load the data and the path to save the figures should be changed at your convenience.

Environment: **Docker**


## Settings

### Set library path

```{r setup, include=FALSE}
.libPaths("/home/rstudio/project/renv/library/R-4.3/aarch64-unknown-linux-gnu/")
knitr::opts_chunk$set(echo = TRUE)
```


### Set environment variables

```{r}
# Set the directory 
myfolder <- "/home/rstudio/project/output/HDCA_heart_sc_analysis_docker/"
```


### Load Libraries

```{r, message = FALSE}
library(Seurat)
library(Matrix)
library(future)
library(RcppHNSW)
library(igraph)
library(future.apply)
library(niceRplots)
library(DoubletFinder)
library(ggplot2)
library(ROCR)
library(KernSmooth)
library(dplyr)
library(patchwork)
library(harmony)
library(gprofiler2)
library(plotly)
library(openxlsx)
library(tibble)
library(gtools)
library(tidyr)
```


### Load data

Do not load all the object at the same time. Some calculations requires a lot of memory so it is recommended to load only the object required for the specific calculation. This affects mainly the objects `data` and `data_unfiltered`.

```{r}
data <- readRDS(paste0(myfolder, "rds_objects/HDCA_heart_sc_full_extended.rds"))
data_unfiltered <- readRDS(paste0(myfolder, "rds_objects/HDCA_heart_sc_unfiltered.rds"))
ST <- readRDS("/home/rstudio/project/data/ST/220114_STdata_harmony_semla.rds")
CM <- readRDS(file = paste0(myfolder, "rds_objects/cardiomyocytes02.rds"))
EN <- readRDS(file = paste0(myfolder, "rds_objects/endothelial.rds"))
FB <- readRDS(paste0(myfolder, "rds_objects/fibroblasts.rds"))
IN <- readRDS(paste0(myfolder, "rds_objects/innervation.rds"))
```


### Set seed

```{r}
set.seed(1)
```


### Number of cells per HL cluster (Extended Tab. 10)

```{r}
number_cells <- as.matrix(table(data$cell.types))

colnames(number_cells) <- "number of cells"

# Save the detable as a csv file.
write.csv2(number_cells, paste0(myfolder, "supplementary_tables/extended_table_10_number_cells_HL.csv"))
```


## Complementary figures with data filtered and processed

### Add factor to orig.ident to plot single-cell by age

```{r}
data$orig.ident <- factor(data$orig.ident, levels = c("10X303",
                                                      "10X151",
                                                      "10X289",
                                                      "10X299",
                                                      "10X300",
                                                      "10X158",
                                                      "10X142",
                                                      "10X166",
                                                      "10X127",
                                                      "10X285",
                                                      "10X274",
                                                      "10X263",
                                                      "10X246",
                                                      "10X253",
                                                      "10X256"))
```


### XIST expression (Extended Fig. 2A)

```{r}
pdf(file = paste0(myfolder, "complementary/Violin_XIST.pdf"),  width = 15, height = 4)
VlnPlot(object = data, features = 'XIST', group.by = 'orig.ident', pt.size = 0) + theme(legend.position = 'none',
                                                                                        axis.text.x = element_text(size = 20),
                                                                                        axis.text.y = element_text(size = 20))
dev.off()
```


## UMAP all clusters embedded (Extended Fig. 2E)

#### nCount_RNA

```{r}
pdf(paste0(myfolder, "complementary/umap_nCount_RNA.pdf"), width = 20, height = 15)
FeaturePlot(data, reduction = "umap_oi", features = 'nCount_RNA', order = TRUE)
dev.off()
gc()
```


#### nFeature_RNA

```{r}
pdf(paste0(myfolder, "complementary/umap_nFeature_RNA.pdf"), width = 20, height = 15)
FeaturePlot(data, reduction = "umap_oi", features = 'nFeature_RNA', order = TRUE)
dev.off()
gc()
```


#### Level mito

```{r}
pdf(paste0(myfolder, "complementary/umap_percent_mito_after_filter.pdf"), width = 20, height = 15)
FeaturePlot(data, reduction = "umap_oi", features = "percent_mito", order = TRUE)
dev.off()
gc()
```


#### Level ribo

```{r}
pdf(paste0(myfolder, "complementary/umap_percent_ribo_after_filter.pdf"), width = 20, height = 15)
FeaturePlot(data, reduction = "umap_oi", features = "percent_ribo", order = TRUE)
dev.off()
gc()
```


#### HL clusters (Extended Fig. 2C)

```{r}
pdf(paste0(myfolder, "complementary/umap_high_level_clusters.pdf"), width = 20, height = 15)
DimPlot(data, reduction = "umap_oi", group.by = "high_level_clusters", label = T)
dev.off()
gc()
```


#### Sample ID (Extended Fig. 2D)

```{r}
pdf(paste0(myfolder, "complementary/umap_sampleID.pdf"), width = 20, height = 15)
DimPlot(data, reduction = "umap_oi", group.by = "orig.ident", label = F)
dev.off()
gc()
```


## Differentially expressed genes (DEG) after annotation - top5 - subsection clusters (Supp Fig. 2A)

```{r}
# Select a new samples size for each cluster, to get groups that are equally (or almost equally) large.
sample_size <- table(data$high_level_clusters_removed)  #high_level_clusters_removed
sample_size[sample_size > 150] <- 150
sample_size

# Order the clusters
data$high_level_clusters_removed <- factor(data$high_level_clusters_removed, levels = mixedsort(unique(data@meta.data[["high_level_clusters_removed"]])))


# Randomly sample the cells that will be included when calculating DGE (same number of cells as selected above).
# Samples without replacement, so each cell only can be included once.
DGE_cells <- lapply(names(sample_size), function(x){ 
  set.seed(1)
  sample(colnames(data)[data$high_level_clusters_removed == x], size = sample_size[x]) #high_level_clusters
  })
DGE_cells <- unlist(DGE_cells)

# Select all data that relates to the sampled DGE cells.
DGE_DATA <- data[, DGE_cells]

# reorder the cell types desired
DGE_DATA$cell.types <- factor(DGE_DATA$cell.types, 
                                              levels = c("Mat_vCM",
                                                         "Mat_aCM",
                                                         "MetAct_vCM_1",
                                                         "MetAct_vCM_2",
                                                         "MetAct_aCM",
                                                         "Prol_CM",
                                                         "Immat_CM",
                                                         "TMSB10high_C_1",
                                                         "TMSB10high_C_2",
                                                         "MacroVasc_EC",
                                                         "MicroVasc_EC",
                                                         "PDE4Chigh_EC",
                                                         "LEC",
                                                         "Endoc_EC",
                                                         "EndocCush_EC",
                                                         "Valve_MC",
                                                         "PDE4Chigh_FB",
                                                         "Int_FB",
                                                         "Peric_MC",
                                                         "OFT_FB",
                                                         "EPDC",
                                                         "AnnFibr_FB",
                                                         "Prol_FB",
                                                         "OFT_SMC",
                                                         "CA_SMC",
                                                         "PC",
                                                         "EpC",
                                                         "NB-N",
                                                         "SCP-GC",
                                                         "MyC",
                                                         "LyC"))

#Find differentially expressed genes based on groups defined by "high_level_clusters".
DGE_DATA <- SetIdent(DGE_DATA, value = "cell.types") #high_level_clusters

# The logfc threshold is set to 0.1 as compared to the default 0.25, which makes the detection of weaker signals possible (but it takes more computational power).
detable <- FindAllMarkers(DGE_DATA,
                          only.pos = TRUE,
                          max.cells.per.ident = 300,
                          logfc.threshold = .1,
                          assay = "RNA",
                          min.pct = 0.05)

# Only keep DEGs that have a p-value below 0.05.
detable <- detable[detable$p_val < 0.05, ]

# Adding two extra columns giving the difference (and log difference) between pct1 and pct2.
# pct1 is the percentage of cells in the specific cluster that express this gene.
# pct2 is the percentage of cells in all other clusters that express this gene.
detable$pct.diff <- detable$pct.1 - detable$pct.2
detable$log.pct.diff <- log2((detable$pct.1*99+1) / (detable$pct.2*99+1))

# Save the detable as a csv file.
#write.csv2(detable, paste0(myfolder, "detable_rough_clusters.csv"))

# Pick the 60 genes with the lowest p value for each group, then 40 highest based on pct.diff, then 20 highest based on log.pct.diff.
# This gives a total of 140 genes (20 per group). They seem to be unevenly spread over the cell clusters though.
# Why do we want to do it like this?
detable %>%
  group_by(cluster) %>%
  top_n(-40, p_val) %>%
  top_n(20, pct.diff) %>%
  top_n(5, avg_log2FC) ->
  top5

# Unclear what this does since there is no documentation for getcluster(). But my guess is that the genes get matched to one specific cluster somehow, and the factor is leveled based on cluster number.
#ord <- factor(sapply(unique(as.character(top5$gene)), function(x){niceRplots::getcluster(DGE_DATA, x, "high_level_clusters_removed")})) #high_level_clusters

pdf(paste0(myfolder, "complementary/DEG_top5_removed.pdf"), width = 12, height = 30)
par(mfrow=c(1,1), mar=c(3,6,2,5))
DotPlot(DGE_DATA, features = top5 %>% pull(gene) %>% unique()) +
  coord_flip() +
  theme(axis.text.y = element_text(angle = 0, hjust = 1, vjust = 0.5)) +
  scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 11, name = "RdBu") %>% rev()) +
  scale_x_discrete(limits = rev) +
  scale_y_discrete() +
  labs(x = "", y = "") +
  theme(axis.text.x = element_text(angle = 30, hjust = 1, size = 8))
dev.off()

gc()
```


## Complementary figures with data unfiltered and before processed

```{r}
data_unfiltered <- readRDS(paste0(myfolder, "rds_objects/HDCA_heart_sc_unfiltered.rds"))
rm(data)
gc()
```


### Feature barplots (Extended Fig. 2B)

```{r}
# Create dataframe from meta.data for feature plotting
metadataframe <- data_unfiltered@meta.data

# Add the age group to the metadataframe
df <- data.frame(age = c(5.5, 6, 7, 8, 8.5, 9, 10, 11.5, 12, 13.25, 14), age_group = c("5-6", "5-6", "7-8", "7-8", "7-8", "9-11", "9-11", "9-11", "12-14", "12-14", "12-14"))
age_to_age_group <- setNames(df$age_group, nm = df$age)
metadataframe$age_groups <- factor(age_to_age_group[as.character(metadataframe$age)], levels = c("5-6", "7-8", "9-11", "12-14"))
```


#### nCounts (nCount_RNA < 4000)

```{r}
metadataframe <- metadataframe %>% 
  arrange(orig.ident, nCount_RNA) %>% # order data.frame by age group and nFeature_RNA
  mutate(ord = 1:n()) %>% # Add a new column with values from 1 to last row number
  mutate(filter = case_when(nCount_RNA < 4000 ~ 0.7, TRUE ~ 1)) # Add a new column based on ifelse statement

pdf(paste0(myfolder, "complementary/bar_nCounts.pdf"), width = 20, height = 4.2)
ggplot(metadataframe, aes(ord, nCount_RNA, fill = orig.ident)) +
  geom_bar(stat = "identity", width = 1, alpha = metadataframe$filter) +
  scale_y_log10() +
  geom_hline(yintercept = 4000, linetype = "dashed") +
  labs(x = "cells", y = "log10(nCount)") + 
  theme_classic() + 
  theme(axis.text.x = element_text(size = 20),
        axis.text.y = element_text(size = 20))
dev.off()
```


#### nFeatures (nFeature_RNA < 250)

```{r}
metadataframe <- metadataframe %>% 
  arrange(orig.ident, nFeature_RNA) %>% # order data.frame by age group and nFeature_RNA
  mutate(ord = 1:n()) %>% # Add a new column with values from 1 to last row number
  mutate(filter = case_when(nFeature_RNA < 250 ~ 0.7, TRUE ~ 1)) # Add a new column based on ifelse statement

pdf(paste0(myfolder, "complementary/bar_nFeature.pdf"), width = 20, height = 4.2)
ggplot(metadataframe, aes(ord, nFeature_RNA, fill = orig.ident)) +
  geom_bar(stat = "identity", width = 1, alpha = metadataframe$filter) +
  scale_y_log10() +
  geom_hline(yintercept = 250, linetype = "dashed") + 
  labs(x = "cells", y = "log10(nFeature)") + 
  theme_classic() + 
  theme(axis.text.x = element_text(size = 20),
        axis.text.y = element_text(size = 20))
dev.off()
```


#### percent.mito (percent_mito > 30)

```{r}
metadataframe <- metadataframe %>% 
  arrange(orig.ident, percent_mito) %>% 
  mutate(ord = 1:n()) %>% 
  mutate(filter = case_when(percent_mito > 30 ~ 0.7, TRUE ~ 1))

pdf(paste0(myfolder, "complementary/bar_percent_mito.pdf"), width = 20, height = 4.2)
ggplot(metadataframe, aes(ord, percent_mito, fill = orig.ident)) +
  geom_bar(stat = "identity", width = 1, alpha = metadataframe$filter) +
  geom_hline(yintercept = 30, linetype = "dashed") + 
  labs(x = "cells", y = "% mito") + 
  theme_classic() + 
  theme(axis.text.x = element_text(size = 20),
        axis.text.y = element_text(size = 20))
dev.off()
```


#### percent_ribo (percent_ribo < 3)

```{r}
metadataframe <- metadataframe %>% 
  arrange(orig.ident, percent_ribo) %>% 
  mutate(ord = 1:n()) %>% 
  mutate(filter = case_when(percent_ribo < 3 ~ 0.7, TRUE ~ 1))

pdf(paste0(myfolder, "complementary/bar_percent_ribo.pdf"), width = 20, height = 4.2)
ggplot(metadataframe, aes(ord, percent_ribo, fill = orig.ident)) +
  geom_bar(stat = "identity", width = 1, alpha = metadataframe$filter) +
  geom_hline(yintercept = 3, linetype = "dashed") + 
  labs(x = "cells", y = "% ribo") + 
  theme_classic() + 
  theme(axis.text.x = element_text(size = 20),
        axis.text.y = element_text(size = 20))
dev.off()
```


#### percent_hemo (percent_hb < 10)

```{r}
metadataframe <- metadataframe %>% 
  arrange(orig.ident, percent_hb) %>% 
  mutate(ord = 1:n()) %>% 
  mutate(filter = case_when(percent_hb > 10 ~ 0.7, TRUE ~ 1))

pdf(paste0(myfolder, "complementary/bar_percent_hb.pdf"), width = 20, height = 4.2)
ggplot(metadataframe, aes(ord, percent_hb, fill = orig.ident)) +
  geom_bar(stat = "identity", width = 1, alpha = metadataframe$filter) +
  geom_hline(yintercept = 10, linetype = "dashed") + 
  labs(x = "cells", y = "% hb") + 
  theme_classic() + 
  theme(axis.text.x = element_text(size = 20),
        axis.text.y = element_text(size = 20))
dev.off()
```


## New age groups

Since the subgroups for age groups 9-11 and 12-14 above gave rise to groups with very few cells. We decided to merge those into a single "latest" age group 9-14.

### Add additional age groups to the metadata (age_group_refined)

```{r}
df <- data.frame(age = c(5.5, 6, 7, 8, 8.5, 9, 10, 11.5, 12, 13.25, 14), age_group = c("5-6", "5-6", "7-8", "7-8", "7-8", "9-14", "9-14", "9-14", "9-14", "9-14", "9-14"))
age_to_age_group <- setNames(df$age_group, nm = df$age)
data$age_group_refined <- factor(age_to_age_group[as.character(data$age)], levels = c("5-6", "7-8", "9-14"))
rm(df, age_to_age_group)
gc()
```


## CM Dotplot - ion channels

### Selection

#### Feature percentage all clusters from object

```{r}
#parameters
nrow <- dim(CM)[2]
my_markers <- c("HCN1", "HCN4", "SCN5A", "CACNA1C", "CACNA1D", "CACNA1G", "CACNA2D1", "CACNA2D2", "CACNB2", "RYR2", "KCNA5", "KCNH7", "KCNJ2", "KCNJ3", "KCNJ4", "KCNMB2", "KCNN2", "KCNQ1", "KCNQ3", "KCNQ5", "KCNT2", "CLCN3", "CLCN6", "CLIC1", "CLIC4", "CLIC5", "ANO4", "ANO5", "ANO6", "GRIA1", "GRID1", "GRID2", "GRIK1", "GRIK2", "GRIN2B", "P2RX1", "GABRB2", "GABRB3", "GABRE", "LRRC8B", "TRPC1", "TRPC3", "TRPM3", "TRPM4", "TRPM7", "PKD2", "GJA1", "GJA3", "GJA5", "GJC1", "AQP1")

#count function
count_greater_than_zero <- function(col) {
  as.integer(col > 0.00)
}

#create matrix with percentage of expression of each mentionned markers per cells
new_df <- data.frame(matrix(nrow = nrow, ncol = 0))
for (marker in my_markers) {
print(marker)
new_df <- cbind(new_df, PercentageFeatureSet(object = CM, features = marker))
}
colnames(new_df) <- my_markers

# Initialize an empty matrix to store the counts
marker_cluster_counts <- matrix(0, nrow = length(unique(CM$clusters_subset_split)), ncol = length(my_markers))
rownames(marker_cluster_counts) <- unique(CM$clusters_subset_split)
colnames(marker_cluster_counts) <- my_markers

# Loop through clusters
for (cluster in unique(CM$clusters_subset_split)) {
  # New data frame with cells only in the specific cluster
  new_df_spe <- new_df[rownames(new_df) %in% colnames(CM[, CM$clusters_subset_split == cluster]),]
  
  # Apply the count function to each column
  result_df <- apply(new_df_spe, 2, count_greater_than_zero)
  
  # Sum the counts for each marker
  sum_of_ones <- colSums(result_df)
  
  # Assign the counts to the corresponding row in the matrix
  marker_cluster_counts[as.character(cluster), ] <- sum_of_ones
}

#convert to dataframe
marker_cluster_counts <- as.data.frame(marker_cluster_counts)

#save as table
write.csv2(marker_cluster_counts, file = paste0(myfolder, "complementary/cells_marker_cluster_CM.csv"))
write.csv2(marker_cluster_counts, file = paste0(myfolder, "complementary/cells_marker_cluster_GPRC1_CM.csv"))
write.csv2(marker_cluster_counts, file = paste0(myfolder, "complementary/cells_marker_cluster_GPRC2_CM.csv"))


## Total number of cells per cluster
table1 <- as.data.frame(table(CM$clusters_subset_split))
colnames(table1) <- c("cluster", "number of cells")
write.csv2(table1, file = paste0(myfolder, "complementary/total_cells_CM.csv"))
```


#### Feature percentage Selection of clusters

(19-2, 19-1, 12, 9, 14, 18, 4-2, 4-1, 6)

```{r}
#parameters
nrow <- dim(CM)[2]
my_markers <- c("HCN1", "HCN4", "SCN5A", "CACNA1C", "CACNA1D", "CACNA1G", "CACNA2D1", "CACNA2D2", "CACNB2", "RYR2", "KCNA5", "KCNH7", "KCNJ2", "KCNJ3", "KCNJ4", "KCNMB2", "KCNN2", "KCNQ1", "KCNQ3", "KCNQ5", "KCNT2", "CLCN3", "CLCN6", "CLIC1", "CLIC4", "CLIC5", "ANO4", "ANO5", "ANO6", "GRIA1", "GRID1", "GRID2", "GRIK1", "GRIK2", "GRIN2B", "P2RX1", "GABRB2", "GABRB3", "GABRE", "LRRC8B", "TRPC1", "TRPC3", "TRPM3", "TRPM4", "TRPM7", "PKD2", "GJA1", "GJA3", "GJA5", "GJC1", "AQP1")

#count function
count_greater_than_zero <- function(col) {
  as.integer(col > 0.00)
}

#create matrix with percentage of expression of each mentionned markers per cells
new_df <- data.frame(matrix(nrow = nrow, ncol = 0))
for (marker in my_markers) {
print(marker)
new_df <- cbind(new_df, PercentageFeatureSet(object = CM, features = marker))
}
colnames(new_df) <- my_markers

# Initialize an empty matrix to store the counts
marker_cluster_counts <- matrix(0, nrow = length(unique(CM$clusters_subset_split)), ncol = length(my_markers))
rownames(marker_cluster_counts) <- unique(CM$clusters_subset_split)
colnames(marker_cluster_counts) <- my_markers

# Loop through clusters
for (cluster in unique(CM$clusters_subset_split)) {
  # New data frame with cells only in the specific cluster
  new_df_spe <- new_df[rownames(new_df) %in% colnames(CM[, CM$clusters_subset_split == cluster]),]
  
  # Apply the count function to each column
  result_df <- apply(new_df_spe, 2, count_greater_than_zero)
  
  # Sum the counts for each marker
  sum_of_ones <- colSums(result_df)
  
  # Assign the counts to the corresponding row in the matrix
  marker_cluster_counts[as.character(cluster), ] <- sum_of_ones
}

#convert to dataframe
marker_cluster_counts <- as.data.frame(marker_cluster_counts)

#save as table
write.csv2(marker_cluster_counts, file = paste0(myfolder, "complementary/cells_marker_cluster_CM_selection.csv"))

## Total number of cells per cluster
table1 <- as.data.frame(table(CM$clusters_subset_split))
colnames(table1) <- c("cluster", "number of cells")
write.csv2(table1, file = paste0(myfolder, "complementary/total_cells_CM_selection.csv"))
```


### VCS-Ion channel - 1 (Supp Fig. 3F)

```{r}
my_markers <- c("HCN4", "SCN5A", "CACNA1C", "CACNA1D", "CACNA1G", "CACNA2D1", "CACNB2", 
                "KCNJ2", "KCNH7", "KCNMB2", "KCNQ1", "KCNQ3",
                "CLCN3", "CLCN6", "CLIC1", "CLIC5", "ANO4", "ANO5",
                "GRID1", "P2RX1", "GABRB2", "GABRE",
                "TRPC1", "TRPC3", "TRPM3", "PKD2", 
                "GJA1", "GJA3", "GJA5", "GJC1")

CM_subset <- CM[, CM$clusters_subset_split %in% c( "18",
                                                  "4-2",
                                                  "4-1",
                                                  "6")]

CM_subset$cell.types <- factor(CM_subset$cell.types, levels = c("AVB-BB_CM",
                                                                "PF_CM",
                                                                "TsPF_CM",
                                                                "vCM_4"))

CM_subset <- SetIdent(CM_subset, value = "cell.types")

pdf(paste0(myfolder, "complementary/DGE_CM_VCS-Ion_channel-1.pdf"), width = 12, height = 3)
DotPlot(CM_subset, features = my_markers) +
  theme(axis.text.y = element_text(angle = 0, hjust = 1, vjust = 0.5)) +
  scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 11, name = "RdBu") %>% rev()) +
  scale_x_discrete() +
  scale_y_discrete(limits = rev) +
  labs(x = "", y = "") +
  theme(axis.text.x = element_text(angle = 30, hjust = 1, size = 8))
dev.off()

gc()
```


#### prob genes

```{r}
my_markers <- c("HCN4", "SCN5A", "CACNA1C", "CACNA1D", "CACNA1G", "CACNA2D1", "CACNB2", 
                "KCNJ2", "KCNH7", "KCNMB2", "KCNQ1", "KCNQ3",
                "CLCN3", "CLCN6", "CLIC1", "CLIC5", "ANO4", "ANO5",
                "GRID1", "P2RX1", "GABRB2", "GABRE",
                "TRPC1", "TRPC3", "TRPM3", "PKD2", 
                "GJA1", "GJA3", "GJA5", "GJC1")


CM <- SetIdent(CM, value = "cell.types")

pdf(paste0(myfolder, "complementary/DGE_CM_VCS-Ion_channel-1_all_cluster.pdf"), width = 12, height = 3)
DotPlot(CM, features = my_markers) +
  theme(axis.text.y = element_text(angle = 0, hjust = 1, vjust = 0.5)) +
  scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 11, name = "RdBu") %>% rev()) +
  scale_x_discrete() +
  scale_y_discrete(limits = rev) +
  labs(x = "", y = "") +
  theme(axis.text.x = element_text(angle = 30, hjust = 1, size = 8))
dev.off()

gc()
```


#### DE table

```{r}
values_to_keep <- c("18", "4-2", "4-1", "6", "1", "10")

data_CM <- CM[, CM$clusters_subset_split %in% values_to_keep]

# Select a new samples size for each cluster, to get groups that are equally (or almost equally) large.
sample_size <- table(data_CM$clusters_subset_split)
sample_size[sample_size > 150] <- 150
sample_size

# Randomly sample the cells that will be included when calculating DGE (same number of cells as selected above).
# Samples without replacement, so each cell only can be included once.
DGE_cells <- lapply(names(sample_size), function(x){ 
  set.seed(1)
  sample(colnames(data_CM)[data_CM$clusters_subset_split == x], size = sample_size[x])
  })
DGE_cells <- unlist(DGE_cells)

# Select all data that relates to the sampled DGE cells.
DGE_DATA <- data_CM[, DGE_cells]

# reorder the cell types desired
DGE_DATA$clusters_subset_split <- factor(DGE_DATA$clusters_subset_split, levels = values_to_keep)

# reorder the cell types desired
DGE_DATA$cell.types <- factor(DGE_DATA$clusters_subset_split, levels = c("18", "4-2", "4-1", "6", "1", "10"))

#Find differential expressed genes based on groups defined by "clusters_subset_split".
DGE_DATA <- SetIdent(DGE_DATA, value = "clusters_subset_split")

# The logfc threshold is set to 0.1 as compared to the default 0.25, which makes the detection of weaker signals possible (but it takes more computational power).
detable <- FindAllMarkers(DGE_DATA,
                          only.pos = TRUE,
                          max.cells.per.ident = 300,
                          logfc.threshold = .1,
                          assay = "RNA",
                          min.pct = 0.05)

# Only keep DEG that have a p-value below 0.05.
detable <- detable[detable$p_val < 0.05, ]

# Adding two extra columns giving the difference (and log difference) between pct1 and pct2.
# pct1 is the percentage of cells in the specific cluster that express this gene.
# pct2 is the percentage of cells in all other clusters that express this gene.
detable$pct.diff <- detable$pct.1 - detable$pct.2
detable$log.pct.diff <- log2((detable$pct.1*99+1) / (detable$pct.2*99+1))


# Save the detable as a csv file.
write.csv2(detable, paste0(myfolder, "cardiomyocytes/detable_18_4-2_4-1_6_1_10_new.csv"))
```


#### DE table 19-2, 19-1

```{r}
values_to_keep <- c("19-2", "19-1")

data_CM <- CM[, CM$clusters_subset_split %in% values_to_keep]

# Select a new samples size for each cluster, to get groups that are equally (or almost equally) large.
sample_size <- table(data_CM$clusters_subset_split)
sample_size[sample_size > 150] <- 150
sample_size

# Randomly sample the cells that will be included when calculating DGE (same number of cells as selected above).
# Samples without replacement, so each cell only can be included once.
DGE_cells <- lapply(names(sample_size), function(x){ 
  set.seed(1)
  sample(colnames(data_CM)[data_CM$clusters_subset_split == x], size = sample_size[x])
  })
DGE_cells <- unlist(DGE_cells)

# Select all data that relates to the sampled DGE cells.
DGE_DATA <- data_CM[, DGE_cells]

# reorder the cell types desired
DGE_DATA$clusters_subset_split <- factor(DGE_DATA$clusters_subset_split, levels = values_to_keep)

# reorder the cell types desired
DGE_DATA$cell.types <- factor(DGE_DATA$clusters_subset_split, levels = c("19-2", "19-1"))

#Find differential expressed genes based on groups defined by "clusters_subset_split".
DGE_DATA <- SetIdent(DGE_DATA, value = "clusters_subset_split")

# The logfc threshold is set to 0.1 as compared to the default 0.25, which makes the detection of weaker signals possible (but it takes more computational power).
detable <- FindAllMarkers(DGE_DATA,
                          only.pos = TRUE,
                          max.cells.per.ident = 300,
                          logfc.threshold = .1,
                          assay = "RNA",
                          min.pct = 0.05)

# Only keep DEG that have a p-value below 0.05.
detable <- detable[detable$p_val < 0.05, ]

# Adding two extra columns giving the difference (and log difference) between pct1 and pct2.
# pct1 is the percentage of cells in the specific cluster that express this gene.
# pct2 is the percentage of cells in all other clusters that express this gene.
detable$pct.diff <- detable$pct.1 - detable$pct.2
detable$log.pct.diff <- log2((detable$pct.1*99+1) / (detable$pct.2*99+1))


# Save the detable as a csv file.
write.csv2(detable, paste0(myfolder, "cardiomyocytes/detable_19-2_19-1_new.csv"))
```


#### DE table 19-2, 19-1

```{r}
values_to_keep <- c("19-2", "19-1", "18", "4-2")

data_CM <- CM[, CM$clusters_subset_split %in% values_to_keep]

# Select a new samples size for each cluster, to get groups that are equally (or almost equally) large.
sample_size <- table(data_CM$clusters_subset_split)
sample_size[sample_size > 150] <- 150
sample_size

# Randomly sample the cells that will be included when calculating DGE (same number of cells as selected above).
# Samples without replacement, so each cell only can be included once.
DGE_cells <- lapply(names(sample_size), function(x){ 
  set.seed(1)
  sample(colnames(data_CM)[data_CM$clusters_subset_split == x], size = sample_size[x])
  })
DGE_cells <- unlist(DGE_cells)

# Select all data that relates to the sampled DGE cells.
DGE_DATA <- data_CM[, DGE_cells]

# reorder the cell types desired
DGE_DATA$clusters_subset_split <- factor(DGE_DATA$clusters_subset_split, levels = values_to_keep)

# reorder the cell types desired
DGE_DATA$cell.types <- factor(DGE_DATA$clusters_subset_split, levels = c("19-2", "19-1", "18", "4-2"))

#Find differential expressed genes based on groups defined by "clusters_subset_split".
DGE_DATA <- SetIdent(DGE_DATA, value = "clusters_subset_split")

# The logfc threshold is set to 0.1 as compared to the default 0.25, which makes the detection of weaker signals possible (but it takes more computational power).
detable <- FindAllMarkers(DGE_DATA,
                          only.pos = TRUE,
                          max.cells.per.ident = 300,
                          logfc.threshold = .1,
                          assay = "RNA",
                          min.pct = 0.05)

# Only keep DEG that have a p-value below 0.05.
detable <- detable[detable$p_val < 0.05, ]

# Adding two extra columns giving the difference (and log difference) between pct1 and pct2.
# pct1 is the percentage of cells in the specific cluster that express this gene.
# pct2 is the percentage of cells in all other clusters that express this gene.
detable$pct.diff <- detable$pct.1 - detable$pct.2
detable$log.pct.diff <- log2((detable$pct.1*99+1) / (detable$pct.2*99+1))


# Save the detable as a csv file.
write.csv2(detable, paste0(myfolder, "cardiomyocytes/detable_19-2_19-1_18_4-2_new.csv"))
```


#### DE table 19-2, 19-1, 12, 9, 14 - downsampled

```{r}
values_to_keep <- c("19-2", "19-1", "12", "9", "14")

data_CM <- CM[, CM$clusters_subset_split %in% values_to_keep]

# Select a new samples size for each cluster, to get groups that are equally (or almost equally) large.
sample_size <- table(data_CM$clusters_subset_split)
sample_size[sample_size > 150] <- 150
sample_size

# Randomly sample the cells that will be included when calculating DGE (same number of cells as selected above).
# Samples without replacement, so each cell only can be included once.
DGE_cells <- lapply(names(sample_size), function(x){ 
  set.seed(1)
  sample(colnames(data_CM)[data_CM$clusters_subset_split == x], size = sample_size[x])
  })
DGE_cells <- unlist(DGE_cells)

# Select all data that relates to the sampled DGE cells.
DGE_DATA <- data_CM[, DGE_cells]

# reorder the cell types desired
DGE_DATA$clusters_subset_split <- factor(DGE_DATA$clusters_subset_split, levels = values_to_keep)

# reorder the cell types desired
DGE_DATA$cell.types <- factor(DGE_DATA$clusters_subset_split, levels = c("19-2", "19-1", "12", "9", "14"))

#Find differential expressed genes based on groups defined by "clusters_subset_split".
DGE_DATA <- SetIdent(DGE_DATA, value = "clusters_subset_split")

# The logfc threshold is set to 0.1 as compared to the default 0.25, which makes the detection of weaker signals possible (but it takes more computational power).
detable <- FindAllMarkers(DGE_DATA,
                          only.pos = TRUE,
                          max.cells.per.ident = 300,
                          logfc.threshold = .1,
                          assay = "RNA",
                          min.pct = 0.05)

# Only keep DEG that have a p-value below 0.05.
detable <- detable[detable$p_val < 0.05, ]

# Adding two extra columns giving the difference (and log difference) between pct1 and pct2.
# pct1 is the percentage of cells in the specific cluster that express this gene.
# pct2 is the percentage of cells in all other clusters that express this gene.
detable$pct.diff <- detable$pct.1 - detable$pct.2
detable$log.pct.diff <- log2((detable$pct.1*99+1) / (detable$pct.2*99+1))


# Save the detable as a csv file.
write.csv2(detable, paste0(myfolder, "cardiomyocytes/detable_19-2_19-1_12_9_14_downsampled.csv"))
```

#### DE table 19-2, 19-1, 12, 9, 14 - all cells

```{r}
values_to_keep <- c("19-2", "19-1", "12", "9", "14")

data_CM <- CM[, CM$clusters_subset_split %in% values_to_keep]

# # Select a new samples size for each cluster, to get groups that are equally (or almost equally) large.
# sample_size <- table(data_CM$clusters_subset_split)
# sample_size[sample_size > 150] <- 150
# sample_size
# 
# # Randomly sample the cells that will be included when calculating DGE (same number of cells as selected above).
# # Samples without replacement, so each cell only can be included once.
# DGE_cells <- lapply(names(sample_size), function(x){ 
#   set.seed(1)
#   sample(colnames(data_CM)[data_CM$clusters_subset_split == x], size = sample_size[x])
#   })
# DGE_cells <- unlist(DGE_cells)

# Select all data that relates to the sampled DGE cells.
DGE_DATA <- data_CM

# reorder the cell types desired
DGE_DATA$clusters_subset_split <- factor(DGE_DATA$clusters_subset_split, levels = values_to_keep)

# reorder the cell types desired
DGE_DATA$cell.types <- factor(DGE_DATA$clusters_subset_split, levels = c("19-2", "19-1", "12", "9", "14"))

#Find differential expressed genes based on groups defined by "clusters_subset_split".
DGE_DATA <- SetIdent(DGE_DATA, value = "clusters_subset_split")

# The logfc threshold is set to 0.1 as compared to the default 0.25, which makes the detection of weaker signals possible (but it takes more computational power).
detable <- FindAllMarkers(DGE_DATA,
                          only.pos = TRUE,
                          max.cells.per.ident = 300,
                          logfc.threshold = .1,
                          assay = "RNA",
                          min.pct = 0.05)

# Only keep DEG that have a p-value below 0.05.
detable <- detable[detable$p_val < 0.05, ]

# Adding two extra columns giving the difference (and log difference) between pct1 and pct2.
# pct1 is the percentage of cells in the specific cluster that express this gene.
# pct2 is the percentage of cells in all other clusters that express this gene.
detable$pct.diff <- detable$pct.1 - detable$pct.2
detable$log.pct.diff <- log2((detable$pct.1*99+1) / (detable$pct.2*99+1))


# Save the detable as a csv file.
write.csv2(detable, paste0(myfolder, "cardiomyocytes/detable_19-2_19-1_12_9_14_allcells.csv"))
```


#### DE table 18, 4-2, 4-1, 6 - downsampled

```{r}
values_to_keep <- c("18", "4-2", "4-1", "6")

data_CM <- CM[, CM$clusters_subset_split %in% values_to_keep]

# Select a new samples size for each cluster, to get groups that are equally (or almost equally) large.
sample_size <- table(data_CM$clusters_subset_split)
sample_size[sample_size > 150] <- 150
sample_size

# Randomly sample the cells that will be included when calculating DGE (same number of cells as selected above).
# Samples without replacement, so each cell only can be included once.
DGE_cells <- lapply(names(sample_size), function(x){ 
  set.seed(1)
  sample(colnames(data_CM)[data_CM$clusters_subset_split == x], size = sample_size[x])
  })
DGE_cells <- unlist(DGE_cells)

# Select all data that relates to the sampled DGE cells.
DGE_DATA <- data_CM[, DGE_cells]

# reorder the cell types desired
DGE_DATA$clusters_subset_split <- factor(DGE_DATA$clusters_subset_split, levels = values_to_keep)

# reorder the cell types desired
DGE_DATA$cell.types <- factor(DGE_DATA$clusters_subset_split, levels = c("18", "4-2", "4-1", "6"))

#Find differential expressed genes based on groups defined by "clusters_subset_split".
DGE_DATA <- SetIdent(DGE_DATA, value = "clusters_subset_split")

# The logfc threshold is set to 0.1 as compared to the default 0.25, which makes the detection of weaker signals possible (but it takes more computational power).
detable <- FindAllMarkers(DGE_DATA,
                          only.pos = TRUE,
                          max.cells.per.ident = 300,
                          logfc.threshold = .1,
                          assay = "RNA",
                          min.pct = 0.05)

# Only keep DEG that have a p-value below 0.05.
detable <- detable[detable$p_val < 0.05, ]

# Adding two extra columns giving the difference (and log difference) between pct1 and pct2.
# pct1 is the percentage of cells in the specific cluster that express this gene.
# pct2 is the percentage of cells in all other clusters that express this gene.
detable$pct.diff <- detable$pct.1 - detable$pct.2
detable$log.pct.diff <- log2((detable$pct.1*99+1) / (detable$pct.2*99+1))


# Save the detable as a csv file.
write.csv2(detable, paste0(myfolder, "cardiomyocytes/detable_18_4-2_4-1_6_downsampled.csv"))
```


#### DE table 18, 4-2, 4-1, 6 - all cells

```{r}
values_to_keep <- c("18", "4-2", "4-1", "6")

data_CM <- CM[, CM$clusters_subset_split %in% values_to_keep]

# Select a new samples size for each cluster, to get groups that are equally (or almost equally) large.
# sample_size <- table(data_CM$clusters_subset_split)
# sample_size[sample_size > 150] <- 150
# sample_size
# 
# # Randomly sample the cells that will be included when calculating DGE (same number of cells as selected above).
# # Samples without replacement, so each cell only can be included once.
# DGE_cells <- lapply(names(sample_size), function(x){ 
#   set.seed(1)
#   sample(colnames(data_CM)[data_CM$clusters_subset_split == x], size = sample_size[x])
#   })
# DGE_cells <- unlist(DGE_cells)

# Select all data that relates to the sampled DGE cells.
DGE_DATA <- data_CM

# reorder the cell types desired
DGE_DATA$clusters_subset_split <- factor(DGE_DATA$clusters_subset_split, levels = values_to_keep)

# reorder the cell types desired
DGE_DATA$cell.types <- factor(DGE_DATA$clusters_subset_split, levels = c("18", "4-2", "4-1", "6"))

#Find differential expressed genes based on groups defined by "clusters_subset_split".
DGE_DATA <- SetIdent(DGE_DATA, value = "clusters_subset_split")

# The logfc threshold is set to 0.1 as compared to the default 0.25, which makes the detection of weaker signals possible (but it takes more computational power).
detable <- FindAllMarkers(DGE_DATA,
                          only.pos = TRUE,
                          max.cells.per.ident = 300,
                          logfc.threshold = .1,
                          assay = "RNA",
                          min.pct = 0.05)

# Only keep DEG that have a p-value below 0.05.
detable <- detable[detable$p_val < 0.05, ]

# Adding two extra columns giving the difference (and log difference) between pct1 and pct2.
# pct1 is the percentage of cells in the specific cluster that express this gene.
# pct2 is the percentage of cells in all other clusters that express this gene.
detable$pct.diff <- detable$pct.1 - detable$pct.2
detable$log.pct.diff <- log2((detable$pct.1*99+1) / (detable$pct.2*99+1))


# Save the detable as a csv file.
write.csv2(detable, paste0(myfolder, "cardiomyocytes/detable_18_4-2_4-1_6_allcells.csv"))
```


#### plot dot plot 6 clusters

```{r}
my_markers <- c("HCN4", "SCN5A", "CACNA1C", "CACNA1D", "CACNA1G", "CACNA2D1", "CACNB2", 
                "KCNJ2", "KCNH7", "KCNMB2", "KCNQ1", "KCNQ3",
                "CLCN3", "CLCN6", "CLIC1", "CLIC5", "ANO4", "ANO5",
                "GRID1", "P2RX1", "GABRB2", "GABRE",
                "TRPC1", "TRPC3", "TRPM3", "PKD2", 
                "GJA1", "GJA3", "GJA5", "GJC1")

CM_subset <- CM[, CM$clusters_subset_split %in% c("18", "4-2", "4-1", "6", "1", "10")]

CM_subset$clusters_subset_split <- factor(CM_subset$clusters_subset_split, levels = c("18", "4-2", "4-1", "6", "1", "10"))

CM_subset <- SetIdent(CM_subset, value = "clusters_subset_split")

pdf(paste0(myfolder, "complementary/DGE_CM_VCS-18_4-2_4-1_6_1_10.pdf"), width = 12, height = 3)
DotPlot(CM_subset, features = my_markers) +
  theme(axis.text.y = element_text(angle = 0, hjust = 1, vjust = 0.5)) +
  scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 11, name = "RdBu") %>% rev()) +
  scale_x_discrete() +
  scale_y_discrete(limits = rev) +
  labs(x = "", y = "") +
  theme(axis.text.x = element_text(angle = 30, hjust = 1, size = 8))
dev.off()

gc()
```



```{r}
values_to_keep <- c("18", "4-2", "4-1", "6")

data_CM <- CM[, CM$clusters_subset_split %in% values_to_keep]

# Select a new samples size for each cluster, to get groups that are equally (or almost equally) large.
sample_size <- table(data_CM$clusters_subset_split)
# sample_size[sample_size > 150] <- 150
# sample_size

# Randomly sample the cells that will be included when calculating DGE (same number of cells as selected above).
# Samples without replacement, so each cell only can be included once.
# DGE_cells <- lapply(names(sample_size), function(x){ 
#   set.seed(1)
#   sample(colnames(data_CM)[data_CM$clusters_subset_split == x], size = sample_size[x])
#   })
# DGE_cells <- unlist(DGE_cells)

# Select all data that relates to the sampled DGE cells.
#DGE_DATA <- data_CM[, DGE_cells]
DGE_DATA <- data_CM

# reorder the cell types desired
DGE_DATA$clusters_subset_split <- factor(DGE_DATA$clusters_subset_split, levels = values_to_keep)

# reorder the cell types desired
DGE_DATA$cell.types <- factor(DGE_DATA$clusters_subset_split, levels = c("18", "4-2", "4-1", "6"))

#Find differential expressed genes based on groups defined by "clusters_subset_split".
DGE_DATA <- SetIdent(DGE_DATA, value = "clusters_subset_split")

# The logfc threshold is set to 0.1 as compared to the default 0.25, which makes the detection of weaker signals possible (but it takes more computational power).
detable <- FindAllMarkers(DGE_DATA,
                          only.pos = TRUE,
                          max.cells.per.ident = 300,
                          logfc.threshold = .1,
                          assay = "RNA",
                          min.pct = 0.05)

# Only keep DEG that have a p-value below 0.05.
detable <- detable[detable$p_val < 0.05, ]

# Adding two extra columns giving the difference (and log difference) between pct1 and pct2.
# pct1 is the percentage of cells in the specific cluster that express this gene.
# pct2 is the percentage of cells in all other clusters that express this gene.
detable$pct.diff <- detable$pct.1 - detable$pct.2
detable$log.pct.diff <- log2((detable$pct.1*99+1) / (detable$pct.2*99+1))


# Save the detable as a csv file.
write.csv2(detable, paste0(myfolder, "cardiomyocytes/detable_18_4-2_4-1_6_new_all_cells.csv"))
```


### VCS-Ion channel - 2

```{r}
my_markers <- c("RYR2", "KCNN2", "KCNQ5", "CLIC4", "ANO6", "TRPM4", "TRPM7")

CM_subset <- CM[, CM$clusters_subset_split %in% c( "18",
                                                  "4-2",
                                                  "4-1",
                                                  "6")]

CM_subset$clusters_subset_split <- factor(CM_subset$clusters_subset_split, levels = c( "18",
                                                  "4-2",
                                                  "4-1",
                                                  "6"))

CM_subset <- SetIdent(CM_subset, value = "clusters_subset_split")

pdf(paste0(myfolder, "complementary/DGE_CM_VCS-Ion_channel-2.pdf"), width = 12, height = 3)
DotPlot(CM_subset, features = my_markers) +
  theme(axis.text.y = element_text(angle = 0, hjust = 1, vjust = 0.5)) +
  scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 11, name = "RdBu") %>% rev()) +
  scale_x_discrete() +
  scale_y_discrete(limits = rev) +
  labs(x = "", y = "") +
  theme(axis.text.x = element_text(angle = 30, hjust = 1, size = 8))
dev.off()

gc()
```


### ACS-Ion channel - 1 (Fig. 3F)

```{r}
my_markers <- c("HCN1", "HCN4", "CACNA1C", "CACNA1D", "CACNA1G", "CACNA2D2", "CACNB2", "KCNJ3", "KCNN2", "KCNQ1", "KCNQ3", "KCNQ5", "KCNT2", "CLCN3", "CLIC4", "ANO6", "GRID2", "GRIK1", "LRRC8B", "TRPC1", "TRPC3",  "TRPM4", "TRPM7", "PKD2", "GJA3", "GJC1")

CM_subset <- CM[, CM$clusters_subset_split %in% c("19-2",
                                                  "19-1",
                                                  "12",
                                                  "9",
                                                  "14")]

CM_subset$cell.types <- factor(CM_subset$cell.types, levels = c("SAN_CM", "AVN_CM", "Cond_aCM", "Left_aCM", "Right_aCM"))

CM_subset <- SetIdent(CM_subset, value = "cell.types")

pdf(paste0(myfolder, "complementary/DGE_CM_ACS-Ion channel-1.pdf"), width = 12, height = 3)
DotPlot(CM_subset, features = my_markers) +
  theme(axis.text.y = element_text(angle = 0, hjust = 1, vjust = 0.5)) +
  scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 11, name = "RdBu") %>% rev()) +
  scale_x_discrete() +
  scale_y_discrete(limits = rev) +
  labs(x = "", y = "") +
  theme(axis.text.x = element_text(angle = 30, hjust = 1, size = 8))
dev.off()

gc()
```


#### DE table

```{r}
values_to_keep <- c("19-2", "19-1",  "12", "9", "14")

data_CM <- CM[, CM$clusters_subset_split %in% values_to_keep]

# Select a new samples size for each cluster, to get groups that are equally (or almost equally) large.
sample_size <- table(data_CM$clusters_subset_split)
sample_size[sample_size > 150] <- 150
sample_size

# Randomly sample the cells that will be included when calculating DGE (same number of cells as selected above).
# Samples without replacement, so each cell only can be included once.
DGE_cells <- lapply(names(sample_size), function(x){ 
  set.seed(1)
  sample(colnames(data_CM)[data_CM$clusters_subset_split == x], size = sample_size[x])
  })
DGE_cells <- unlist(DGE_cells)

# Select all data that relates to the sampled DGE cells.
DGE_DATA <- data_CM[, DGE_cells]

# reorder the cell types desired
DGE_DATA$clusters_subset_split <- factor(DGE_DATA$clusters_subset_split, levels = values_to_keep)

# reorder the cell types desired
DGE_DATA$cell.types <- factor(DGE_DATA$clusters_subset_split, levels = c("19-2", "19-1",  "12", "9", "14"))

#Find differential expressed genes based on groups defined by "clusters_subset_split".
DGE_DATA <- SetIdent(DGE_DATA, value = "clusters_subset_split")

# The logfc threshold is set to 0.1 as compared to the default 0.25, which makes the detection of weaker signals possible (but it takes more computational power).
detable <- FindAllMarkers(DGE_DATA,
                          only.pos = TRUE,
                          max.cells.per.ident = 300,
                          logfc.threshold = .1,
                          assay = "RNA",
                          min.pct = 0.05)

# Only keep DEG that have a p-value below 0.05.
detable <- detable[detable$p_val < 0.05, ]

# Adding two extra columns giving the difference (and log difference) between pct1 and pct2.
# pct1 is the percentage of cells in the specific cluster that express this gene.
# pct2 is the percentage of cells in all other clusters that express this gene.
detable$pct.diff <- detable$pct.1 - detable$pct.2
detable$log.pct.diff <- log2((detable$pct.1*99+1) / (detable$pct.2*99+1))


# Save the detable as a csv file.
write.csv2(detable, paste0(myfolder, "cardiomyocytes/detable_19-2_19-1_12_9_14.csv"))
```



### ACS-Ion channel - 2

```{r}
my_markers <- c("SCN5A", "CACNA2D1", "RYR2", "KCNA5", "KCNH7",  "KCNMB2", "CLCN6", "CLIC1", "CLIC5", "ANO5", "GRIN2B", "GABRB2", "GABRB3", "TRPM3", "GJA1", "GJA5")

CM_subset <- CM[, CM$clusters_subset_split %in% c("19-2",
                                                  "19-1",
                                                  "12",
                                                  "9",
                                                  "14")]

CM_subset$clusters_subset_split <- factor(CM_subset$clusters_subset_split, levels = c("19-2","19-1","12","9","14"))

CM_subset <- SetIdent(CM_subset, value = "clusters_subset_split")

pdf(paste0(myfolder, "complementary/DGE_CM_ACS-Ion channel-2.pdf"), width = 12, height = 3)
DotPlot(CM_subset, features = my_markers) +
  theme(axis.text.y = element_text(angle = 0, hjust = 1, vjust = 0.5)) +
  scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 11, name = "RdBu") %>% rev()) +
  scale_x_discrete() +
  scale_y_discrete(limits = rev) +
  labs(x = "", y = "") +
  theme(axis.text.x = element_text(angle = 30, hjust = 1, size = 8))
dev.off()

gc()
```


### FeaturePlot (Fig. 3B)

```{r}
## remove: 11, 13, 15, 16, 20, 21, 22
CM_subset <- CM[, CM$clusters_subset_split %in% c("19-2", "19-1", "12", "9", "14", "4-2", "4-1", "18", "8-2", "8-1", "8-3", "17", "6", "1", "10", "3", "7", "2", "5")]

CM_subset$clusters_subset_split <- factor(CM_subset$clusters_subset_split, levels = c("19-2", "19-1", "12", "9", "14", "4-2", "4-1", "18", "8-2", "8-1", "8-3", "17", "6", "1", "10", "3", "7", "2", "5"))
CM_subset <- SetIdent(CM_subset, value = "clusters_subset_split")
```

```{r}
DimPlot(CM_subset, reduction = "umap_subset", raster = FALSE)
```


#### MYH6

```{r}
pdf(paste0(myfolder, "cardiomyocytes/FeaturePlot_MYH6_not_ordered.pdf"), width = 10, height = 10)
FeaturePlot(CM_subset, reduction = "umap_subset", features = "MYH6", raster = FALSE, order = F)
dev.off()
```


#### MYH7

```{r}
pdf(paste0(myfolder, "cardiomyocytes/FeaturePlot_MYH7_not_ordered.pdf"), width = 10, height = 10)
FeaturePlot(CM_subset, reduction = "umap_subset", features = "MYH7", raster = FALSE, order = F)
dev.off()
```


#### CKM

```{r}
pdf(paste0(myfolder, "cardiomyocytes/FeaturePlot_CKM_not_ordered.pdf"), width = 10, height = 10)
FeaturePlot(CM_subset, reduction = "umap_subset", features = "CKM", raster = FALSE, order = F)
dev.off()
```


#### MKI67

```{r}
pdf(paste0(myfolder, "cardiomyocytes/FeaturePlot_MKI67_not_ordered.pdf"), width = 10, height = 10)
FeaturePlot(CM_subset, reduction = "umap_subset", features = "MKI67", raster = FALSE, order = F)
dev.off()
```


## Violin plot

### CM (Extended Fig. 3B)

```{r}
# Create a single violin plot
pdf(file = paste0(myfolder, "complementary/Vln_feature_CM_before_split.pdf"),  width = 15, height = 4)
VlnPlot(CM, group.by = "clusters_subset", features = "nFeature_RNA", pt.size = 0) +
stat_summary(fun.y = median, geom='point', size = 25, colour = "black", shape = 95) +
theme(legend.position = 'none')
dev.off()
```

#### HBA1 across all CM clusters (without exclusions, unordered) (Extended Fig. 3C)

```{r}
pdf(file = paste0(myfolder, "complementary/Vln_feature_CM_HBA1.pdf"),  width = 15, height = 4)
VlnPlot(CM, group.by = "clusters_subset", features = "HBA1", pt.size = 0) +
theme(legend.position = 'none')
dev.off()
```


### EN (Extended Fig. 3B)

```{r}
# Create a single violin plot
pdf(file = paste0(myfolder, "complementary/Vln_feature_EN.pdf"),  width = 15, height = 4)
VlnPlot(EN, group.by = "clusters_subset", features = "nFeature_RNA", pt.size = 0) +
stat_summary(fun.y = median, geom='point', size = 25, colour = "black", shape = 95) +
theme(legend.position = 'none')
dev.off()
```


#### HBA1 across all EN clusters (without exclusions, unordered) (Extended Fig. 3C)

```{r}
pdf(file = paste0(myfolder, "complementary/Vln_feature_EN_HBA1.pdf"),  width = 15, height = 4)
VlnPlot(EN, group.by = "clusters_subset", features = "HBA1", pt.size = 0) +
theme(legend.position = 'none')
dev.off()
```


### FB (Extended Fig. 3B)

```{r}
# Create a single violin plot
pdf(file = paste0(myfolder, "complementary/Vln_feature_FB.pdf"),  width = 15, height = 4)
VlnPlot(FB, group.by = "clusters_subset", features = "nFeature_RNA", pt.size = 0) +
stat_summary(fun.y = median, geom='point', size = 25, colour = "black", shape = 95) +
theme(legend.position = 'none')
dev.off()
```


### HBA1 across all FB clusters (again without exclusions, unordered) (Extended Fig. 3C)

```{r}
pdf(file = paste0(myfolder, "complementary/Vln_feature_FB_HBA1.pdf"),  width = 15, height = 4)
VlnPlot(FB, group.by = "clusters_subset", features = "HBA1", pt.size = 0) +
theme(legend.position = 'none')
dev.off()
```


### Data

#### Data - all clusters (clusters_louvain_oi) (Extended Fig. 2F)

```{r}
# Create a single violin plot
pdf(file = paste0(myfolder, "complementary/Vln_feature_data_all_clusters.pdf"),  width = 15, height = 4)
VlnPlot(data, group.by = "clusters_louvain_oi", features = "nFeature_RNA", pt.size = 0) +
stat_summary(fun.y = median, geom='point', size = 25, colour = "black", shape = 95) +
theme(legend.position = 'none')
dev.off()
```


#### HBA1 across all HL clusters (without exclusions, unordered) (Extended Fig. 2F)

```{r}
pdf(file = paste0(myfolder, "complementary/Vln_feature_HL_HBA1.pdf"),  width = 15, height = 4)
VlnPlot(data, group.by = "clusters_louvain_oi", features = "HBA1", pt.size = 0) +
theme(legend.position = 'none')
dev.off()
```


## DE analysis multi clusters

```{r}
data_curated <- data[, data$high_level_clusters %in% c("1", "2", "3", "4_35", "5", "6", "7", "8", "9", "10", "11", "12", "13", "14", "15", "16", "17", "18_1", "18_2", "20", "21", "22", "23", "24", "25", "26", "28", "29", "30", "31", "32", "33", "34")]

data_curated <- FindVariableFeatures(data_curated, nfeatures = 5000)
var.features <- data_curated@assays$RNA@var.features

sampled_cells <- data_curated[[]] %>% 
  tibble::rownames_to_column(var = "barcode") %>% 
  group_by(high_level_clusters)
rm(data_curated)

annotation <- sampled_cells %>% 
  select(barcode, high_level_clusters, age_groups) %>% 
  data.frame(row.names = 1)
colnames(annotation) <- c("bio_celltype", "age_groups")
```

```{r}
# Load annotations prepared earlier and filter
# CM
cardiomyocyte_clusters_keep <- c("1", "2", "3", "4-1", "4-2", "5", "6", "7", "8-1", "8-2", "8-3", "9", "10", "12", "14", "17", "18", "19-1", "19-2")
cardiomyocyte_metadata <- readRDS(paste0(myfolder, "rds_objects/cardiomyocyte_metadata.rds")) %>% 
  mutate(clusters_subset = paste0("CM_", as.character(clusters_subset))) %>%  # rename cardiomyocyte clusters by adding a prefix CM_
  mutate_if(is.factor, as.character) %>% 
  filter(clusters_subset_split %in% cardiomyocyte_clusters_keep)

# EN
endothelial_clusters_keep <- c("1", "2", "3", "5", "6", "7", "10", "11", "12", "13", "14", "15", "16", "17", "18", "19", "20", "21")
endothelial_metadata <- readRDS(paste0(myfolder, "rds_objects/endothelial_metadata.rds")) %>% 
  filter(clusters_subset %in% endothelial_clusters_keep) %>% 
  mutate(clusters_subset = paste0("EN_", as.character(clusters_subset))) # rename endothelial clusters by adding a prefix EN_

# FB
fibroblasts_clusters_keep <- c("1", "2", "3", "4", "5", "6", "7", "8", "9", "11", "12", "13", "14", "15", "17", "18", "19", "20")
fibroblasts_metadata <- readRDS(paste0(myfolder, "rds_objects/fibroblasts_metadata.rds")) %>% 
  filter(clusters_subset %in% fibroblasts_clusters_keep) %>% 
  mutate(clusters_subset = paste0("FB_", as.character(clusters_subset))) # rename fibroblasts clusters by adding a prefix FB_

# IN
innervation_clusters_keep <- c("1", "3", "4", "7", "8", "11", "12", "13", "17", "19")
innervation_metadata <- readRDS(paste0(myfolder, "rds_objects/innervation_metadata.rds")) %>% 
  filter(clusters_subset %in% innervation_clusters_keep) %>% 
  mutate(clusters_subset = paste0("IN_", as.character(clusters_subset))) # rename innervation clusters by adding a prefix IN_

# Merge deep level clusters from the three different subsets
deep_clusters <- cardiomyocyte_metadata %>% 
  select(clusters_subset, clusters_subset_split) %>% # Only select clustering results
  bind_rows(endothelial_metadata %>% select(clusters_subset)) %>% # Add endothelial clusters
  bind_rows(fibroblasts_metadata %>% select(clusters_subset)) %>% # Add fibroblast clusters
  bind_rows(innervation_metadata %>% select(clusters_subset)) %>% # Add innervation clusters
  mutate(clusters_subset_split = case_when(is.na(clusters_subset_split) ~ clusters_subset, # rename cardiomyocyte clusters by adding a prefix CM_
                                           TRUE ~ paste0("CM_", clusters_subset_split)))

# Define clusters to keep
keep_clusters <- c(paste0("HL_", c("6", "8", "10", "15", "18_1", "20", "22", "29", "34")), 
                   paste0("CM_", cardiomyocyte_clusters_keep),
                   paste0("EN_", endothelial_clusters_keep),
                   paste0("FB_", fibroblasts_clusters_keep),
                   paste0("IN_", innervation_clusters_keep))


# Create a new annotations tibble
annotation_split <- annotation 
annotation_split <- annotation_split %>% 
  rownames_to_column(var = "ID") %>% 
  left_join(y = deep_clusters %>% rownames_to_column(var = "ID"), by = "ID") %>% 
  mutate(clusters_subset_split = case_when(is.na(clusters_subset_split) ~ paste0("HL_", bio_celltype),
                                           TRUE ~ clusters_subset_split)) %>% 
  filter(clusters_subset_split %in% keep_clusters) # keep selected clusters

# Convert clusters_subset_split to a factor and set order of clusters
annotation_split <- annotation_split %>% 
  mutate(clusters_subset_split = factor(clusters_subset_split, keep_clusters))

# Check overlap (sanity check)
all(annotation$ID %in% colnames(data))

rm(annotation, cardiomyocyte_metadata, deep_clusters, endothelial_metadata, fibroblasts_metadata, sampled_cells, cardiomyocyte_clusters_keep, endothelial_clusters_keep, fibroblasts_clusters_keep, keep_clusters)
gc()
```

```{r}
data <- data[, colnames(data) %in% annotation_split$ID]
data$clusters_subset_split <- as.character(annotation_split$clusters_subset_split)
```


### Number of cells per DL cluster (Extended Tab. 11)

#### Add the annotation

```{r}
annotation <- read.csv2(paste0("/home/rstudio/project/", "data/metadata/HDCA_heart_SC_annotations_DL_240204.csv"), header = 1, sep = ";")
cells_clusters <- setNames(annotation$cell_type, nm = annotation$cluster)
```

```{r}
number_cells <- as.matrix(table(data$clusters_subset_split))
number_cells_annot <- number_cells
rownames(number_cells_annot) <- cells_clusters[as.character(rownames(number_cells_annot))]
colnames(number_cells_annot) <- "number of cells"

# Save the detable as a csv file.
write.csv2(number_cells_annot, paste0(myfolder, "supplementary_tables/extended_table_11_number_cells_DL.csv"))
```


#### HL_22 FB_15

```{r}
clusters_to_analyze <- c("HL_22", "FB_15")

my_niche <- annotation_split[annotation_split$clusters_subset_split %in% clusters_to_analyze,]
data_sub <- data[, colnames(data) %in% my_niche$ID]

table(data_sub$clusters_subset_split)
```

```{r}
data_sub <- SetIdent(data_sub, value = "clusters_subset_split")

# The logfc threshold is set to 0.1 as compared to the default 0.25, which makes the detection of weaker signals possible (but it takes more computational power).
detable <- FindAllMarkers(data_sub,
                          only.pos = TRUE,
                          max.cells.per.ident = 300,
                          logfc.threshold = .1,
                          assay = "RNA",
                          min.pct = 0.05)

# Save the detable as a csv file.
write.csv2(detable, paste0(myfolder, "complementary/DEG_HL_22_FB_15.csv"))

# Only keep DEG that have a p-value below 0.05.
detable <- detable[detable$p_val < 0.05, ]

# Adding two extra columns giving the difference (and log difference) between pct1 and pct2.
# pct1 is the percentage of cells in the specific cluster that express this gene.
# pct2 is the percentage of cells in all other clusters that express this gene.
detable$pct.diff <- detable$pct.1 - detable$pct.2
detable$log.pct.diff <- log2((detable$pct.1*99+1) / (detable$pct.2*99+1))

# Pick the 40 genes with the lowest p value for each cluster, then 20 highest based on pct.diff, then 5 highest based on avg_log2FC.
detable %>%
  group_by(cluster) %>%
  top_n(-40, p_val) %>%
  top_n(20, pct.diff) %>%
  top_n(10, avg_log2FC) ->
  top10

pdf(paste0(myfolder, "complementary/DEG_HL_22_FB_15.pdf"), width = 4.5, height = 8)
par(mfrow=c(1,1), mar=c(3,6,2,5))
DotPlot(data_sub, features = top10 %>% pull(gene) %>% unique()) +
  coord_flip() +
  theme(axis.text.y = element_text(angle = 0, hjust = 1, vjust = 0.5)) +
  scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 11, name = "RdBu") %>% rev()) +
  scale_x_discrete(limits = rev) +
  scale_y_discrete() +
  labs(x = "", y = "") +
  theme(axis.text.x = element_text(angle = 30, hjust = 0.5))
dev.off()
```


#### HL_20, FB_7 and FB_3

```{r}
clusters_to_analyze <- c("HL_20", "FB_7", "FB_3")

my_niche <- annotation_split[annotation_split$clusters_subset_split %in% clusters_to_analyze,]
data_sub <- data[, colnames(data) %in% my_niche$ID]

table(data_sub$clusters_subset_split)
```

```{r}
data_sub <- SetIdent(data_sub, value = "clusters_subset_split")

# The logfc threshold is set to 0.1 as compared to the default 0.25, which makes the detection of weaker signals possible (but it takes more computational power).
detable <- FindAllMarkers(data_sub,
                          only.pos = TRUE,
                          max.cells.per.ident = 300,
                          logfc.threshold = .1,
                          assay = "RNA",
                          min.pct = 0.05)

# Save the detable as a csv file.
write.csv2(detable, paste0(myfolder, "complementary/DEG_HL_20_FB_7_FB_3.csv"))

# Only keep DEG that have a p-value below 0.05.
detable <- detable[detable$p_val < 0.05, ]

# Adding two extra columns giving the difference (and log difference) between pct1 and pct2.
# pct1 is the percentage of cells in the specific cluster that express this gene.
# pct2 is the percentage of cells in all other clusters that express this gene.
detable$pct.diff <- detable$pct.1 - detable$pct.2
detable$log.pct.diff <- log2((detable$pct.1*99+1) / (detable$pct.2*99+1))

# Pick the 40 genes with the lowest p value for each cluster, then 20 highest based on pct.diff, then 5 highest based on avg_log2FC.
detable %>%
  group_by(cluster) %>%
  top_n(-40, p_val) %>%
  top_n(20, pct.diff) %>%
  top_n(10, avg_log2FC) ->
  top10

pdf(paste0(myfolder, "complementary/DEG_HL_20_FB_7_FB_3.pdf"), width = 4.5, height = 8)
par(mfrow=c(1,1), mar=c(3,6,2,5))
DotPlot(data_sub, features = top10 %>% pull(gene) %>% unique()) +
  coord_flip() +
  theme(axis.text.y = element_text(angle = 0, hjust = 1, vjust = 0.5)) +
  scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 11, name = "RdBu") %>% rev()) +
  scale_x_discrete(limits = rev) +
  scale_y_discrete() +
  labs(x = "", y = "") +
  theme(axis.text.x = element_text(angle = 30, hjust = 0.5))
dev.off()
```


### DEG EN selection of markers EN_18, EN_17, EN_7, FB_4, FB_11, FB_13 comparison (split & selection)

```{r}
clusters_to_analyze <- c("EN_18", "EN_17", "EN_7", "FB_4", "FB_11", "FB_13")

my_niche <- annotation_split[annotation_split$clusters_subset_split %in% clusters_to_analyze,]
data_sub <- data[, colnames(data) %in% my_niche$ID]

table(data_sub$clusters_subset_split)
```

```{r}
data_sub$clusters_subset_split <- factor(data_sub$clusters_subset_split, levels = clusters_to_analyze)

data_sub <- SetIdent(data_sub, value = "clusters_subset_split")

# specific list of marker to project the dotplot on
my_markers <- c("MSX1", "ESRRG", "FOXP2", "LEF1", "MAF", "FREM1", "SOX6", "FOS", "MAFF", "JUNB", "NFATC2", "RUNX1", "ATF3", "EGR1", "CEBPD", "ZNF93", "NPAS3", "IRF1", "NFE2L3", "MEIS2", "ARID5B", "HEY2", "LMO2", "ARID5A", "SMAD6", "KLF4", "SP6", "GLIS3", "RFX2", "NFKB1", "NFATC1", "FOSB", "SEMA4A", "TBX20", "SALL3", "ZNF385D", "MECOM", "PRDM16", "FOXC2", "HAND2", "PROX1", "HMGA2", "SOX5", "TWIST1", "SETBP1")

p <- DotPlot(data_sub, features = my_markers) +
  coord_flip() +
  theme(axis.text.y = element_text(angle = 0, hjust = 1, vjust = 0.5)) +
  scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 11, name = "RdBu") %>% rev()) +
  scale_x_discrete(limits = rev) +
  scale_y_discrete() +
  labs(x = "", y = "") + 
  theme(axis.text.x = element_text(angle = 30, hjust = 1, size = 8))

pdf(paste0(myfolder, "dotplot_markers_EN_18_17_7_FB_4_11_13.pdf"), width = 4.5, height = 9)
print(p)
dev.off()

rm(p)
gc()
```


#### Patho expression EC_17, EC_7, FB_4, FB_13, FB_11 (Fig. 7D)

```{r}
clusters_to_analyze <- c("EN_17", "EN_7", "FB_4", "FB_13", "FB_11")

my_niche <- annotation_split[annotation_split$clusters_subset_split %in% clusters_to_analyze,]
data_sub <- data[, colnames(data) %in% my_niche$ID]

table(data_sub$clusters_subset_split)
```


#### WNT signaling:

```{r}
my_markers <- c("WNT2", "WNT2B", "WNT4", "WNT5B", "WNT9B", "WNT11", "FZD10", "LRP6", "ROR1", "ROR2", "LGR4", "LGR5", "RSPO2", "RSPO3", "ZNRF3", "DKK2", "DKK3", "SFRP1", "SFRP4")
```

```{r}
#get matrix average expression from marker genes
mat <- AverageExpression(object = data_sub, features = my_markers, group.by = 'clusters_subset_split')$RNA

# plot heatmap
pal <- colorRampPalette(RColorBrewer::brewer.pal(n = 8, name = "Reds"))(25)

# normalise average expression gene by gene
gg_mat <- mat |>
  as.data.frame() |>
  tibble::rownames_to_column(var = "gene") |>
  pivot_longer(-gene) |>
  group_by(gene) |>
  mutate(value = scale(value, center = F, scale = T)) |>
  ungroup() |>
  mutate(gene = factor(gene, levels = rev(my_markers)))

gg_mat$name <- factor(gg_mat$name, levels = c("EN_17", "EN_7", "FB_4", "FB_13", "FB_11"))

#plot
p <- ggplot(gg_mat, aes(name, gene, fill = value)) + 
  geom_tile() +
  scale_fill_gradientn(colours = pal) +
  theme(panel.grid = element_blank(),
        axis.text.x = element_text(angle = 90)) +
  labs(fill = "normalised \n score")

pdf(paste0(myfolder, "patho_enrich/congenital_heart_disease_spe_markers_niche_gg_with_norm_wnt.pdf"), width = 3, height = 5)
p
dev.off()
```


#### BMP/inhibin-activin signaling:

```{r}
my_markers <- c("BMP2", "BMP4", "BMP6", "INHBA", "BMPR1A", "BMPR1B", "ACVR1", "BMPR2", "NBL1", "SOST", "BMPER", "FST", "FSTL4", "BAMBI")
```

```{r}
#get matrix average expression from marker genes
mat <- AverageExpression(object = data_sub, features = my_markers, group.by = 'clusters_subset_split')$RNA

# plot heatmap
pal <- colorRampPalette(RColorBrewer::brewer.pal(n = 8, name = "Reds"))(25)

# normalise average expression gene by gene
gg_mat <- mat |>
  as.data.frame() |>
  tibble::rownames_to_column(var = "gene") |>
  pivot_longer(-gene) |>
  group_by(gene) |>
  mutate(value = scale(value, center = F, scale = T)) |>
  ungroup() |>
  mutate(gene = factor(gene, levels = rev(my_markers)))

gg_mat$name <- factor(gg_mat$name, levels = c("EN_17", "EN_7", "FB_4", "FB_13", "FB_11"))

#plot
p <- ggplot(gg_mat, aes(name, gene, fill = value)) + 
  geom_tile() +
  scale_fill_gradientn(colours = pal) +
  theme(panel.grid = element_blank(),
        axis.text.x = element_text(angle = 90)) +
  labs(fill = "normalised \n score")

pdf(paste0(myfolder, "patho_enrich/congenital_heart_disease_spe_markers_niche_gg_with_norm_bmp.pdf"), width = 3, height = 3.68)
p
dev.off()
```


#### TGFbeta signaling:

```{r}
my_markers <- c("TGFB1", "TGFB2", "TGFBR1", "LTBP1", "FMOD", "BAMBI", "THBS1")
```

```{r}
#get matrix average expression from marker genes
mat <- AverageExpression(object = data_sub, features = my_markers, group.by = 'clusters_subset_split')$RNA

# plot heatmap
pal <- colorRampPalette(RColorBrewer::brewer.pal(n = 8, name = "Reds"))(25)

# normalise average expression gene by gene
gg_mat <- mat |>
  as.data.frame() |>
  tibble::rownames_to_column(var = "gene") |>
  pivot_longer(-gene) |>
  group_by(gene) |>
  mutate(value = scale(value, center = F, scale = T)) |>
  ungroup() |>
  mutate(gene = factor(gene, levels = rev(my_markers)))

gg_mat$name <- factor(gg_mat$name, levels = c("EN_17", "EN_7", "FB_4", "FB_13", "FB_11"))

#plot
p <- ggplot(gg_mat, aes(name, gene, fill = value)) + 
  geom_tile() +
  scale_fill_gradientn(colours = pal) +
  theme(panel.grid = element_blank(),
        axis.text.x = element_text(angle = 90)) +
  labs(fill = "normalised \n score")

pdf(paste0(myfolder, "patho_enrich/congenital_heart_disease_spe_markers_niche_gg_with_norm_TGFbeta.pdf"), width = 3, height = 2.2)
p
dev.off()
```


#### Patho gene expression dev valvular (Fig. 7E)

```{r}
my_markers <- c("ADAMTS19", "TEX41", "SMAD6", "MUC4", "PALMD", "GATA4", "GATA5", "GATA6", "NOTCH1", "ROBO4", "MAT2A", "ACTA2", "FLNA", "DCHS1", "DZIP1", "SMAD2", "SMAD4", "BMPR1A", "COL1A1", "ELN", "FBN1", "SH3PXD2B", "B3GAT3", "HCN4", "TGFB2", "TGFB3", "TGFBR1", "TGFBR2", "SMAD3")
```

```{r}
#get matrix average expression from marker genes
mat <- AverageExpression(object = data_sub, features = my_markers, group.by = 'clusters_subset_split')$RNA

# plot heatmap
pal <- colorRampPalette(RColorBrewer::brewer.pal(n = 8, name = "Reds"))(25)

# normalise average expression gene by gene
gg_mat <- mat |>
  as.data.frame() |>
  tibble::rownames_to_column(var = "gene") |>
  pivot_longer(-gene) |>
  group_by(gene) |>
  mutate(value = scale(value, center = F, scale = T)) |>
  ungroup() |>
  mutate(gene = factor(gene, levels = rev(my_markers)))

gg_mat$name <- factor(gg_mat$name, levels = c("EN_17", "EN_7", "FB_4", "FB_13", "FB_11"))

#plot
p <- ggplot(gg_mat, aes(name, gene, fill = value)) + 
  geom_tile() +
  scale_fill_gradientn(colours = pal) +
  theme(panel.grid = element_blank(),
        axis.text.x = element_text(angle = 90)) +
  labs(fill = "normalised \n score")

pdf(paste0(myfolder, "patho_enrich/congenital_heart_disease_spe_markers_niche_gg_with_norm_valvar.pdf"), width = 3, height = 6)
p
dev.off()
```


## High Level UMAP embedded with Deep Level cluster

## High level clusters
HL stereoscope: exclude 6, 19, 27, 34

```{r}
data_curated <- data[, data$high_level_clusters %in% c("1", "2", "3", "4_35", "5", "6", "7", "8", "9", "10", "11", "12", "13", "14", "15", "16", "17", "18_1", "18_2", "20", "21", "22", "23", "24", "25", "26", "28", "29", "30", "31", "32", "33", "34")]
```

```{r, message = FALSE, warning=FALSE}
data_curated <- FindVariableFeatures(data_curated, nfeatures = 5000)
var.features <- data_curated@assays$RNA@var.features
```

```{r}
sampled_cells <- data_curated[[]] %>% 
  tibble::rownames_to_column(var = "barcode") %>% 
  group_by(high_level_clusters)
rm(data_curated)
```

```{r, , message = FALSE}
annotation <- sampled_cells %>% 
  select(barcode, high_level_clusters,  age_groups) %>% 
  data.frame(row.names = 1)
colnames(annotation) <- c("bio_celltype", "age_groups")
```

```{r}
# Load annotations prepared earlier and filter
# CM
cardiomyocyte_clusters_keep <- c("1", "2", "3", "4-1", "4-2", "5", "6", "7", "8-1", "8-2", "8-3", "9", "10", "12", "14", "17", "18", "19-1", "19-2")
cardiomyocyte_metadata <- readRDS(paste0(myfolder, "rds_objects/cardiomyocyte_metadata.rds")) %>% 
  mutate(clusters_subset = paste0("CM_", as.character(clusters_subset))) %>%  # rename cardiomyocyte clusters by adding a prefix CM_
  mutate_if(is.factor, as.character) %>% 
  filter(clusters_subset_split %in% cardiomyocyte_clusters_keep)

# EN
endothelial_clusters_keep <- c("1", "2", "3", "5", "6", "7", "10", "11", "12", "13", "14", "15", "16", "17", "18", "19", "20", "21")
endothelial_metadata <- readRDS(paste0(myfolder, "rds_objects/endothelial_metadata.rds")) %>% 
  filter(clusters_subset %in% endothelial_clusters_keep) %>% 
  mutate(clusters_subset = paste0("EN_", as.character(clusters_subset))) # rename endothelial clusters by adding a prefix EN_

# FB
fibroblasts_clusters_keep <- c("1", "2", "3", "4", "5", "6", "7", "8", "9", "11", "12", "13", "14", "15", "17", "18", "19", "20")
fibroblasts_metadata <- readRDS(paste0(myfolder, "rds_objects/fibroblasts_metadata.rds")) %>% 
  filter(clusters_subset %in% fibroblasts_clusters_keep) %>% 
  mutate(clusters_subset = paste0("FB_", as.character(clusters_subset))) # rename fibroblasts clusters by adding a prefix FB_

# IN
innervation_clusters_keep <- c("1", "3", "4", "7", "8", "11", "12", "13", "17", "19")
innervation_metadata <- readRDS(paste0(myfolder, "rds_objects/innervation_metadata.rds")) %>% 
  filter(clusters_subset %in% innervation_clusters_keep) %>% 
  mutate(clusters_subset = paste0("IN_", as.character(clusters_subset))) # rename innervation clusters by adding a prefix IN_

# Merge deep level clusters from the three different subsets
deep_clusters <- cardiomyocyte_metadata %>% 
  select(clusters_subset, clusters_subset_split) %>% # Only select clustering results
  bind_rows(endothelial_metadata %>% select(clusters_subset)) %>% # Add endothelial clusters
  bind_rows(fibroblasts_metadata %>% select(clusters_subset)) %>% # Add fibroblast clusters
  bind_rows(innervation_metadata %>% select(clusters_subset)) %>% # Add innervation clusters
  mutate(clusters_subset_split = case_when(is.na(clusters_subset_split) ~ clusters_subset, # rename cardiomyocyte clusters by adding a prefix CM_
                                           TRUE ~ paste0("CM_", clusters_subset_split)))

# Define clusters to keep
keep_clusters <- c(paste0("HL_", c("6", "8", "10", "15", "18_1", "20", "22", "29", "34")), 
                   paste0("CM_", cardiomyocyte_clusters_keep),
                   paste0("EN_", endothelial_clusters_keep),
                   paste0("FB_", fibroblasts_clusters_keep),
                   paste0("IN_", innervation_clusters_keep))


# Create a new annotations tibble
annotation_split <- annotation 
annotation_split <- annotation_split %>% 
  rownames_to_column(var = "ID") %>% 
  left_join(y = deep_clusters %>% rownames_to_column(var = "ID"), by = "ID") %>% 
  mutate(clusters_subset_split = case_when(is.na(clusters_subset_split) ~ paste0("HL_", bio_celltype),
                                           TRUE ~ clusters_subset_split)) %>% 
  filter(clusters_subset_split %in% keep_clusters) # keep selected clusters

# Convert clusters_subset_split to a factor and set order of clusters
annotation_split <- annotation_split %>% 
  mutate(clusters_subset_split = factor(clusters_subset_split, keep_clusters))

# Check overlap (sanity check)
all(annotation$ID %in% colnames(data))

# Check which cells we keep
table(annotation_split$clusters_subset_split)
```

```{r}
data_to_plot <- data[, colnames(data) %in% annotation_split$ID]
data_to_plot$clusters_subset_split <- as.character(annotation_split$clusters_subset_split)
```

```{r}
table(data_to_plot@meta.data[["clusters_subset_split"]])
```


#### Plot HL umap fine-graine embedded (Fig. 7A)

```{r}
pdf(paste0(myfolder, "complementary/umap_DL_embedding.pdf"), width = 20, height = 15)
DimPlot(data_to_plot, reduction = "umap_oi", group.by = "clusters_subset_split", label = T)
dev.off()
gc()
```


================================================================================
================================================================================
================================================================================
================================================================================
================================================================================









