---
title: "09_HDCA_heart_complementary_QC_figures"
author: "RaphaÃ«l Mauron"
output: html_document
editor_options: 
  chunk_output_type: console
---

# HDCA heart Quality Controle figures

In this script, additional QC figures are plotted.

NOTE: The path to load the data and the path to save the figures should be changed at your convenience.

Environment: **Docker**


## Settings

### Set library path

```{r setup, include=FALSE}
.libPaths("/home/rstudio/project/renv/library/R-4.3/aarch64-unknown-linux-gnu/")
knitr::opts_chunk$set(echo = TRUE)
```


### Set environment variables

```{r}
# Set the directory 
myfolder <- "/home/rstudio/project/output/HDCA_heart_sc_analysis_docker/"
```


### Load Libraries

```{r, message = FALSE}
library(Seurat)
library(Matrix)
library(future)
library(RcppHNSW)
library(igraph)
library(future.apply)
library(niceRplots)
library(DoubletFinder)
library(ggplot2)
library(ROCR)
library(KernSmooth)
library(dplyr)
library(patchwork)
library(harmony)
library(gprofiler2)
library(plotly)
library(openxlsx)
library(tibble)
```


### Load data

Do not load all the object at the same time. Some calculations requires a lot of memory so it is recommended to load only the object required for the specific calculation. This affects mainly the objects `data` and `data_unfiltered`.

```{r}
data <- readRDS(paste0(myfolder, "rds_objects/HDCA_heart_sc_full_extended.rds"))
data_unfiltered <- readRDS(paste0(myfolder, "rds_objects/HDCA_heart_sc_unfiltered.rds"))
ST <- readRDS("/home/rstudio/project/data/ST/220114_STdata_harmony_semla.rds")
CM <- readRDS(file = paste0(myfolder, "rds_objects/cardiomyocytes02.rds"))
EN <- readRDS(file = paste0(myfolder, "rds_objects/endothelial.rds"))
FB <- readRDS(paste0(myfolder, "rds_objects/fibroblasts.rds"))
IN <- readRDS(paste0(myfolder, "rds_objects/innervation.rds"))
```


### Set seed

```{r}
set.seed(1)
```


## Complementary figures with data filtered and processed

### Add factor to orig.ident to plot single-cell by age

```{r}
data$orig.ident <- factor(data$orig.ident, levels = c("10X303",
                                                      "10X151",
                                                      "10X289",
                                                      "10X299",
                                                      "10X300",
                                                      "10X158",
                                                      "10X142",
                                                      "10X166",
                                                      "10X127",
                                                      "10X285",
                                                      "10X274",
                                                      "10X263",
                                                      "10X246",
                                                      "10X253",
                                                      "10X256"))
```


### XIST expression

```{r}
pdf(file = paste0(myfolder, "complementary/Violin_XIST.pdf"),  width = 15, height = 4)
VlnPlot(object = data, features = 'XIST', group.by = 'orig.ident', pt.size = 0) + theme(legend.position = 'none',
                                                                                        axis.text.x = element_text(size = 20),
                                                                                        axis.text.y = element_text(size = 20))
dev.off()
```


### KDM5D expression

```{r}
pdf(file = paste0(myfolder, "complementary/Violin_KDM5D.pdf"),  width = 15, height = 4)
VlnPlot(object = data, features = 'KDM5D', group.by = 'orig.ident', pt.size = 0) + theme(legend.position = 'none')
dev.off()
```


### UTY expression

```{r}
pdf(file = paste0(myfolder, "complementary/Violin_UTY.pdf"),  width = 15, height = 4)
VlnPlot(object = data, features = 'UTY', group.by = 'orig.ident', pt.size = 0) + theme(legend.position = 'none')
dev.off()
```


### ZFY expression

```{r}
pdf(file = paste0(myfolder, "complementary/Violin_ZFY.pdf"),  width = 15, height = 4)
VlnPlot(object = data, features = 'ZFY', group.by = 'orig.ident', pt.size = 0) + theme(legend.position = 'none')
dev.off()
```


### USP9Y expression

```{r}
pdf(file = paste0(myfolder, "complementary/Violin_USP9Y.pdf"),  width = 15, height = 4)
VlnPlot(object = data, features = 'USP9Y', group.by = 'orig.ident', pt.size = 0) + theme(legend.position = 'none')
dev.off()
```



## UMAP all clusters embedded

#### nCount_RNA

```{r}
pdf(paste0(myfolder, "complementary/umap_nCount_RNA.pdf"), width = 20, height = 15)
FeaturePlot(data, reduction = "umap_oi", features = 'nCount_RNA', order = TRUE)
dev.off()
gc()
```


#### nFeature_RNA

```{r}
pdf(paste0(myfolder, "complementary/umap_nFeature_RNA.pdf"), width = 20, height = 15)
FeaturePlot(data, reduction = "umap_oi", features = 'nFeature_RNA', order = TRUE)
dev.off()
gc()
```


#### Sample ID

```{r}
pdf(paste0(myfolder, "complementary/umap_sampleID.pdf"), width = 20, height = 15)
DimPlot(data, reduction = "umap_oi", group.by = "orig.ident", label = F)
dev.off()
gc()
```


#### Age

```{r}
pdf(paste0(myfolder, "complementary/umap_age_groups.pdf"), width = 20, height = 15)
DimPlot(data, reduction = "umap_oi", group.by = "age_groups", label = F)
dev.off()
gc()
```


#### HL clusters

```{r}
pdf(paste0(myfolder, "complementary/umap_high_level_clusters.pdf"), width = 20, height = 15)
DimPlot(data, reduction = "umap_oi", group.by = "high_level_clusters", label = T)
dev.off()
gc()
```


#### HL cell types

```{r}
pdf(paste0(myfolder, "complementary/umap_cell_types.pdf"), width = 20, height = 15)
DimPlot(data, reduction = "umap_oi", group.by = "cell.types", label = T)
dev.off()
gc()
```


#### Level 'HEY2'

```{r}
pdf(paste0(myfolder, "complementary/umap_HEY2.pdf"), width = 20, height = 15)
FeaturePlot(data[, as.character(data$high_level_clusters_removed) %in% levels(data$high_level_clusters_removed)], reduction = "umap_oi", features = 'HEY2', order = TRUE)
dev.off()
gc()
```


#### Level mito

```{r}
pdf(paste0(myfolder, "complementary/umap_percent_mito_after_filter.pdf"), width = 20, height = 15)
FeaturePlot(data, reduction = "umap_oi", features = "percent_mito", order = TRUE)
dev.off()
gc()
```


#### Level ribo

```{r}
pdf(paste0(myfolder, "complementary/umap_percent_ribo_after_filter.pdf"), width = 20, height = 15)
FeaturePlot(data, reduction = "umap_oi", features = "percent_ribo", order = TRUE)
dev.off()
gc()
```


#### Level G2/M

```{r}
pdf(paste0(myfolder, "complementary/umap_G2M_filter.pdf"), width = 20, height = 15)
FeaturePlot(data, reduction = "umap_oi", features = "G2M.Score", order = TRUE)
dev.off()
gc()
```


#### Level S

```{r}
pdf(paste0(myfolder, "complementary/umap_S_filter.pdf"), width = 20, height = 15)
FeaturePlot(data, reduction = "umap_oi", features = "S.Score", order = TRUE)
dev.off()
gc()
```


### CM

#### nCount_RNA

```{r}
pdf(paste0(myfolder, "complementary/umap_nCount_RNA_CM.pdf"), width = 20, height = 15)
FeaturePlot(CM, reduction = "umap_subset", features = 'nCount_RNA', order = TRUE)
dev.off()
gc()
```


#### nFeature_RNA

```{r}
pdf(paste0(myfolder, "complementary/umap_nFeature_RNA_CM.pdf"), width = 20, height = 15)
FeaturePlot(CM, reduction = "umap_subset", features = 'nFeature_RNA', order = TRUE)
dev.off()
gc()
```


### EN

#### nCount_RNA

```{r}
pdf(paste0(myfolder, "complementary/umap_nCount_RNA_EN.pdf"), width = 20, height = 15)
FeaturePlot(EN, reduction = "umap_subset", features = 'nCount_RNA', order = TRUE)
dev.off()
gc()
```


#### nFeature_RNA

```{r}
pdf(paste0(myfolder, "complementary/umap_nFeature_RNA_EN.pdf"), width = 20, height = 15)
FeaturePlot(EN, reduction = "umap_subset", features = 'nFeature_RNA', order = TRUE)
dev.off()
gc()
```


### FB

#### nCount_RNA

```{r}
pdf(paste0(myfolder, "complementary/umap_nCount_RNA_FB.pdf"), width = 20, height = 15)
FeaturePlot(FB, reduction = "umap_subset", features = 'nCount_RNA', order = TRUE)
dev.off()
gc()
```


#### nFeature_RNA

```{r}
pdf(paste0(myfolder, "complementary/umap_nFeature_RNA_FB.pdf"), width = 20, height = 15)
FeaturePlot(FB, reduction = "umap_subset", features = 'nFeature_RNA', order = TRUE)
dev.off()
gc()
```


### IN

#### nCount_RNA

```{r}
pdf(paste0(myfolder, "complementary/umap_nCount_RNA_IN.pdf"), width = 20, height = 15)
FeaturePlot(IN, reduction = "umap_subset", features = 'nCount_RNA', order = TRUE)
dev.off()
gc()
```


#### nFeature_RNA

```{r}
pdf(paste0(myfolder, "complementary/umap_nFeature_RNA_IN.pdf"), width = 20, height = 15)
FeaturePlot(IN, reduction = "umap_subset", features = 'nFeature_RNA', order = TRUE)
dev.off()
gc()
```


#### Origin

```{r}
pdf(paste0(myfolder, "innervation/innervationl_umap_origin.pdf"), width = 10, height = 10)
DimPlot(IN, reduction = "umap_subset", group.by = "high_level_clusters_removed", label = T)
dev.off()
```


#### Cell cycle

```{r}
pdf(paste0(myfolder, "innervation/innervationl_umap_phase_cell_cycle.pdf"), width = 10, height = 10)
DimPlot(IN, reduction = "umap_subset", group.by = "Phase", label = T)
dev.off()
```


#### Age distribution

```{r}
pdf(paste0(myfolder, "innervation/innervationl_umap_age_groups.pdf"), width = 10, height = 10)
DimPlot(IN, reduction = "umap_subset", group.by = "age_groups", label = T)
dev.off()
```


## Differentially expressed genes (DEG) after annotation - top5 - all clusters

```{r}
# Select a new samples size for each cluster, to get groups that are equally (or almost equally) large.
sample_size <- table(data$high_level_clusters)  #high_level_clusters
sample_size[sample_size > 150] <- 150
sample_size

# Order the clusters
data$high_level_clusters <- factor(data$high_level_clusters, levels = mixedsort(unique(data@meta.data[["high_level_clusters"]])))

# Randomly sample the cells that will be included when calculating DGE (same number of cells as selected above).
# Samples without replacement, so each cell only can be included once.
DGE_cells <- lapply(names(sample_size), function(x){ 
  set.seed(1)
  sample(colnames(data)[data$high_level_clusters == x], size = sample_size[x]) #high_level_clusters
  })
DGE_cells <- unlist(DGE_cells)

# Select all data that relates to the sampled DGE cells.
DGE_DATA <- data[, DGE_cells]

#Find differentially expressed genes based on groups defined by "high_level_clusters".
DGE_DATA <- SetIdent(DGE_DATA, value = "high_level_clusters") #high_level_clusters

# The logfc threshold is set to 0.1 as compared to the default 0.25, which makes the detection of weaker signals possible (but it takes more computational power).
detable <- FindAllMarkers(DGE_DATA,
                          only.pos = TRUE,
                          max.cells.per.ident = 300,
                          logfc.threshold = .1,
                          assay = "RNA",
                          min.pct = 0.05)

# Only keep DEGs that have a p-value below 0.05.
detable <- detable[detable$p_val < 0.05, ]

# Adding two extra columns giving the difference (and log difference) between pct1 and pct2.
# pct1 is the percentage of cells in the specific cluster that express this gene.
# pct2 is the percentage of cells in all other clusters that express this gene.
detable$pct.diff <- detable$pct.1 - detable$pct.2
detable$log.pct.diff <- log2((detable$pct.1*99+1) / (detable$pct.2*99+1))

# Save the detable as a csv file.
#write.csv2(detable, paste0(myfolder, "detable_rough_clusters.csv"))

# Pick the 60 genes with the lowest p value for each group, then 40 highest based on pct.diff, then 20 highest based on log.pct.diff.
# This gives a total of 140 genes (20 per group). They seem to be unevenly spread over the cell clusters though.
# Why do we want to do it like this?
detable %>%
  group_by(cluster) %>%
  top_n(-40, p_val) %>%
  top_n(20, pct.diff) %>%
  top_n(5, avg_log2FC) ->
  top5

# Unclear what this does since there is no documentation for getcluster(). But my guess is that the genes get matched to one specific cluster somehow, and the factor is leveled based on cluster number.
#ord <- factor(sapply(unique(as.character(top5$gene)), function(x){niceRplots::getcluster(DGE_DATA, x, "high_level_clusters")})) #high_level_clusters

pdf(paste0(myfolder, "complementary/DEG_top5.pdf"), width = 12, height = 30)
par(mfrow=c(1,1), mar=c(3,6,2,5))
DotPlot(DGE_DATA, features = top5 %>% pull(gene) %>% unique()) +
  coord_flip() +
  theme(axis.text.y = element_text(angle = 0, hjust = 1, vjust = 0.5)) +
  scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 11, name = "RdBu") %>% rev()) +
  scale_x_discrete(limits = rev) +
  scale_y_discrete() +
  labs(x = "", y = "") +
  theme(axis.text.x = element_text(angle = 0, hjust = 0.5))
dev.off()

gc()
```


## Differentially expressed genes (DEG) after annotation - top5 - subsection clusters

```{r}
# Select a new samples size for each cluster, to get groups that are equally (or almost equally) large.
sample_size <- table(data$high_level_clusters_removed)  #high_level_clusters_removed
sample_size[sample_size > 150] <- 150
sample_size

# Order the clusters
data$high_level_clusters_removed <- factor(data$high_level_clusters_removed, levels = mixedsort(unique(data@meta.data[["high_level_clusters_removed"]])))


# Randomly sample the cells that will be included when calculating DGE (same number of cells as selected above).
# Samples without replacement, so each cell only can be included once.
DGE_cells <- lapply(names(sample_size), function(x){ 
  set.seed(1)
  sample(colnames(data)[data$high_level_clusters_removed == x], size = sample_size[x]) #high_level_clusters
  })
DGE_cells <- unlist(DGE_cells)

# Select all data that relates to the sampled DGE cells.
DGE_DATA <- data[, DGE_cells]

# reorder the cell types desired
DGE_DATA$high_level_clusters_removed <- factor(DGE_DATA$cell.types, 
                                              levels = c("Ventricular CM_1",
                                                         "Atrial CM_1",
                                                         "Ventricular CM_2",
                                                         "Ventricular CM_3",
                                                         "Atrial CM_2",
                                                         "Proliferating CM",
                                                         "Immature CM",
                                                         "TMSB10high_1",
                                                         "TMSB10high_2",
                                                         "Macrovascular EC",
                                                         "Microvascular EC",
                                                         "SEMA4Ahigh EC",
                                                         "Lymphatic EC",
                                                         "Endocardial EC",
                                                         "Valve-related EC",
                                                         "Valve-related FB",
                                                         "SEMA4Ahigh FB",
                                                         "Interstitial FB",
                                                         "Vessel-related FB",
                                                         "Outflow tract FB",
                                                         "Epicardium-derived progenitor cells",
                                                         "Atrioventricular plane FB",
                                                         "Proliferating FB",
                                                         "Outflow tract SMC",
                                                         "Coronary SMC",
                                                         "Pericytes",
                                                         "Epicardial cells",
                                                         "Cardiac neural crest cells",
                                                         "Schwann cell progenitor cells",
                                                         "Myeloid cells",
                                                         "Lymphoid cells"
                                                         ))

#Find differentially expressed genes based on groups defined by "high_level_clusters".
DGE_DATA <- SetIdent(DGE_DATA, value = "high_level_clusters_removed") #high_level_clusters

# The logfc threshold is set to 0.1 as compared to the default 0.25, which makes the detection of weaker signals possible (but it takes more computational power).
detable <- FindAllMarkers(DGE_DATA,
                          only.pos = TRUE,
                          max.cells.per.ident = 300,
                          logfc.threshold = .1,
                          assay = "RNA",
                          min.pct = 0.05)

# Only keep DEGs that have a p-value below 0.05.
detable <- detable[detable$p_val < 0.05, ]

# Adding two extra columns giving the difference (and log difference) between pct1 and pct2.
# pct1 is the percentage of cells in the specific cluster that express this gene.
# pct2 is the percentage of cells in all other clusters that express this gene.
detable$pct.diff <- detable$pct.1 - detable$pct.2
detable$log.pct.diff <- log2((detable$pct.1*99+1) / (detable$pct.2*99+1))

# Save the detable as a csv file.
#write.csv2(detable, paste0(myfolder, "detable_rough_clusters.csv"))

# Pick the 60 genes with the lowest p value for each group, then 40 highest based on pct.diff, then 20 highest based on log.pct.diff.
# This gives a total of 140 genes (20 per group). They seem to be unevenly spread over the cell clusters though.
# Why do we want to do it like this?
detable %>%
  group_by(cluster) %>%
  top_n(-40, p_val) %>%
  top_n(20, pct.diff) %>%
  top_n(5, avg_log2FC) ->
  top5

# Unclear what this does since there is no documentation for getcluster(). But my guess is that the genes get matched to one specific cluster somehow, and the factor is leveled based on cluster number.
#ord <- factor(sapply(unique(as.character(top5$gene)), function(x){niceRplots::getcluster(DGE_DATA, x, "high_level_clusters_removed")})) #high_level_clusters

pdf(paste0(myfolder, "complementary/DEG_top5_removed.pdf"), width = 12, height = 30)
par(mfrow=c(1,1), mar=c(3,6,2,5))
DotPlot(DGE_DATA, features = top5 %>% pull(gene) %>% unique()) +
  coord_flip() +
  theme(axis.text.y = element_text(angle = 0, hjust = 1, vjust = 0.5)) +
  scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 11, name = "RdBu") %>% rev()) +
  scale_x_discrete(limits = rev) +
  scale_y_discrete() +
  labs(x = "", y = "") +
  theme(axis.text.x = element_text(angle = 30, hjust = 1, size = 8))
dev.off()

gc()
```


## Complementary figures with data unfiltered and before processed

```{r}
data_unfiltered <- readRDS(paste0(myfolder, "rds_objects/HDCA_heart_sc_unfiltered.rds"))
rm(data)
gc()
```

#### nCount_RNA_before_filtering

```{r}
pdf(paste0(myfolder, "complementary/umap_nCount_RNA_before_filtering.pdf"), width = 20, height = 15)
FeaturePlot(data_unfiltered, reduction = "umap_unfilt", features = 'nCount_RNA', order = TRUE, raster=FALSE)
dev.off()
gc()
```


#### nFeature_RNA_before_filtering

```{r}
pdf(paste0(myfolder, "complementary/umap_nFeature_RNA_before_filtering.pdf"), width = 20, height = 15)
FeaturePlot(data_unfiltered, reduction = "umap_unfilt", features = 'nFeature_RNA', order = TRUE, raster=FALSE)
dev.off()
gc()
```


### Add factor to orig.ident to plot single-cell by age

```{r}
data_unfiltered$orig.ident <- factor(data_unfiltered$orig.ident, levels = c("10X303",
                                                                            "10X151",
                                                                            "10X289",
                                                                            "10X299",
                                                                            "10X300",
                                                                            "10X158",
                                                                            "10X142",
                                                                            "10X166",
                                                                            "10X127",
                                                                            "10X285",
                                                                            "10X274",
                                                                            "10X263",
                                                                            "10X246",
                                                                            "10X253",
                                                                            "10X256"))
```

### Feature barplots

```{r}
# Create dataframe from meta.data for feature plotting
metadataframe <- data_unfiltered@meta.data

# Add the age group to the metadataframe
df <- data.frame(age = c(5.5, 6, 7, 8, 8.5, 9, 10, 11.5, 12, 13.25, 14), age_group = c("5-6", "5-6", "7-8", "7-8", "7-8", "9-11", "9-11", "9-11", "12-14", "12-14", "12-14"))
age_to_age_group <- setNames(df$age_group, nm = df$age)
metadataframe$age_groups <- factor(age_to_age_group[as.character(metadataframe$age)], levels = c("5-6", "7-8", "9-11", "12-14"))
```


#### nCounts (nCount_RNA < 4000)

```{r}
metadataframe <- metadataframe %>% 
  arrange(orig.ident, nCount_RNA) %>% # order data.frame by age group and nFeature_RNA
  mutate(ord = 1:n()) %>% # Add a new column with values from 1 to last row number
  mutate(filter = case_when(nCount_RNA < 4000 ~ 0.7, TRUE ~ 1)) # Add a new column based on ifelse statement

pdf(paste0(myfolder, "complementary/bar_nCounts.pdf"), width = 20, height = 4.2)
ggplot(metadataframe, aes(ord, nCount_RNA, fill = orig.ident)) +
  geom_bar(stat = "identity", width = 1, alpha = metadataframe$filter) +
  scale_y_log10() +
  geom_hline(yintercept = 4000, linetype = "dashed") +
  labs(x = "cells", y = "log10(nCount)") + 
  theme_classic() + 
  theme(axis.text.x = element_text(size = 20),
        axis.text.y = element_text(size = 20))
dev.off()
```


#### nFeatures (nFeature_RNA < 250)

```{r}
metadataframe <- metadataframe %>% 
  arrange(orig.ident, nFeature_RNA) %>% # order data.frame by age group and nFeature_RNA
  mutate(ord = 1:n()) %>% # Add a new column with values from 1 to last row number
  mutate(filter = case_when(nFeature_RNA < 250 ~ 0.7, TRUE ~ 1)) # Add a new column based on ifelse statement

pdf(paste0(myfolder, "complementary/bar_nFeature.pdf"), width = 20, height = 4.2)
ggplot(metadataframe, aes(ord, nFeature_RNA, fill = orig.ident)) +
  geom_bar(stat = "identity", width = 1, alpha = metadataframe$filter) +
  scale_y_log10() +
  geom_hline(yintercept = 250, linetype = "dashed") + 
  labs(x = "cells", y = "log10(nFeature)") + 
  theme_classic() + 
  theme(axis.text.x = element_text(size = 20),
        axis.text.y = element_text(size = 20))
dev.off()
```


#### percent.mito (percent_mito > 30)

```{r}
metadataframe <- metadataframe %>% 
  arrange(orig.ident, percent_mito) %>% 
  mutate(ord = 1:n()) %>% 
  mutate(filter = case_when(percent_mito > 30 ~ 0.7, TRUE ~ 1))

pdf(paste0(myfolder, "complementary/bar_percent_mito.pdf"), width = 20, height = 4.2)
ggplot(metadataframe, aes(ord, percent_mito, fill = orig.ident)) +
  geom_bar(stat = "identity", width = 1, alpha = metadataframe$filter) +
  geom_hline(yintercept = 30, linetype = "dashed") + 
  labs(x = "cells", y = "% mito") + 
  theme_classic() + 
  theme(axis.text.x = element_text(size = 20),
        axis.text.y = element_text(size = 20))
dev.off()
```


#### percent_ribo (percent_ribo < 3)

```{r}
metadataframe <- metadataframe %>% 
  arrange(orig.ident, percent_ribo) %>% 
  mutate(ord = 1:n()) %>% 
  mutate(filter = case_when(percent_ribo < 3 ~ 0.7, TRUE ~ 1))

pdf(paste0(myfolder, "complementary/bar_percent_ribo.pdf"), width = 20, height = 4.2)
ggplot(metadataframe, aes(ord, percent_ribo, fill = orig.ident)) +
  geom_bar(stat = "identity", width = 1, alpha = metadataframe$filter) +
  geom_hline(yintercept = 3, linetype = "dashed") + 
  labs(x = "cells", y = "% ribo") + 
  theme_classic() + 
  theme(axis.text.x = element_text(size = 20),
        axis.text.y = element_text(size = 20))
dev.off()
```


#### percent_hemo (percent_hb < 10)

```{r}
metadataframe <- metadataframe %>% 
  arrange(orig.ident, percent_hb) %>% 
  mutate(ord = 1:n()) %>% 
  mutate(filter = case_when(percent_hb > 10 ~ 0.7, TRUE ~ 1))

pdf(paste0(myfolder, "complementary/bar_percent_hb.pdf"), width = 20, height = 4.2)
ggplot(metadataframe, aes(ord, percent_hb, fill = orig.ident)) +
  geom_bar(stat = "identity", width = 1, alpha = metadataframe$filter) +
  geom_hline(yintercept = 10, linetype = "dashed") + 
  labs(x = "cells", y = "% hb") + 
  theme_classic() + 
  theme(axis.text.x = element_text(size = 20),
        axis.text.y = element_text(size = 20))
dev.off()
```


#### S.score

```{r}
# metadataframe <- metadataframe %>% 
#   arrange(orig.ident, S.Score) %>% 
#   mutate(ord = 1:n())
# 
# pdf(paste0(myfolder, "complementary/bar_s_score.pdf"), width = 20, height = 4.2)
# ggplot(metadataframe, aes(ord, S.Score, fill = orig.ident)) +
#   geom_bar(stat = "identity", width = 1, alpha = metadataframe$filter) +
#   #geom_hline(yintercept = 10, linetype = "dashed") + 
#   labs(x = "cells", y = "S.Score") + 
#   theme_classic() + 
#   theme(axis.text.x = element_text(size = 20),
#         axis.text.y = element_text(size = 20))
# dev.off()
```


#### G2M score

```{r}
# metadataframe <- metadataframe %>% 
#   arrange(orig.ident, G2M.Score) %>% 
#   mutate(ord = 1:n())
# 
# pdf(paste0(myfolder, "complementary/bar_G2M_score.pdf"), width = 20, height = 4.2)
# ggplot(metadataframe, aes(ord, G2M.Score, fill = orig.ident)) +
#   geom_bar(stat = "identity", width = 1, alpha = metadataframe$filter) +
#   #geom_hline(yintercept = 10, linetype = "dashed") + 
#   labs(x = "cells", y = "G2M.Score") + 
#   theme_classic() + 
#   theme(axis.text.x = element_text(size = 20),
#         axis.text.y = element_text(size = 20))
# dev.off()
```


## General Dotplot - down sampled (10-109 cells)

```{r}
# Select a new samples size for each cluster, to get groups that are equally (or almost equally) large.
sample_size <- table(data$high_level_clusters_removed)
sample_size[sample_size > 150] <- 150
sample_size

# Randomly sample the cells that will be included when calculating DGE (same number of cells as selected above).
# Samples without replacement, so each cell only can be included once.
DGE_cells <- lapply(names(sample_size), function(x){ 
  set.seed(1)
  sample(colnames(data)[data$high_level_clusters_removed == x], size = sample_size[x])
  })
DGE_cells <- unlist(DGE_cells)

# Select all data that relates to the sampled DGE cells.
DGE_DATA <- data[, DGE_cells]

#Find differential expressed genes based on groups defined by "clusters_subset".
DGE_DATA <- SetIdent(DGE_DATA, value = "high_level_clusters_removed")

# The logfc threshold is set to 0.1 as compared to the default 0.25, which makes the detection of weaker signals possible (but it takes more computational power).
detable <- FindAllMarkers(DGE_DATA,
                          only.pos = TRUE,
                          max.cells.per.ident = 300,
                          logfc.threshold = .1,
                          assay = "RNA",
                          min.pct = 0.05)

# Only keep DEG that have a p-value below 0.05.
detable <- detable[detable$p_val < 0.05, ]

# Adding two extra columns giving the difference (and log difference) between pct1 and pct2.
# pct1 is the percentage of cells in the specific cluster that express this gene.
# pct2 is the percentage of cells in all other clusters that express this gene.
detable$pct.diff <- detable$pct.1 - detable$pct.2
detable$log.pct.diff <- log2((detable$pct.1*99+1) / (detable$pct.2*99+1))

# Save the detable as a csv file.
write.csv2(detable, paste0(myfolder, "complementary/detable_general_downsampled.csv"))

# Pick the 60 genes with the lowest p value for each age group, then 40 highest based on pct.diff, then 20 highest based on log.pct.diff.
detable %>%
  group_by(cluster) %>%
  top_n(-40, p_val) %>%
  top_n(20, pct.diff) %>%
  top_n(5, avg_log2FC) ->
  top5

pdf(paste0(myfolder, "complementary/DGE_general_downsampled.pdf"), width = 10, height = 32)
par(mfrow=c(1,1), mar=c(3,6,2,5))
DotPlot(DGE_DATA,  features = top5 %>% pull(gene) %>% unique()) +
  coord_flip() +
  theme(axis.text.y = element_text(angle = 0, hjust = 1, vjust = 0.5)) +
  scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 11, name = "RdBu") %>% rev()) +
  scale_x_discrete(limits = rev) +
  scale_y_discrete() +
  #labs(x = "", y = "") +
  theme(axis.text.x = element_text(angle = 0, hjust = 0.5))
dev.off()

rm(detable, DGE_DATA, top20, DGE_cells, ord, sample_size)
```


## General Dotplot - all cells

```{r}
# Select a new samples size for each cluster, to get groups that are equally (or almost equally) large.
sample_size <- table(data$high_level_clusters_removed)
#sample_size[sample_size > 150] <- 150
sample_size

# Randomly sample the cells that will be included when calculating DGE (same number of cells as selected above).
# Samples without replacement, so each cell only can be included once.
DGE_cells <- lapply(names(sample_size), function(x){ 
  set.seed(1)
  colnames(data)[data$high_level_clusters_removed == x]
  })
DGE_cells <- unlist(DGE_cells)

# Select all data that relates to the sampled DGE cells.
DGE_DATA <- data[, DGE_cells]

#Find differential expressed genes based on groups defined by "clusters_subset".
DGE_DATA <- SetIdent(DGE_DATA, value = "high_level_clusters_removed")

# The logfc threshold is set to 0.1 as compared to the default 0.25, which makes the detection of weaker signals possible (but it takes more computational power).
detable <- FindAllMarkers(DGE_DATA,
                          only.pos = TRUE,
                          max.cells.per.ident = 300,
                          logfc.threshold = .1,
                          assay = "RNA",
                          min.pct = 0.05)

# Only keep DEG that have a p-value below 0.05.
detable <- detable[detable$p_val < 0.05, ]

# Adding two extra columns giving the difference (and log difference) between pct1 and pct2.
# pct1 is the percentage of cells in the specific cluster that express this gene.
# pct2 is the percentage of cells in all other clusters that express this gene.
detable$pct.diff <- detable$pct.1 - detable$pct.2
detable$log.pct.diff <- log2((detable$pct.1*99+1) / (detable$pct.2*99+1))

# Save the detable as a csv file.
write.csv2(detable, paste0(myfolder, "complementary/detable_general_complete.csv"))

# Pick the 60 genes with the lowest p value for each age group, then 40 highest based on pct.diff, then 20 highest based on log.pct.diff.
detable %>%
  group_by(cluster) %>%
  top_n(-40, p_val) %>%
  top_n(20, pct.diff) %>%
  top_n(5, avg_log2FC) ->
  top5

pdf(paste0(myfolder, "complementary/DGE_general_complete.pdf"), width = 10, height = 32)
par(mfrow=c(1,1), mar=c(3,6,2,5))
DotPlot(DGE_DATA,  features = top5 %>% pull(gene) %>% unique()) +
  coord_flip() +
  theme(axis.text.y = element_text(angle = 0, hjust = 1, vjust = 0.5)) +
  scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 11, name = "RdBu") %>% rev()) +
  scale_x_discrete(limits = rev) +
  scale_y_discrete() +
  #labs(x = "", y = "") +
  theme(axis.text.x = element_text(angle = 0, hjust = 0.5))
dev.off()

rm(detable, DGE_DATA, top20, DGE_cells, ord, sample_size)
```

# Temporal Dotplots

## General Dotplot - age

```{r}
data$groups <- paste0("cluster_", data$high_level_clusters_removed, "_age_", data$age_groups)

data$groups <- factor(data$groups, levels = c("cluster_1_age_5-6", "cluster_1_age_7-8", "cluster_1_age_9-11", "cluster_1_age_12-14",
                                              "cluster_2_age_5-6", "cluster_2_age_7-8", "cluster_2_age_9-11", "cluster_2_age_12-14",
                                              "cluster_3_age_5-6", "cluster_3_age_7-8", "cluster_3_age_9-11", "cluster_3_age_12-14",
                                              "cluster_4_35_age_5-6", "cluster_4_35_age_7-8", "cluster_4_35_age_9-11", "cluster_4_35_age_12-14",
                                              "cluster_5_age_5-6", "cluster_5_age_7-8", "cluster_5_age_9-11", "cluster_5_age_12-14",
                                              "cluster_7_age_5-6", "cluster_7_age_7-8", "cluster_7_age_9-11", "cluster_7_age_12-14",
                                              "cluster_8_age_5-6", "cluster_8_age_7-8", "cluster_8_age_9-11", "cluster_8_age_12-14",
                                              "cluster_9_age_5-6", "cluster_9_age_7-8", "cluster_9_age_9-11", "cluster_9_age_12-14",
                                              "cluster_10_age_5-6", "cluster_10_age_7-8", "cluster_10_age_9-11", "cluster_10_age_12-14",
                                              "cluster_11_age_5-6", "cluster_11_age_7-8", "cluster_11_age_9-11", "cluster_11_age_12-14",
                                              "cluster_12_age_5-6", "cluster_12_age_7-8", "cluster_12_age_9-11", "cluster_12_age_12-14",
                                              "cluster_13_age_5-6", "cluster_13_age_7-8", "cluster_13_age_9-11", "cluster_13_age_12-14",
                                              "cluster_14_age_5-6", "cluster_14_age_7-8", "cluster_14_age_9-11", "cluster_14_age_12-14",
                                              "cluster_15_age_5-6", "cluster_15_age_7-8", "cluster_15_age_9-11", "cluster_15_age_12-14",
                                              "cluster_16_age_5-6", "cluster_16_age_7-8", "cluster_16_age_9-11", "cluster_16_age_12-14",
                                              "cluster_17_age_5-6", "cluster_17_age_7-8", "cluster_17_age_9-11", "cluster_17_age_12-14",
                                              "cluster_18_1_age_5-6", "cluster_18_1_age_7-8", "cluster_18_1_age_9-11", "cluster_18_1_age_12-14",
                                              "cluster_18_2_age_5-6", "cluster_18_2_age_7-8", "cluster_18_2_age_9-11", "cluster_18_2_age_12-14",
                                              "cluster_20_age_5-6", "cluster_20_age_7-8", "cluster_20_age_9-11", "cluster_20_age_12-14",
                                              "cluster_21_age_5-6", "cluster_21_age_7-8", "cluster_21_age_9-11", "cluster_21_age_12-14",
                                              "cluster_22_age_5-6", "cluster_22_age_7-8", "cluster_22_age_9-11", "cluster_22_age_12-14",
                                              "cluster_23_age_5-6", "cluster_23_age_7-8", "cluster_23_age_9-11", "cluster_23_age_12-14",
                                              "cluster_24_age_5-6", "cluster_24_age_7-8", "cluster_24_age_9-11", "cluster_24_age_12-14",
                                              "cluster_25_age_5-6", "cluster_25_age_7-8", "cluster_25_age_9-11", "cluster_25_age_12-14",
                                              "cluster_26_age_5-6", "cluster_26_age_7-8", "cluster_26_age_9-11", "cluster_26_age_12-14",
                                              "cluster_28_age_5-6", "cluster_28_age_7-8", "cluster_28_age_9-11", "cluster_28_age_12-14",
                                              "cluster_29_age_5-6", "cluster_29_age_7-8", "cluster_29_age_9-11", "cluster_29_age_12-14",
                                              "cluster_30_age_5-6", "cluster_30_age_7-8", "cluster_30_age_9-11", "cluster_30_age_12-14",
                                              "cluster_31_age_5-6", "cluster_31_age_7-8", "cluster_31_age_9-11", "cluster_31_age_12-14",
                                              "cluster_32_age_5-6", "cluster_32_age_7-8", "cluster_32_age_9-11", "cluster_32_age_12-14",
                                              "cluster_33_age_5-6", "cluster_33_age_7-8", "cluster_33_age_9-11", "cluster_33_age_12-14"
                                              ))

table(data$groups)
unique(data$groups)

sample_size <- table(data$groups)
#sample_size[sample_size > 500] <- 500

# Randomly sample the cells that will be included when calculating DGE (same number of cells as selected above).
# Samples without replacement, so each cell only can be included once.
DGE_cells <- lapply(names(sample_size), function(x){
  set.seed(1)
  colnames(data)[data$groups == x]
  })
DGE_cells <- unlist(DGE_cells)

# Select all data that relates to the sampled DGE cells.
DGE_DATA <- data[, DGE_cells]


#Find differential expressed genes based on groups defined by "clusters_subset".
DGE_DATA <- SetIdent(DGE_DATA, value = "groups")

# The logfc threshold is set to 0.1 as compared to the default 0.25, which makes the detection of weaker signals possible (but it takes more computational power).
detable <- FindAllMarkers(DGE_DATA,
                          only.pos = TRUE,
                          max.cells.per.ident = 300,
                          logfc.threshold = .1,
                          assay = "RNA",
                          min.pct = 0.05)

 # Only keep DEG that have a p-value below 0.05.
detable <- detable[detable$p_val < 0.05, ]

# Adding two extra columns giving the difference (and log difference) between pct1 and pct2.
# pct1 is the percentage of cells in the specific cluster that express this gene.
# pct2 is the percentage of cells in all other clusters that express this gene.
detable$pct.diff <- detable$pct.1 - detable$pct.2
detable$log.pct.diff <- log2((detable$pct.1*99+1) / (detable$pct.2*99+1))

# Save the detable as a csv file.
write.csv2(detable, paste0(myfolder, "complementary/detable_general_age.csv"))

# Pick the 60 genes with the lowest p value for each age group, then 40 highest based on pct.diff, then 20 highest based on log.pct.diff.
# This gives a total of 140 genes (20 per age group). They seem to be unevenly spread over the cell clusters though.
# Why do we want to do it like this?
detable %>%
  group_by(cluster) %>%
  top_n(-40, p_val) %>%
  top_n(20, pct.diff) %>%
  top_n(5, avg_log2FC) ->
  top5

pdf(paste0(myfolder, "complementary/DGE_general_age.pdf"), width = 25, height = 75)
par(mfrow=c(1,1), mar=c(3,6,2,5))
DotPlot(DGE_DATA,  features = top5 %>% pull(gene) %>% unique()) +
  coord_flip() +
  theme(axis.text.y = element_text(angle = 0, hjust = 1, vjust = 0.5)) +
  scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 11, name = "RdBu") %>% rev()) +
  scale_x_discrete(limits = rev) +
  scale_y_discrete() +
  #labs(x = "", y = "") +
  theme(axis.text.x = element_text(angle = 90, hjust = 0.5))
dev.off()
```


## General Dotplot - New age groups

Since the subgroups for age groups 9-11 and 12-14 above gave rise to groups with very few cells. We decided to merge those into a single "latest" age group 9-14.

### Add additional age groups to the metadata (age_group_refined)

```{r}
df <- data.frame(age = c(5.5, 6, 7, 8, 8.5, 9, 10, 11.5, 12, 13.25, 14), age_group = c("5-6", "5-6", "7-8", "7-8", "7-8", "9-14", "9-14", "9-14", "9-14", "9-14", "9-14"))
age_to_age_group <- setNames(df$age_group, nm = df$age)
data$age_group_refined <- factor(age_to_age_group[as.character(data$age)], levels = c("5-6", "7-8", "9-14"))
rm(df, age_to_age_group)
gc()
```

```{r}
data$group_refined <- paste0("cluster_", data$high_level_clusters_removed, "_age_", data$age_group_refined)

data$group_refined <- factor(data$group_refined, levels = c(
                                              "cluster_1_age_5-6", "cluster_1_age_7-8", "cluster_1_age_9-14",
                                              "cluster_2_age_5-6", "cluster_2_age_7-8", "cluster_2_age_9-14",
                                              "cluster_3_age_5-6", "cluster_3_age_7-8", "cluster_3_age_9-14",
                                              "cluster_4_35_age_5-6", "cluster_4_35_age_7-8", "cluster_4_35_age_9-14",
                                              "cluster_5_age_5-6", "cluster_5_age_7-8", "cluster_5_age_9-14",
                                              "cluster_7_age_5-6", "cluster_7_age_7-8", "cluster_7_age_9-14",
                                              "cluster_8_age_5-6", "cluster_8_age_7-8", "cluster_8_age_9-14",
                                              "cluster_9_age_5-6", "cluster_9_age_7-8", "cluster_9_age_9-14",
                                              "cluster_10_age_5-6", "cluster_10_age_7-8", "cluster_10_age_9-14",
                                              "cluster_11_age_5-6", "cluster_11_age_7-8", "cluster_11_age_9-14",
                                              "cluster_12_age_5-6", "cluster_12_age_7-8", "cluster_12_age_9-14",
                                              "cluster_13_age_5-6", "cluster_13_age_7-8", "cluster_13_age_9-14",
                                              "cluster_14_age_5-6", "cluster_14_age_7-8", "cluster_14_age_9-14",
                                              "cluster_15_age_5-6", "cluster_15_age_7-8", "cluster_15_age_9-14",
                                              "cluster_16_age_5-6", "cluster_16_age_7-8", "cluster_16_age_9-14",
                                              "cluster_17_age_5-6", "cluster_17_age_7-8", "cluster_17_age_9-14",
                                              "cluster_18_1_age_5-6", "cluster_18_1_age_7-8", "cluster_18_1_age_9-14",
                                              "cluster_18_2_age_5-6", "cluster_18_2_age_7-8", "cluster_18_2_age_9-14",
                                              "cluster_20_age_5-6", "cluster_20_age_7-8", "cluster_20_age_9-14",
                                              "cluster_21_age_5-6", "cluster_21_age_7-8", "cluster_21_age_9-14",
                                              "cluster_22_age_5-6", "cluster_22_age_7-8", "cluster_22_age_9-14",
                                              "cluster_23_age_5-6", "cluster_23_age_7-8", "cluster_23_age_9-14",
                                              "cluster_24_age_5-6", "cluster_24_age_7-8", "cluster_24_age_9-14",
                                              "cluster_25_age_5-6", "cluster_25_age_7-8", "cluster_25_age_9-14",
                                              "cluster_26_age_5-6", "cluster_26_age_7-8", "cluster_26_age_9-14",
                                              "cluster_28_age_5-6", "cluster_28_age_7-8", "cluster_28_age_9-14",
                                              "cluster_29_age_5-6", "cluster_29_age_7-8", "cluster_29_age_9-14",
                                              "cluster_30_age_5-6", "cluster_30_age_7-8", "cluster_30_age_9-14",
                                              "cluster_31_age_5-6", "cluster_31_age_7-8", "cluster_31_age_9-14",
                                              "cluster_32_age_5-6", "cluster_32_age_7-8", "cluster_32_age_9-14",
                                              "cluster_33_age_5-6", "cluster_33_age_7-8", "cluster_33_age_9-14"
                                              ))

table(data$group_refined)
unique(data$group_refined)

sample_size <- table(data$group_refined)
#sample_size[sample_size > 500] <- 500

# Randomly sample the cells that will be included when calculating DGE (same number of cells as selected above).
# Samples without replacement, so each cell only can be included once.
DGE_cells <- lapply(names(sample_size), function(x){
  set.seed(1)
  colnames(data)[data$group_refined == x]
  })
DGE_cells <- unlist(DGE_cells)

# Select all data that relates to the sampled DGE cells.
DGE_DATA <- data[, DGE_cells]


#Find differential expressed genes based on groups defined by "clusters_subset".
DGE_DATA <- SetIdent(DGE_DATA, value = "group_refined")

# The logfc threshold is set to 0.1 as compared to the default 0.25, which makes the detection of weaker signals possible (but it takes more computational power).
detable <- FindAllMarkers(DGE_DATA,
                          only.pos = TRUE,
                          max.cells.per.ident = 300,
                          logfc.threshold = .1,
                          assay = "RNA",
                          min.pct = 0.05)

 # Only keep DEG that have a p-value below 0.05.
detable <- detable[detable$p_val < 0.05, ]

# Adding two extra columns giving the difference (and log difference) between pct1 and pct2.
# pct1 is the percentage of cells in the specific cluster that express this gene.
# pct2 is the percentage of cells in all other clusters that express this gene.
detable$pct.diff <- detable$pct.1 - detable$pct.2
detable$log.pct.diff <- log2((detable$pct.1*99+1) / (detable$pct.2*99+1))

# Save the detable as a csv file.
write.csv2(detable, paste0(myfolder, "complementary/detable_general_age_refined.csv"))

# Pick the 60 genes with the lowest p value for each age group, then 40 highest based on pct.diff, then 20 highest based on log.pct.diff.
# This gives a total of 140 genes (20 per age group). They seem to be unevenly spread over the cell clusters though.
# Why do we want to do it like this?
detable %>%
  group_by(cluster) %>%
  top_n(-40, p_val) %>%
  top_n(20, pct.diff) %>%
  top_n(5, avg_log2FC) ->
  top5

pdf(paste0(myfolder, "complementary/DGE_general_age_refined.pdf"), width = 25, height = 75)
par(mfrow=c(1,1), mar=c(3,6,2,5))
DotPlot(DGE_DATA,  features = top5 %>% pull(gene) %>% unique()) +
  coord_flip() +
  theme(axis.text.y = element_text(angle = 0, hjust = 1, vjust = 0.5)) +
  scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 11, name = "RdBu") %>% rev()) +
  scale_x_discrete(limits = rev) +
  scale_y_discrete() +
  #labs(x = "", y = "") +
  theme(axis.text.x = element_text(angle = 90, hjust = 0.5))
dev.off()
```


## CM temporal dotplot

### Add additional age groups to the metadata (age_group_refined)

```{r}
df <- data.frame(age = c(5.5, 6, 7, 8, 8.5, 9, 10, 11.5, 12, 13.25, 14), age_group = c("5-6", "5-6", "7-8", "7-8", "7-8", "9-14", "9-14", "9-14", "9-14", "9-14", "9-14"))
age_to_age_group <- setNames(df$age_group, nm = df$age)
CM$age_group_refined <- factor(age_to_age_group[as.character(CM$age)], levels = c("5-6", "7-8", "9-14"))
rm(df)
gc()
```

```{r}
CM$group_refined <- paste0("cluster_", CM$clusters_subset_split, "_age_", CM$age_group_refined)

CM$group_refined <- factor(CM$group_refined, levels = c(
                                              "cluster_6_age_5-6", "cluster_6_age_7-8", "cluster_6_age_9-14",
                                              "cluster_1_age_5-6", "cluster_1_age_7-8", "cluster_1_age_9-14",
                                              "cluster_10_age_5-6", "cluster_10_age_7-8", "cluster_10_age_9-14",
                                              "cluster_8-1_age_5-6", "cluster_8-1_age_7-8", "cluster_8-1_age_9-14",
                                              "cluster_8-3_age_5-6", "cluster_8-3_age_7-8", "cluster_8-3_age_9-14",
                                              "cluster_17_age_5-6", "cluster_17_age_7-8", "cluster_17_age_9-14",
                                              "cluster_8-2_age_5-6", "cluster_8-2_age_7-8", "cluster_8-2_age_9-14",
                                              "cluster_18_age_5-6", "cluster_18_age_7-8", "cluster_18_age_9-14",
                                              "cluster_4-2_age_5-6", "cluster_4-2_age_7-8", "cluster_4-2_age_9-14",
                                              "cluster_4-1_age_5-6", "cluster_4-1_age_7-8", "cluster_4-1_age_9-14",
                                              "cluster_3_age_5-6", "cluster_3_age_7-8", "cluster_3_age_9-14",
                                              "cluster_14_age_5-6", "cluster_14_age_7-8", "cluster_14_age_9-14",
                                              "cluster_9_age_5-6", "cluster_9_age_7-8", "cluster_9_age_9-14",
                                              "cluster_12_age_5-6", "cluster_12_age_7-8", "cluster_12_age_9-14",
                                              "cluster_19-2_age_5-6", "cluster_19-2_age_7-8", "cluster_19-2_age_9-14",
                                              "cluster_19-1_age_5-6", "cluster_19-1_age_7-8", "cluster_19-1_age_9-14",
                                              "cluster_7_age_5-6", "cluster_7_age_7-8", "cluster_7_age_9-14",
                                              "cluster_2_age_5-6", "cluster_2_age_7-8", "cluster_2_age_9-14",
                                              "cluster_5_age_5-6", "cluster_5_age_7-8", "cluster_5_age_9-14"

                                              # "cluster_11_age_5-6", "cluster_11_age_7-8", "cluster_11_age_9-14",
                                              # "cluster_13_age_5-6", "cluster_13_age_7-8", "cluster_13_age_9-14",
                                              # "cluster_15_age_5-6", "cluster_15_age_7-8", "cluster_15_age_9-14",
                                              # "cluster_16_age_5-6", "cluster_16_age_7-8", "cluster_16_age_9-14",
                                              # "cluster_20_age_5-6", "cluster_20_age_7-8", "cluster_20_age_9-14",
                                              # "cluster_21_age_5-6", "cluster_21_age_7-8", "cluster_21_age_9-14"
                                                                                            ))

table(CM$group_refined)
unique(CM$group_refined)

sample_size <- table(CM$group_refined)
#sample_size[sample_size > 500] <- 500

# Randomly sample the cells that will be included when calculating DGE (same number of cells as selected above).
# Samples without replacement, so each cell only can be included once.
DGE_cells <- lapply(names(sample_size), function(x){
  set.seed(1)
  colnames(CM)[CM$group_refined == x]
  })
DGE_cells <- unlist(DGE_cells)

# Select all data that relates to the sampled DGE cells.
DGE_DATA <- CM[, DGE_cells]


#Find differential expressed genes based on groups defined by "clusters_subset".
DGE_DATA <- SetIdent(DGE_DATA, value = "group_refined")

# The logfc threshold is set to 0.1 as compared to the default 0.25, which makes the detection of weaker signals possible (but it takes more computational power).
detable <- FindAllMarkers(DGE_DATA,
                          only.pos = TRUE,
                          max.cells.per.ident = 300,
                          logfc.threshold = .1,
                          assay = "RNA",
                          min.pct = 0.05)

 # Only keep DEG that have a p-value below 0.05.
detable <- detable[detable$p_val < 0.05, ]

# Adding two extra columns giving the difference (and log difference) between pct1 and pct2.
# pct1 is the percentage of cells in the specific cluster that express this gene.
# pct2 is the percentage of cells in all other clusters that express this gene.
detable$pct.diff <- detable$pct.1 - detable$pct.2
detable$log.pct.diff <- log2((detable$pct.1*99+1) / (detable$pct.2*99+1))

# Save the detable as a csv file.
write.csv2(detable, paste0(myfolder, "complementary/CM_detable_general_age_refined.csv"))

# Pick the 60 genes with the lowest p value for each age group, then 40 highest based on pct.diff, then 20 highest based on log.pct.diff.
# This gives a total of 140 genes (20 per age group). They seem to be unevenly spread over the cell clusters though.
# Why do we want to do it like this?
detable %>%
  group_by(cluster) %>%
  top_n(-40, p_val) %>%
  top_n(20, pct.diff) %>%
  top_n(5, avg_log2FC) ->
  top5

pdf(paste0(myfolder, "complementary/DGE_general_age_refined_CM.pdf"), width = 18, height = 55)
par(mfrow=c(1,1), mar=c(3,6,2,5))
DotPlot(DGE_DATA,  features = top5 %>% pull(gene) %>% unique()) +
  coord_flip() +
  theme(axis.text.y = element_text(angle = 0, hjust = 1, vjust = 0.5)) +
  scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 11, name = "RdBu") %>% rev()) +
  scale_x_discrete(limits = rev) +
  scale_y_discrete() +
  #labs(x = "", y = "") +
  theme(axis.text.x = element_text(angle = 90, hjust = 0.5))
dev.off()
```


### Markers from overall DE CM

```{r}
my_markers <- c("IL1RAPL1", "PCDH7", "KCNQ5", "NAV1", "SLC1A3", "MYH7", "PLCXD3", "PDK4", "MYL2", "MDH1", "ATP5PO", "ATP5F1C", "PPIF", "HECW2", "CREB5", "JAK2", "LRIF1", "HEY2", "FRMD5", "CEMIP2", "ASPM", "CENPF", "TOP2A", "APOLD1", "MIR924HG", "XPO4", "TENM3", "SPON1", "KCNIP4", "TNFRSF19", "RCAN1", "MIR100HG", "UNC5B-AS1", "AC109466.1", "BRINP3", "CSMD1", "SEMA3A", "SGCD", "LINC02208", "SIPA1L1", "MGAT4C", "GJA1", "PLXNA4", "MYL4", "SLC25A4", "CSRP3", "CKM", "NDUFB1", "NTM", "ANGPT1", "BMP10", "RELN", "SLIT3", "COL2A1", "PLCB1", "MYH6", "TRPM3", "AC009878.2", "CAMK1D", "CHST11", "POLG2", "LINC01619", "ADAMTSL4-AS1", "GIPC2", "SEMA4A", "ZNF385B", "LRRC4C", "TENM2", "RGS6", "CACNA1D", "ADAMTS19", "NR2F2-AS1", "NRXN3", "PGAM2", "PRDX1", "NPPA", "SNHG29", "YBX1", "VIM", "NACA", "FTL", "PTTG1", "UBE2C", "KPNA2", "CKS2", "CCNB1")

CM_subset <- CM[, CM$clusters_subset_split %in% c("6", "1", "10", "8-1", "8-3", "17", "8-2", "18", "4-2", "4-1", "3", "14", "9", "12", "19-2", "19-1", "7", "2", "5")]

CM_subset$group_refined <- paste0("cluster_", CM_subset$clusters_subset_split, "_age_", CM_subset$age_group_refined)

CM_subset$group_refined <- factor(CM_subset$group_refined, levels = c(
                                              "cluster_6_age_5-6", "cluster_6_age_7-8", "cluster_6_age_9-14",
                                              "cluster_1_age_5-6", "cluster_1_age_7-8", "cluster_1_age_9-14",
                                              "cluster_10_age_5-6", "cluster_10_age_7-8", "cluster_10_age_9-14",
                                              "cluster_8-1_age_5-6", "cluster_8-1_age_7-8", "cluster_8-1_age_9-14",
                                              "cluster_8-3_age_5-6", "cluster_8-3_age_7-8", "cluster_8-3_age_9-14",
                                              "cluster_17_age_5-6", "cluster_17_age_7-8", "cluster_17_age_9-14",
                                              "cluster_8-2_age_5-6", "cluster_8-2_age_7-8", "cluster_8-2_age_9-14",
                                              "cluster_18_age_5-6", "cluster_18_age_7-8", "cluster_18_age_9-14",
                                              "cluster_4-2_age_5-6", "cluster_4-2_age_7-8", "cluster_4-2_age_9-14",
                                              "cluster_4-1_age_5-6", "cluster_4-1_age_7-8", "cluster_4-1_age_9-14",
                                              "cluster_3_age_5-6", "cluster_3_age_7-8", "cluster_3_age_9-14",
                                              "cluster_14_age_5-6", "cluster_14_age_7-8", "cluster_14_age_9-14",
                                              "cluster_9_age_5-6", "cluster_9_age_7-8", "cluster_9_age_9-14",
                                              "cluster_12_age_5-6", "cluster_12_age_7-8", "cluster_12_age_9-14",
                                              "cluster_19-2_age_5-6", "cluster_19-2_age_7-8", "cluster_19-2_age_9-14",
                                              "cluster_19-1_age_5-6", "cluster_19-1_age_7-8", "cluster_19-1_age_9-14",
                                              "cluster_7_age_5-6", "cluster_7_age_7-8", "cluster_7_age_9-14",
                                              "cluster_2_age_5-6", "cluster_2_age_7-8", "cluster_2_age_9-14",
                                              "cluster_5_age_5-6", "cluster_5_age_7-8", "cluster_5_age_9-14"

                                              # "cluster_11_age_5-6", "cluster_11_age_7-8", "cluster_11_age_9-14",
                                              # "cluster_13_age_5-6", "cluster_13_age_7-8", "cluster_13_age_9-14",
                                              # "cluster_15_age_5-6", "cluster_15_age_7-8", "cluster_15_age_9-14",
                                              # "cluster_16_age_5-6", "cluster_16_age_7-8", "cluster_16_age_9-14",
                                              # "cluster_20_age_5-6", "cluster_20_age_7-8", "cluster_20_age_9-14",
                                              # "cluster_21_age_5-6", "cluster_21_age_7-8", "cluster_21_age_9-14"
                                                                                           ))

CM_subset <- SetIdent(CM_subset, value = "group_refined")

pdf(paste0(myfolder, "complementary/DGE_general_age_refined_CM_overall_markers.pdf"), width = 18, height = 25)
par(mfrow=c(1,1), mar=c(3,6,2,5))
DotPlot(CM_subset,  features = my_markers) +
  coord_flip() +
  theme(axis.text.y = element_text(angle = 0, hjust = 1, vjust = 0.5)) +
  scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 11, name = "RdBu") %>% rev()) +
  scale_x_discrete(limits = rev) +
  scale_y_discrete() +
  #labs(x = "", y = "") +
  theme(axis.text.x = element_text(angle = 90, hjust = 0.5))
dev.off()
```


## EN temporal dotplot

### Add additional age groups to the metadata (age_group_refined)

```{r}
df <- data.frame(age = c(5.5, 6, 7, 8, 8.5, 9, 10, 11.5, 12, 13.25, 14), age_group = c("5-6", "5-6", "7-8", "7-8", "7-8", "9-14", "9-14", "9-14", "9-14", "9-14", "9-14"))
age_to_age_group <- setNames(df$age_group, nm = df$age)
EN$age_group_refined <- factor(age_to_age_group[as.character(EN$age)], levels = c("5-6", "7-8", "9-14"))
rm(df)
gc()
```

```{r}
EN$group_refined <- paste0("cluster_", EN$clusters_subset, "_age_", EN$age_group_refined)

EN$group_refined <- factor(EN$group_refined, levels = c(
                                              "cluster_1_age_5-6", "cluster_1_age_7-8", "cluster_1_age_9-14",
                                              "cluster_12_age_5-6", "cluster_12_age_7-8", "cluster_12_age_9-14",
                                              "cluster_3_age_5-6", "cluster_3_age_7-8", "cluster_3_age_9-14",
                                              "cluster_6_age_5-6", "cluster_6_age_7-8", "cluster_6_age_9-14",
                                              "cluster_18_age_5-6", "cluster_18_age_7-8", "cluster_18_age_9-14",
                                              "cluster_17_age_5-6", "cluster_17_age_7-8", "cluster_17_age_9-14",
                                              "cluster_7_age_5-6", "cluster_7_age_7-8", "cluster_7_age_9-14",
                                              "cluster_19_age_5-6", "cluster_19_age_7-8", "cluster_19_age_9-14",
                                              "cluster_15_age_5-6", "cluster_15_age_7-8", "cluster_15_age_9-14",
                                              "cluster_13_age_5-6", "cluster_13_age_7-8", "cluster_13_age_9-14",
                                              "cluster_10_age_5-6", "cluster_10_age_7-8", "cluster_10_age_9-14",
                                              "cluster_2_age_5-6", "cluster_2_age_7-8", "cluster_2_age_9-14",
                                              "cluster_5_age_5-6", "cluster_5_age_7-8", "cluster_5_age_9-14",
                                              "cluster_21_age_5-6", "cluster_21_age_7-8", "cluster_21_age_9-14",
                                              "cluster_16_age_5-6", "cluster_16_age_7-8", "cluster_16_age_9-14",
                                              "cluster_20_age_5-6", "cluster_20_age_7-8", "cluster_20_age_9-14",
                                              "cluster_11_age_5-6", "cluster_11_age_7-8", "cluster_11_age_9-14",
                                              "cluster_14_age_5-6", "cluster_14_age_7-8", "cluster_14_age_9-14"

                                              # "cluster_4_age_5-6", "cluster_4_age_7-8", "cluster_4_age_9-14",
                                              # "cluster_9_age_5-6", "cluster_9_age_7-8", "cluster_9_age_9-14",
                                              # "cluster_8_age_5-6", "cluster_8_age_7-8", "cluster_8_age_9-14",
                                                                                            ))

table(EN$group_refined)
unique(EN$group_refined)

sample_size <- table(EN$group_refined)
#sample_size[sample_size > 500] <- 500

# Randomly sample the cells that will be included when calculating DGE (same number of cells as selected above).
# Samples without replacement, so each cell only can be included once.
DGE_cells <- lapply(names(sample_size), function(x){
  set.seed(1)
  colnames(EN)[EN$group_refined == x]
  })
DGE_cells <- unlist(DGE_cells)

# Select all data that relates to the sampled DGE cells.
DGE_DATA <- EN[, DGE_cells]


#Find differential expressed genes based on groups defined by "clusters_subset".
DGE_DATA <- SetIdent(DGE_DATA, value = "group_refined")

# The logfc threshold is set to 0.1 as compared to the default 0.25, which makes the detection of weaker signals possible (but it takes more computational power).
detable <- FindAllMarkers(DGE_DATA,
                          only.pos = TRUE,
                          max.cells.per.ident = 300,
                          logfc.threshold = .1,
                          assay = "RNA",
                          min.pct = 0.05)

# Only keep DEG that have a p-value below 0.05.
detable <- detable[detable$p_val < 0.05, ]

# Adding two extra columns giving the difference (and log difference) between pct1 and pct2.
# pct1 is the percentage of cells in the specific cluster that express this gene.
# pct2 is the percentage of cells in all other clusters that express this gene.
detable$pct.diff <- detable$pct.1 - detable$pct.2
detable$log.pct.diff <- log2((detable$pct.1*99+1) / (detable$pct.2*99+1))

# Save the detable as a csv file.
write.csv2(detable, paste0(myfolder, "complementary/EN_detable_general_age_refined.csv"))

# Pick the 60 genes with the lowest p value for each age group, then 40 highest based on pct.diff, then 20 highest based on log.pct.diff.
# This gives a total of 140 genes (20 per age group). They seem to be unevenly spread over the cell clusters though.
# Why do we want to do it like this?
detable %>%
  group_by(cluster) %>%
  top_n(-40, p_val) %>%
  top_n(20, pct.diff) %>%
  top_n(5, avg_log2FC) ->
  top5

# Genes get matched to one specific cluster, and the factor is leveled based on cluster number.
# ord <- factor(sapply(unique(as.character(top20$gene)), function(x){niceRplots::getcluster(DGE_DATA, x, "high_level_clusters_removed")}))

pdf(paste0(myfolder, "complementary/DGE_general_age_refined_EN.pdf"), width = 18, height = 55)
par(mfrow=c(1,1), mar=c(3,6,2,5))
DotPlot(DGE_DATA,  features = top5 %>% pull(gene) %>% unique()) +
  coord_flip() +
  theme(axis.text.y = element_text(angle = 0, hjust = 1, vjust = 0.5)) +
  scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 11, name = "RdBu") %>% rev()) +
  scale_x_discrete(limits = rev) +
  scale_y_discrete() +
  #labs(x = "", y = "") +
  theme(axis.text.x = element_text(angle = 90, hjust = 0.5))
dev.off()
```


### Markers from overall DE EN

```{r}
my_markers <- c("AC083864.5", "PLA2G5", "CDH11", "NTNG1", "PCDH7", "PLAC9", "LYPD2", "NPPC", "OPCML", "ENPP1", "NPR3", "SORCS1", "PRKG1", "KCNMB2", "NRG1", "PCDH15", "LRRC4C", "LINC02208", "NEGR1", "ALCAM", "CCDC80", "SEMA3A", "KCNQ5", "DKK2", "KISS1", "ASPN", "HAPLN1", "TMEM132D", "NCKAP5", "BMP4", "ENOX1", "SOX6", "GJA5", "FBLN5", "SULF1", "FBLN2", "IL1RAPL1", "GRIA2", "EYS", "EGFL8", "GABBR2", "BTNL9", "SLC9C1", "AL357507.1", "PRND", "ARHGAP18", "COL15A1", "RGCC", "KIT", "AL356124.1", "NEURL1B", "CA4", "CLIC5", "DIAPH3", "MKI67", "ASPM", "TOP2A", "CENPF", "RGS6", "APLNR", "TSHZ2", "FAM155A", "CA8", "NTS", "CPE", "NRG3", "KCNIP4", "TFF3", "RELN", "LYVE1", "CCL14", "ACTB", "ACTG1", "THUMPD3-AS1", "TMSB4X", "MOB1A", "SEMA4A", "PDE4C", "ADAMTSL4-AS1", "ATF3", "AL118558.1")


EN_subset <- EN[, EN$clusters_subset %in% c("1", "12", "3", "6", "18", "17", "7", "19", "15", "13", "10", "2", "5", "21", "16", "20", "11", "14")]

EN_subset$group_refined <- paste0("cluster_", EN_subset$clusters_subset, "_age_", EN_subset$age_group_refined)

EN_subset$group_refined <- factor(EN_subset$group_refined, levels = c(
                                              "cluster_1_age_5-6", "cluster_1_age_7-8", "cluster_1_age_9-14",
                                              "cluster_12_age_5-6", "cluster_12_age_7-8", "cluster_12_age_9-14",
                                              "cluster_3_age_5-6", "cluster_3_age_7-8", "cluster_3_age_9-14",
                                              "cluster_6_age_5-6", "cluster_6_age_7-8", "cluster_6_age_9-14",
                                              "cluster_18_age_5-6", "cluster_18_age_7-8", "cluster_18_age_9-14",
                                              "cluster_17_age_5-6", "cluster_17_age_7-8", "cluster_17_age_9-14",
                                              "cluster_7_age_5-6", "cluster_7_age_7-8", "cluster_7_age_9-14",
                                              "cluster_19_age_5-6", "cluster_19_age_7-8", "cluster_19_age_9-14",
                                              "cluster_15_age_5-6", "cluster_15_age_7-8", "cluster_15_age_9-14",
                                              "cluster_13_age_5-6", "cluster_13_age_7-8", "cluster_13_age_9-14",
                                              "cluster_10_age_5-6", "cluster_10_age_7-8", "cluster_10_age_9-14",
                                              "cluster_2_age_5-6", "cluster_2_age_7-8", "cluster_2_age_9-14",
                                              "cluster_5_age_5-6", "cluster_5_age_7-8", "cluster_5_age_9-14",
                                              "cluster_21_age_5-6", "cluster_21_age_7-8", "cluster_21_age_9-14",
                                              "cluster_16_age_5-6", "cluster_16_age_7-8", "cluster_16_age_9-14",
                                              "cluster_20_age_5-6", "cluster_20_age_7-8", "cluster_20_age_9-14",
                                              "cluster_11_age_5-6", "cluster_11_age_7-8", "cluster_11_age_9-14",
                                              "cluster_14_age_5-6", "cluster_14_age_7-8", "cluster_14_age_9-14"

                                              # "cluster_4_age_5-6", "cluster_4_age_7-8", "cluster_4_age_9-14",
                                              # "cluster_9_age_5-6", "cluster_9_age_7-8", "cluster_9_age_9-14",
                                              # "cluster_8_age_5-6", "cluster_8_age_7-8", "cluster_8_age_9-14",
                                                                                            ))

EN_subset <- SetIdent(EN_subset, value = "group_refined")

pdf(paste0(myfolder, "complementary/DGE_general_age_refined_EN_overall_markers.pdf"), width = 18, height = 25)
par(mfrow=c(1,1), mar=c(3,6,2,5))
DotPlot(EN_subset,  features = my_markers) +
  coord_flip() +
  theme(axis.text.y = element_text(angle = 0, hjust = 1, vjust = 0.5)) +
  scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 11, name = "RdBu") %>% rev()) +
  scale_x_discrete(limits = rev) +
  scale_y_discrete() +
  #labs(x = "", y = "") +
  theme(axis.text.x = element_text(angle = 90, hjust = 0.5))
dev.off()
```


## FB temporal dotplot

### Add additional age groups to the metadata (age_group_refined)

```{r}
df <- data.frame(age = c(5.5, 6, 7, 8, 8.5, 9, 10, 11.5, 12, 13.25, 14), age_group = c("5-6", "5-6", "7-8", "7-8", "7-8", "9-14", "9-14", "9-14", "9-14", "9-14", "9-14"))
age_to_age_group <- setNames(df$age_group, nm = df$age)
FB$age_group_refined <- factor(age_to_age_group[as.character(FB$age)], levels = c("5-6", "7-8", "9-14"))
rm(df)
gc()
```

```{r}
FB$group_refined <- paste0("cluster_", FB$clusters_subset, "_age_", FB$age_group_refined)

FB$group_refined <- factor(FB$group_refined, levels = c(
                                              "cluster_7_age_5-6", "cluster_7_age_7-8", "cluster_7_age_9-14",
                                              "cluster_3_age_5-6", "cluster_3_age_7-8", "cluster_3_age_9-14",
                                              "cluster_18_age_5-6", "cluster_18_age_7-8", "cluster_18_age_9-14",
                                              "cluster_14_age_5-6", "cluster_14_age_7-8", "cluster_14_age_9-14",
                                              "cluster_5_age_5-6", "cluster_5_age_7-8", "cluster_5_age_9-14",
                                              "cluster_9_age_5-6", "cluster_9_age_7-8", "cluster_9_age_9-14",
                                              "cluster_19_age_5-6", "cluster_19_age_7-8", "cluster_19_age_9-14",
                                              "cluster_2_age_5-6", "cluster_2_age_7-8", "cluster_2_age_9-14",
                                              "cluster_12_age_5-6", "cluster_12_age_7-8", "cluster_12_age_9-14",
                                              "cluster_4_age_5-6", "cluster_4_age_7-8", "cluster_4_age_9-14",
                                              "cluster_13_age_5-6", "cluster_13_age_7-8", "cluster_13_age_9-14",
                                              "cluster_11_age_5-6", "cluster_11_age_7-8", "cluster_11_age_9-14",
                                              "cluster_17_age_5-6", "cluster_17_age_7-8", "cluster_17_age_9-14",
                                              "cluster_8_age_5-6", "cluster_8_age_7-8", "cluster_8_age_9-14",
                                              "cluster_1_age_5-6", "cluster_1_age_7-8", "cluster_1_age_9-14",
                                              "cluster_6_age_5-6", "cluster_6_age_7-8", "cluster_6_age_9-14",
                                              "cluster_15_age_5-6", "cluster_15_age_7-8", "cluster_15_age_9-14",
                                              "cluster_20_age_5-6", "cluster_20_age_7-8", "cluster_20_age_9-14"
                                              
                                              # "cluster_10_age_5-6", "cluster_10_age_7-8", "cluster_10_age_9-14",
                                              # "cluster_21_age_5-6", "cluster_21_age_7-8", "cluster_21_age_9-14",
                                              # "cluster_16_age_5-6", "cluster_16_age_7-8", "cluster_16_age_9-14",
                                                                                            ))

table(FB$group_refined)
unique(FB$group_refined)

sample_size <- table(FB$group_refined)
#sample_size[sample_size > 500] <- 500

# Randomly sample the cells that will be included when calculating DGE (same number of cells as selected above).
# Samples without replacement, so each cell only can be included once.
DGE_cells <- lapply(names(sample_size), function(x){
  set.seed(1)
  colnames(FB)[FB$group_refined == x]
  })
DGE_cells <- unlist(DGE_cells)

# Select all data that relates to the sampled DGE cells.
DGE_DATA <- FB[, DGE_cells]


#Find differential expressed genes based on groups defined by "clusters_subset".
DGE_DATA <- SetIdent(DGE_DATA, value = "group_refined")

# The logfc threshold is set to 0.1 as compared to the default 0.25, which makes the detection of weaker signals possible (but it takes more computational power).
detable <- FindAllMarkers(DGE_DATA,
                          only.pos = TRUE,
                          max.cells.per.ident = 300,
                          logfc.threshold = .1,
                          assay = "RNA",
                          min.pct = 0.05)

# Only keep DEG that have a p-value below 0.05.
detable <- detable[detable$p_val < 0.05, ]

# Adding two extra columns giving the difference (and log difference) between pct1 and pct2.
# pct1 is the percentage of cells in the specific cluster that express this gene.
# pct2 is the percentage of cells in all other clusters that express this gene.
detable$pct.diff <- detable$pct.1 - detable$pct.2
detable$log.pct.diff <- log2((detable$pct.1*99+1) / (detable$pct.2*99+1))

# Save the detable as a csv file.
write.csv2(detable, paste0(myfolder, "complementary/FB_detable_general_age_refined.csv"))

# Pick the 60 genes with the lowest p value for each age group, then 40 highest based on pct.diff, then 20 highest based on log.pct.diff.
# This gives a total of 140 genes (20 per age group). They seem to be unevenly spread over the cell clusters though.
# Why do we want to do it like this?
detable %>%
  group_by(cluster) %>%
  top_n(-40, p_val) %>%
  top_n(20, pct.diff) %>%
  top_n(5, avg_log2FC) ->
  top5

# Genes get matched to one specific cluster, and the factor is leveled based on cluster number.
# ord <- factor(sapply(unique(as.character(top20$gene)), function(x){niceRplots::getcluster(DGE_DATA, x, "high_level_clusters_removed")}))

pdf(paste0(myfolder, "complementary/DGE_general_age_refined_FB.pdf"), width = 18, height = 55)
par(mfrow=c(1,1), mar=c(3,6,2,5))
DotPlot(DGE_DATA,  features = top5 %>% pull(gene) %>% unique()) +
  coord_flip() +
  theme(axis.text.y = element_text(angle = 0, hjust = 1, vjust = 0.5)) +
  scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 11, name = "RdBu") %>% rev()) +
  scale_x_discrete(limits = rev) +
  scale_y_discrete() +
  #labs(x = "", y = "") +
  theme(axis.text.x = element_text(angle = 90, hjust = 0.5))
dev.off()
```


### Markers from overall DE FB

```{r}
my_markers <- c("MGAT4C", "SLIT3", "DES", "LRRN3", "CCBE1", "DCC", "RYR2", "CEMIP", "FNDC1", "CCL2", "CEBPD", "ZFP36", "NFKB1", "CXCL2", "CLEC3B", "ROBO2", "SEMA3C", "ADAMTSL1", "CCN3", "SCN7A", "CCDC102B", "EGFEM1P", "COL24A1", "ABCA9", "COLEC11", "KCNMB2", "LUZP2", "MIA", "SOX6", "ITIH5", "MYRIP", "GPC5", "PTTG1", "UBE2C", "TOP2A", "CDK1", "CENPF", "DIAPH3", "MKI67", "KNL1", "ASPM", "APCDD1", "EDIL3", "HAPLN1", "TMEM132C", "SLN", "FGF14", "FHOD3", "ERBB4", "GPR158", "PDZRN4", "PENK", "FMOD", "CLU", "RSPO3", "PTPRT", "CALN1", "TECRL", "LRP1B", "SLC24A3", "BMPER", "CACNA2D3", "BRINP3", "NRXN1", "PCDH7", "MEIS1", "ROR1", "AL109930.1", "GRID2", "PTCH2", "GALNT17", "MEOX2", "BNC2", "DLK1", "SEMA5A", "PLA2G5", "HEYL", "ITGA1", "NFE2L3", "AL118558.1", "LINC01619", "PDE4C", "GIPC2")


FB_subset <- FB[, FB$clusters_subset %in% c("7", "3", "18", "14", "5", "9", "19", "2", "12", "4", "13", "11", "17", "8", "1", "6", "15", "20")]


FB_subset$group_refined <- paste0("cluster_", FB_subset$clusters_subset, "_age_", FB_subset$age_group_refined)

FB_subset$group_refined <- factor(FB_subset$group_refined, levels = c(
                                              "cluster_7_age_5-6", "cluster_7_age_7-8", "cluster_7_age_9-14",
                                              "cluster_3_age_5-6", "cluster_3_age_7-8", "cluster_3_age_9-14",
                                              "cluster_18_age_5-6", "cluster_18_age_7-8", "cluster_18_age_9-14",
                                              "cluster_14_age_5-6", "cluster_14_age_7-8", "cluster_14_age_9-14",
                                              "cluster_5_age_5-6", "cluster_5_age_7-8", "cluster_5_age_9-14",
                                              "cluster_9_age_5-6", "cluster_9_age_7-8", "cluster_9_age_9-14",
                                              "cluster_19_age_5-6", "cluster_19_age_7-8", "cluster_19_age_9-14",
                                              "cluster_2_age_5-6", "cluster_2_age_7-8", "cluster_2_age_9-14",
                                              "cluster_12_age_5-6", "cluster_12_age_7-8", "cluster_12_age_9-14",
                                              "cluster_4_age_5-6", "cluster_4_age_7-8", "cluster_4_age_9-14",
                                              "cluster_13_age_5-6", "cluster_13_age_7-8", "cluster_13_age_9-14",
                                              "cluster_11_age_5-6", "cluster_11_age_7-8", "cluster_11_age_9-14",
                                              "cluster_17_age_5-6", "cluster_17_age_7-8", "cluster_17_age_9-14",
                                              "cluster_8_age_5-6", "cluster_8_age_7-8", "cluster_8_age_9-14",
                                              "cluster_1_age_5-6", "cluster_1_age_7-8", "cluster_1_age_9-14",
                                              "cluster_6_age_5-6", "cluster_6_age_7-8", "cluster_6_age_9-14",
                                              "cluster_15_age_5-6", "cluster_15_age_7-8", "cluster_15_age_9-14",
                                              "cluster_20_age_5-6", "cluster_20_age_7-8", "cluster_20_age_9-14"
                                              
                                              # "cluster_10_age_5-6", "cluster_10_age_7-8", "cluster_10_age_9-14",
                                              # "cluster_21_age_5-6", "cluster_21_age_7-8", "cluster_21_age_9-14",
                                              # "cluster_16_age_5-6", "cluster_16_age_7-8", "cluster_16_age_9-14",
                                                                                            ))

FB_subset <- SetIdent(FB_subset, value = "group_refined")

pdf(paste0(myfolder, "complementary/DGE_general_age_refined_FB_overall_markers.pdf"), width = 18, height = 25)
par(mfrow=c(1,1), mar=c(3,6,2,5))
DotPlot(FB_subset,  features = my_markers) +
  coord_flip() +
  theme(axis.text.y = element_text(angle = 0, hjust = 1, vjust = 0.5)) +
  scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 11, name = "RdBu") %>% rev()) +
  scale_x_discrete(limits = rev) +
  scale_y_discrete() +
  #labs(x = "", y = "") +
  theme(axis.text.x = element_text(angle = 90, hjust = 0.5))
dev.off()
```


## Temporal DGE analysis (not used)

```{r}
# # Remove the problematic clusters (previous round)
# # cells_to_remove <- data$clusters_louvain_oi == 18 | data$clusters_louvain_oi == 31 | data$clusters_louvain_oi == 32
# # data <- data[,!cells_to_remove]
# 
# # Create groups based on both clusters and ages and save to the object as a factor.
# groups <- paste0(data$high_level_clusters, '_', as.character(data$age_groups) )
# data$subgroups <- factor(groups, levels = mixedsort(unique(groups)))
# sample_size <- table(data$subgroups)
# sample_size[ sample_size > 150 ] <- 150
# 
# DGE_cells <- lapply(names(sample_size), function(x){
#   set.seed(1)
#   sample( colnames(data) [ data$subgroups == x ] , size = sample_size[x])
#   })
# DGE_cells <- unlist(DGE_cells)
# 
# DGE_DATA <- data[, DGE_cells]
# 
# # Next time, try increasing max.cells.per.ident to 200
# # We set return.thres (p-val) to 1 to include everything and try to avoid an error that made the function crash because of not enough cells in age group 10. Filtering can be done later. logfc.threshold could also be modified and filtered later.
# detable_pat <- lapply(unique(DGE_DATA$high_level_clusters),DGE_DATA=DGE_DATA, function(x,DGE_DATA){
#   message(x)
#   temp <- DGE_DATA[,DGE_DATA$high_level_clusters == x]
#   print(temp)
#   temp <- SetIdent( temp , value = "age_groups")
#   detable <- 0
#   try(detable <- FindAllMarkers(temp,
#                                 only.pos = T,
#                                 logfc.threshold = .1,
#                                 assay = "RNA",
#                                 min.pct = 0.05,
#                                 return.thresh = 1),
#       silent = TRUE)
# 
#   if(detable != 0){
#     return( cbind(detable,cell_cluster= x) )}
# })
# detable <- do.call(rbind,detable_pat)
# 
# detable <- detable[ detable$p_val < 0.05,  ]
# detable$pct.diff <- detable$pct.1 - detable$pct.2
# detable$log.pct.diff <- log2( (detable$pct.1+1) / (detable$pct.2+1) )
# write.csv2(detable, paste0(myfolder, "detable_for_each_cluster_over_time.csv"))
# 
# # Save results as a dot plot
# detable$groups <- paste0(detable$cell_cluster,detable$cluster)
# 
# detable %>%
#   group_by(groups) %>%
#   top_n(-60, p_val) %>%
#   top_n(40, pct.diff) %>%
#   top_n(20, log.pct.diff) ->
#   top20
# 
# # ord <- factor(sapply(unique(as.character(top5$gene)),function(x){getcluster(DGE_DATA, x, "subgroups")}))
# ord <- getcluster(data = DGE_DATA, genes = unique(as.character(top20$gene)), clustering = "subgroups")
# ord
# 
# pdf( paste0(myfolder, "complementary/DGE_temporal.pdf"),width = 60,height = length(ord)/6+3)
# mypar(1,1,mar=c(14,10,1,5))
# plot_dots(DGE_DATA, unique(names(ord)[order(ord)]), clustering = "subgroups", show_grid = T,main = "top cluster markers over time",cex.main=1,font.main=1,cex.col = 1,srt = 90,cex.row = 1.1)
# abline(v=cumsum(c(table(sub( "_.*","",names(table(DGE_DATA$subgroups))))))+0.5)
# dev.off()
```


## CM Dotplot - "8-2", "18", "19-1" - temporal

```{r}
my_markers <- c("ACTA1", "TNFRSF19", "CNN1", "XPO4", "SORBS2", "ZFP36L1", "IRX1", "IRX2", "RCAN1", "MYOM2", "BMP2", "ATP1A1", "HAMP", "FZD7", "TBX3", "PITX2", "SPON1", "VCAM1")

CM_subset <- CM[, CM$clusters_subset_split %in% c("8-2", "18", "19-1")]

CM_subset$group_refined <- paste0("cluster_", CM_subset$clusters_subset_split, "_age_", CM_subset$age_group_refined)

CM_subset$group_refined <- factor(CM_subset$group_refined, levels = c(
                                              "cluster_8-2_age_5-6", "cluster_8-2_age_7-8", "cluster_8-2_age_9-14",
                                              "cluster_18_age_5-6", "cluster_18_age_7-8", "cluster_18_age_9-14",
                                              "cluster_19-1_age_5-6", "cluster_19-1_age_7-8", "cluster_19-1_age_9-14"
                                                                                           ))

CM_subset <- SetIdent(CM_subset, value = "group_refined")

pdf(paste0(myfolder, "complementary/DGE_general_age_refined_CM_8-2_18_19-1_selection_markers.pdf"), width = 5.25, height = 5.5)
par(mfrow=c(1,1), mar=c(3,6,2,5))
DotPlot(CM_subset,  features = my_markers) +
  coord_flip() +
  theme(axis.text.y = element_text(angle = 0, hjust = 1, vjust = 0.5)) +
  scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 11, name = "RdBu") %>% rev()) +
  scale_x_discrete(limits = rev) +
  scale_y_discrete() +
  #labs(x = "", y = "") +
  theme(axis.text.x = element_text(angle = 90, hjust = 0.5))
dev.off()
```





## CM Dotplot - ion channels

```{r}
# my_markers <- c("ABCC8", "ATP1A3", "ATP2A2", "CACNA1C", "KCNIP2", "CACNA2D2", "CACNA1D", "CACNA1G", "CALM3", "CASQ1", "CLCN2", "CLCN6", "GJA1", "GJA4", "GJA5", "GJC1", "GJC3", "HCN1", "HCN3", "HCN4", "ITPR1", "ITPR3", "KCNA2", "KCNA3", "KCNA4", "KCNA5", "KCNA6", "KCNAB3", "KCND2", "KCND3", "KCNE2", "KCNE4", "KCNE5", "KCNJ11", "KCNJ13", "KCNJ2", "KCNJ3", "KCNJ4", "KCNJ5", "KCNJ8", "KCNK1", "KCNK3", "KCNK5", "KCNN2", "PLN", "PPP3CA", "RCAN1", "RYR2", "SCN1B", "SCN9A", "SLC8A1", "ADRB1", "CHRM2")

my_markers <- c("HCN1", "HCN4", "SCN5A", "CACNA1C", "CACNA2D1", "CACNA2D2", "CACNA1D", "CACNA1G", "CACNB2", "KCNJ3", "KCNN2", "KCNQ1", "KCNH7", "KCNA5", "KCNJ5", "KCNQ3", "KCNJ2", "KCNA4", "KCNQ5", "CLCN3", "CLCN6", "GJC1", "GJA5", "GJA1", "ATP2A2", "SLC8A1", "ITPR1", "RYR2", "ADRB1", "CHRM2")

CM_subset <- CM[, CM$clusters_subset_split %in% c("19-2",
                                                  "19-1",
                                                  "21",
                                                  "12",
                                                  "14",
                                                  "9",
                                                  "4-1",
                                                  "4-2",
                                                  "18",
                                                  "8-2",
                                                  "8-1",
                                                  "8-3",
                                                  "6",
                                                  "1",
                                                  "10")]

CM_subset$clusters_subset_split <- factor(CM_subset$clusters_subset_split, levels = c(
                                                  "19-2",
                                                  "19-1",
                                                  "21",
                                                  "12",
                                                  "14",
                                                  "9",
                                                  "4-1",
                                                  "4-2",
                                                  "18",
                                                  "8-2",
                                                  "8-1",
                                                  "8-3",
                                                  "6",
                                                  "1",
                                                  "10"))
CM_subset <- SetIdent(CM_subset, value = "clusters_subset_split")

pdf(paste0(myfolder, "complementary/DGE_CM_ion_channel.pdf"), width = 11, height = 4)
DotPlot(CM_subset, features = my_markers) +
  theme(axis.text.y = element_text(angle = 0, hjust = 1, vjust = 0.5)) +
  scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 11, name = "RdBu") %>% rev()) +
  scale_x_discrete() +
  scale_y_discrete(limits = rev) +
  labs(x = "", y = "") +
  theme(axis.text.x = element_text(angle = 30, hjust = 1, size = 8))
dev.off()

gc()
```


### subset 1

```{r}
my_markers <- c("HCN1", "HCN4", "SCN5A", "CACNA1C", "CACNA1D", "CACNA1G", "CACNA2D1", "CACNA2D2", "CACNB2", "RYR2", "KCNA5", "KCNH7", "KCNJ2", "KCNJ3", "KCNJ4", "KCNMB2", "KCNN2", "KCNQ1", "KCNQ3", "KCNQ5", "CLCN3", "CLCN6", "CLIC1", "CLIC4", "CLIC5", "ANO4", "ANO5", "ANO6", "GRIA1", "GRID1", "GRID2", "GRIK1", "GRIK2", "GRIN2B", "P2RX1", "GABRB2", "GABRB3", "GABRE", "LRRC8B", "TRPC1", "TRPC3", "TRPM3", "TRPM4", "TRPM7", "PKD2", "GJA1", "GJA3", "GJA5", "GJC1", "AQP1", "KCNT2")

CM_subset <- CM[, CM$clusters_subset_split %in% c("19-2",
                                                  "19-1",
                                                  "12",
                                                  "9",
                                                  "14",
                                                  "4-2",
                                                  "4-1",
                                                  "18",
                                                  "8-2",
                                                  "8-1",
                                                  "8-3",
                                                  "17",
                                                  "6",
                                                  "1",
                                                  "10",
                                                  "3",
                                                  "7",
                                                  "2", 
                                                  "5")]

CM_subset$clusters_subset_split <- factor(CM_subset$clusters_subset_split, levels = c(
                                                  "19-2",
                                                  "19-1",
                                                  "12",
                                                  "9",
                                                  "14",
                                                  "4-2",
                                                  "4-1",
                                                  "18",
                                                  "8-2",
                                                  "8-1",
                                                  "8-3",
                                                  "17",
                                                  "6",
                                                  "1",
                                                  "10",
                                                  "3",
                                                  "7",
                                                  "2", 
                                                  "5"))

CM_subset <- SetIdent(CM_subset, value = "clusters_subset_split")

pdf(paste0(myfolder, "complementary/DGE_CM_ion_channel_subset_1.pdf"), width = 12, height = 5)
DotPlot(CM_subset, features = my_markers) +
  theme(axis.text.y = element_text(angle = 0, hjust = 1, vjust = 0.5)) +
  scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 11, name = "RdBu") %>% rev()) +
  scale_x_discrete() +
  scale_y_discrete(limits = rev) +
  labs(x = "", y = "") +
  theme(axis.text.x = element_text(angle = 30, hjust = 1, size = 8))
dev.off()

gc()
```


### subset 1_1

```{r}
my_markers <- c("HCN1", "HCN4", "SCN5A", "CACNA1C", "CACNA1D", "CACNA1G", "CACNA2D1", "CACNA2D2", "CACNB2", "RYR2", "KCNA5", "KCNH7", "KCNJ2", "KCNJ3", "KCNJ4", "KCNMB2", "KCNN2", "KCNQ1", "KCNQ3", "KCNQ5", "CLCN3", "CLCN6", "CLIC1", "CLIC4", "CLIC5", "ANO4", "ANO5", "ANO6", "GRIA1", "GRID1", "GRID2", "GRIK1", "GRIK2", "GRIN2B", "P2RX1", "GABRB2", "GABRB3", "GABRE", "LRRC8B", "TRPC1", "TRPC3", "TRPM3", "TRPM4", "TRPM7", "PKD2", "GJA1", "GJA3", "GJA5", "GJC1", "AQP1", "KCNT2")

CM_subset <- CM[, CM$clusters_subset_split %in% c("4-2",
                                                  "4-1",
                                                  "18",
                                                  "6")]

CM_subset$clusters_subset_split <- factor(CM_subset$clusters_subset_split, levels = c("4-2","4-1","18","6"))

CM_subset <- SetIdent(CM_subset, value = "clusters_subset_split")

pdf(paste0(myfolder, "complementary/DGE_CM_ion_channel_subset_1_1.pdf"), width = 12, height = 3)
DotPlot(CM_subset, features = my_markers) +
  theme(axis.text.y = element_text(angle = 0, hjust = 1, vjust = 0.5)) +
  scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 11, name = "RdBu") %>% rev()) +
  scale_x_discrete() +
  scale_y_discrete(limits = rev) +
  labs(x = "", y = "") +
  theme(axis.text.x = element_text(angle = 30, hjust = 1, size = 8))
dev.off()

gc()
```


### subset 1_2

```{r}
my_markers <- c("HCN1", "HCN4", "SCN5A", "CACNA1C", "CACNA1D", "CACNA1G", "CACNA2D1", "CACNA2D2", "CACNB2", "RYR2", "KCNA5", "KCNH7", "KCNJ2", "KCNJ3", "KCNJ4", "KCNMB2", "KCNN2", "KCNQ1", "KCNQ3", "KCNQ5", "CLCN3", "CLCN6", "CLIC1", "CLIC4", "CLIC5", "ANO4", "ANO5", "ANO6", "GRIA1", "GRID1", "GRID2", "GRIK1", "GRIK2", "GRIN2B", "P2RX1", "GABRB2", "GABRB3", "GABRE", "LRRC8B", "TRPC1", "TRPC3", "TRPM3", "TRPM4", "TRPM7", "PKD2", "GJA1", "GJA3", "GJA5", "GJC1", "AQP1", "KCNT2")

CM_subset <- CM[, CM$clusters_subset_split %in% c("19-2",
                                                  "19-1",
                                                  "12",
                                                  "9",
                                                  "14")]

CM_subset$clusters_subset_split <- factor(CM_subset$clusters_subset_split, levels = c("19-2","19-1","12","9","14"))

CM_subset <- SetIdent(CM_subset, value = "clusters_subset_split")

pdf(paste0(myfolder, "complementary/DGE_CM_ion_channel_subset_1_2.pdf"), width = 12, height = 3)
DotPlot(CM_subset, features = my_markers) +
  theme(axis.text.y = element_text(angle = 0, hjust = 1, vjust = 0.5)) +
  scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 11, name = "RdBu") %>% rev()) +
  scale_x_discrete() +
  scale_y_discrete(limits = rev) +
  labs(x = "", y = "") +
  theme(axis.text.x = element_text(angle = 30, hjust = 1, size = 8))
dev.off()

gc()
```


### subset 1_3

```{r}
my_markers <- c("HCN1", "HCN4", "SCN5A", "CACNA1C", "CACNA1D", "CACNA1G", "CACNA2D1", "CACNA2D2", "CACNB2", "RYR2", "KCNA5", "KCNH7", "KCNJ2", "KCNJ3", "KCNJ4", "KCNMB2", "KCNN2", "KCNQ1", "KCNQ3", "KCNQ5", "CLCN3", "CLCN6", "CLIC1", "CLIC4", "CLIC5", "ANO4", "ANO5", "ANO6", "GRIA1", "GRID1", "GRID2", "GRIK1", "GRIK2", "GRIN2B", "P2RX1", "GABRB2", "GABRB3", "GABRE", "LRRC8B", "TRPC1", "TRPC3", "TRPM3", "TRPM4", "TRPM7", "PKD2", "GJA1", "GJA3", "GJA5", "GJC1", "AQP1", "KCNT2")

CM_subset <- CM[, CM$clusters_subset_split %in% c("4-1",
                                                  "4-2",
                                                  "6")]

CM_subset$clusters_subset_split <- factor(CM_subset$clusters_subset_split, levels = c("4-1",
                                                  "4-2",
                                                  "6"))

CM_subset <- SetIdent(CM_subset, value = "clusters_subset_split")

pdf(paste0(myfolder, "complementary/DGE_CM_ion_channel_subset_1_3.pdf"), width = 12, height = 3)
DotPlot(CM_subset, features = my_markers) +
  theme(axis.text.y = element_text(angle = 0, hjust = 1, vjust = 0.5)) +
  scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 11, name = "RdBu") %>% rev()) +
  scale_x_discrete() +
  scale_y_discrete(limits = rev) +
  labs(x = "", y = "") +
  theme(axis.text.x = element_text(angle = 30, hjust = 1, size = 8))
dev.off()

gc()
```


### subset 1_4

```{r}
my_markers <- c("HCN1", "HCN4", "SCN5A", "CACNA1C", "CACNA1D", "CACNA1G", "CACNA2D1", "CACNA2D2", "CACNB2", "RYR2", "KCNA5", "KCNH7", "KCNJ2", "KCNJ3", "KCNJ4", "KCNMB2", "KCNN2", "KCNQ1", "KCNQ3", "KCNQ5", "CLCN3", "CLCN6", "CLIC1", "CLIC4", "CLIC5", "ANO4", "ANO5", "ANO6", "GRIA1", "GRID1", "GRID2", "GRIK1", "GRIK2", "GRIN2B", "P2RX1", "GABRB2", "GABRB3", "GABRE", "LRRC8B", "TRPC1", "TRPC3", "TRPM3", "TRPM4", "TRPM7", "PKD2", "GJA1", "GJA3", "GJA5", "GJC1", "AQP1", "KCNT2")

CM_subset <- CM[, CM$clusters_subset_split %in% c("19-2", "19-1", "12", "9", "14", "18", "4-2", "4-1", "6")]

CM_subset$clusters_subset_split <- factor(CM_subset$clusters_subset_split, levels = c("19-2", "19-1", "12", "9", "14", "18", "4-2", "4-1", "6"))

CM_subset <- SetIdent(CM_subset, value = "clusters_subset_split")

pdf(paste0(myfolder, "complementary/DGE_CM_ion_channel_subset_1_4.pdf"), width = 12, height = 4)
DotPlot(CM_subset, features = my_markers) +
  theme(axis.text.y = element_text(angle = 0, hjust = 1, vjust = 0.5)) +
  scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 11, name = "RdBu") %>% rev()) +
  scale_x_discrete() +
  scale_y_discrete(limits = rev) +
  labs(x = "", y = "") +
  theme(axis.text.x = element_text(angle = 30, hjust = 1, size = 8))
dev.off()

gc()
```


## CM Dotplot - GPRC

### subset 1

```{r}
my_markers <- c("GNA12", "GNAI1", "GNAL", "GNAO1", "GNAQ", "GNAS", "GNB2", "GNB4", "GNB5", "GNG5", "GNG7", "GNG10", "GNG11", "GNG12", "RGS2", "RGS3", "RGS5", "RGS6", "RGS9", "RGS12")

CM_subset <- CM[, CM$clusters_subset_split %in% c("19-2",
                                                  "19-1",
                                                  "12",
                                                  "9",
                                                  "14",
                                                  "4-2",
                                                  "4-1",
                                                  "18",
                                                  "8-2",
                                                  "8-1",
                                                  "8-3",
                                                  "17",
                                                  "6",
                                                  "1",
                                                  "10",
                                                  "3",
                                                  "7",
                                                  "2", 
                                                  "5")]

CM_subset$clusters_subset_split <- factor(CM_subset$clusters_subset_split, levels = c(
                                                  "19-2",
                                                  "19-1",
                                                  "12",
                                                  "9",
                                                  "14",
                                                  "4-2",
                                                  "4-1",
                                                  "18",
                                                  "8-2",
                                                  "8-1",
                                                  "8-3",
                                                  "17",
                                                  "6",
                                                  "1",
                                                  "10",
                                                  "3",
                                                  "7",
                                                  "2", 
                                                  "5"))

CM_subset <- SetIdent(CM_subset, value = "clusters_subset_split")

pdf(paste0(myfolder, "complementary/DGE_CM_GPRC_subset_1.pdf"), width = 7, height = 4.5)
DotPlot(CM_subset, features = my_markers) +
  theme(axis.text.y = element_text(angle = 0, hjust = 1, vjust = 0.5)) +
  scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 11, name = "RdBu") %>% rev()) +
  scale_x_discrete() +
  scale_y_discrete(limits = rev) +
  labs(x = "", y = "") +
  theme(axis.text.x = element_text(angle = 30, hjust = 1, size = 8))
dev.off()

gc()
```


### subset 1_1

```{r}
my_markers <- c("GNA12", "GNAI1", "GNAL", "GNAO1", "GNAQ", "GNAS", "GNB2", "GNB4", "GNB5", "GNG5", "GNG7", "GNG10", "GNG11", "GNG12", "RGS2", "RGS3", "RGS5", "RGS6", "RGS9", "RGS12")

CM_subset <- CM[, CM$clusters_subset_split %in% c("4-2",
                                                  "4-1",
                                                  "18",
                                                  "6")]

CM_subset$clusters_subset_split <- factor(CM_subset$clusters_subset_split, levels = c("4-2","4-1","18","6"))

CM_subset <- SetIdent(CM_subset, value = "clusters_subset_split")

pdf(paste0(myfolder, "complementary/DGE_CM_GPRC_subset_1_1.pdf"), width = 7, height = 3)
DotPlot(CM_subset, features = my_markers) +
  theme(axis.text.y = element_text(angle = 0, hjust = 1, vjust = 0.5)) +
  scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 11, name = "RdBu") %>% rev()) +
  scale_x_discrete() +
  scale_y_discrete(limits = rev) +
  labs(x = "", y = "") +
  theme(axis.text.x = element_text(angle = 30, hjust = 1, size = 8))
dev.off()

gc()
```


### subset 1_2

```{r}
my_markers <- c("GNA12", "GNAI1", "GNAL", "GNAO1", "GNAQ", "GNAS", "GNB2", "GNB4", "GNB5", "GNG5", "GNG7", "GNG10", "GNG11", "GNG12", "RGS2", "RGS3", "RGS5", "RGS6", "RGS9", "RGS12")

CM_subset <- CM[, CM$clusters_subset_split %in% c("19-2",
                                                  "19-1",
                                                  "12",
                                                  "9",
                                                  "14")]

CM_subset$clusters_subset_split <- factor(CM_subset$clusters_subset_split, levels = c("19-2","19-1","12","9","14"))

CM_subset <- SetIdent(CM_subset, value = "clusters_subset_split")

pdf(paste0(myfolder, "complementary/DGE_CM_GPRC_subset_1_2.pdf"), width = 7, height = 3)
DotPlot(CM_subset, features = my_markers) +
  theme(axis.text.y = element_text(angle = 0, hjust = 1, vjust = 0.5)) +
  scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 11, name = "RdBu") %>% rev()) +
  scale_x_discrete() +
  scale_y_discrete(limits = rev) +
  labs(x = "", y = "") +
  theme(axis.text.x = element_text(angle = 30, hjust = 1, size = 8))
dev.off()

gc()
```


### subset 1_3

```{r}
my_markers <- c("GNA12", "GNAI1", "GNAL", "GNAO1", "GNAQ", "GNAS", "GNB2", "GNB4", "GNB5", "GNG5", "GNG7", "GNG10", "GNG11", "GNG12", "RGS2", "RGS3", "RGS5", "RGS6", "RGS9", "RGS12")

CM_subset <- CM[, CM$clusters_subset_split %in% c("4-1",
                                                  "4-2",
                                                  "6")]

CM_subset$clusters_subset_split <- factor(CM_subset$clusters_subset_split, levels = c("4-1",
                                                  "4-2",
                                                  "6"))

CM_subset <- SetIdent(CM_subset, value = "clusters_subset_split")

pdf(paste0(myfolder, "complementary/DGE_CM_GPRC_subset_1_3.pdf"), width = 7, height = 3)
DotPlot(CM_subset, features = my_markers) +
  theme(axis.text.y = element_text(angle = 0, hjust = 1, vjust = 0.5)) +
  scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 11, name = "RdBu") %>% rev()) +
  scale_x_discrete() +
  scale_y_discrete(limits = rev) +
  labs(x = "", y = "") +
  theme(axis.text.x = element_text(angle = 30, hjust = 1, size = 8))
dev.off()

gc()
```


### subset 2

```{r}
my_markers <- c("ADRB1", "CHRM2", "ADORA1", "ADORA2A", "EDNRA", "HRH2", "HTR4", "PTHER3", "GLP1R", "PTH1R", "VIPR2", "GABBR1", "GRM7", "FZD1", "FZD2", "FZD6", "FZD7", "FZD8", "ADGRA3", "ADGRB2", "ADGRB3", "ADGRD1", "ADGRG6", "ADGRL1", "ADGRL2", "ADGRL3", "ADGRV1", "LGR4", "GPR22", "GPR39", "GPR158", "GPR173", "GPRC5A", "GPRC5B")

CM_subset <- CM[, CM$clusters_subset_split %in% c("19-2",
                                                  "19-1",
                                                  "12",
                                                  "9",
                                                  "14",
                                                  "4-2",
                                                  "4-1",
                                                  "18",
                                                  "8-2",
                                                  "8-1",
                                                  "8-3",
                                                  "17",
                                                  "6",
                                                  "1",
                                                  "10",
                                                  "3",
                                                  "7",
                                                  "2", 
                                                  "5")]

CM_subset$clusters_subset_split <- factor(CM_subset$clusters_subset_split, levels = c(
                                                  "19-2",
                                                  "19-1",
                                                  "12",
                                                  "9",
                                                  "14",
                                                  "4-2",
                                                  "4-1",
                                                  "18",
                                                  "8-2",
                                                  "8-1",
                                                  "8-3",
                                                  "17",
                                                  "6",
                                                  "1",
                                                  "10",
                                                  "3",
                                                  "7",
                                                  "2", 
                                                  "5"))

CM_subset <- SetIdent(CM_subset, value = "clusters_subset_split")

pdf(paste0(myfolder, "complementary/DGE_CM_GPRC_subset_2.pdf"), width = 9, height = 5)
DotPlot(CM_subset, features = my_markers) +
  theme(axis.text.y = element_text(angle = 0, hjust = 1, vjust = 0.5)) +
  scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 11, name = "RdBu") %>% rev()) +
  scale_x_discrete() +
  scale_y_discrete(limits = rev) +
  labs(x = "", y = "") +
  theme(axis.text.x = element_text(angle = 30, hjust = 1, size = 8))
dev.off()

gc()
```


### subset 2_1

```{r}
my_markers <- c("ADRB1", "CHRM2", "ADORA1", "ADORA2A", "EDNRA", "HRH2", "HTR4", "PTHER3", "GLP1R", "PTH1R", "VIPR2", "GABBR1", "GRM7", "FZD1", "FZD2", "FZD6", "FZD7", "FZD8", "ADGRA3", "ADGRB2", "ADGRB3", "ADGRD1", "ADGRG6", "ADGRL1", "ADGRL2", "ADGRL3", "ADGRV1", "LGR4", "GPR22", "GPR39", "GPR158", "GPR173", "GPRC5A", "GPRC5B")

CM_subset <- CM[, CM$clusters_subset_split %in% c("4-2",
                                                  "4-1",
                                                  "18",
                                                  "6")]

CM_subset$clusters_subset_split <- factor(CM_subset$clusters_subset_split, levels = c("4-2","4-1","18","6"))

CM_subset <- SetIdent(CM_subset, value = "clusters_subset_split")

pdf(paste0(myfolder, "complementary/DGE_CM_GPRC_subset_2_1.pdf"), width = 9, height = 3)
DotPlot(CM_subset, features = my_markers) +
  theme(axis.text.y = element_text(angle = 0, hjust = 1, vjust = 0.5)) +
  scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 11, name = "RdBu") %>% rev()) +
  scale_x_discrete() +
  scale_y_discrete(limits = rev) +
  labs(x = "", y = "") +
  theme(axis.text.x = element_text(angle = 30, hjust = 1, size = 8))
dev.off()

gc()
```


### subset 2_2

```{r}
my_markers <- c("ADRB1", "CHRM2", "ADORA1", "ADORA2A", "EDNRA", "HRH2", "HTR4", "PTHER3", "GLP1R", "PTH1R", "VIPR2", "GABBR1", "GRM7", "FZD1", "FZD2", "FZD6", "FZD7", "FZD8", "ADGRA3", "ADGRB2", "ADGRB3", "ADGRD1", "ADGRG6", "ADGRL1", "ADGRL2", "ADGRL3", "ADGRV1", "LGR4", "GPR22", "GPR39", "GPR158", "GPR173", "GPRC5A", "GPRC5B")

CM_subset <- CM[, CM$clusters_subset_split %in% c("19-2",
                                                  "19-1",
                                                  "12",
                                                  "9",
                                                  "14")]

CM_subset$clusters_subset_split <- factor(CM_subset$clusters_subset_split, levels = c("19-2","19-1","12","9","14"))

CM_subset <- SetIdent(CM_subset, value = "clusters_subset_split")

pdf(paste0(myfolder, "complementary/DGE_CM_GPRC_subset_2_2.pdf"), width = 12, height = 3)
DotPlot(CM_subset, features = my_markers) +
  theme(axis.text.y = element_text(angle = 0, hjust = 1, vjust = 0.5)) +
  scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 11, name = "RdBu") %>% rev()) +
  scale_x_discrete() +
  scale_y_discrete(limits = rev) +
  labs(x = "", y = "") +
  theme(axis.text.x = element_text(angle = 30, hjust = 1, size = 8))
dev.off()

gc()
```


### subset 2_3

```{r}
my_markers <- c("ADRB1", "CHRM2", "ADORA1", "ADORA2A", "EDNRA", "HRH2", "HTR4", "PTHER3", "GLP1R", "PTH1R", "VIPR2", "GABBR1", "GRM7", "FZD1", "FZD2", "FZD6", "FZD7", "FZD8", "ADGRA3", "ADGRB2", "ADGRB3", "ADGRD1", "ADGRG6", "ADGRL1", "ADGRL2", "ADGRL3", "ADGRV1", "LGR4", "GPR22", "GPR39", "GPR158", "GPR173", "GPRC5A", "GPRC5B")

CM_subset <- CM[, CM$clusters_subset_split %in% c("4-1",
                                                  "4-2",
                                                  "6")]

CM_subset$clusters_subset_split <- factor(CM_subset$clusters_subset_split, levels = c("4-1",
                                                  "4-2",
                                                  "6"))

CM_subset <- SetIdent(CM_subset, value = "clusters_subset_split")

pdf(paste0(myfolder, "complementary/DGE_CM_GPRC_subset_2_3.pdf"), width = 12, height = 3)
DotPlot(CM_subset, features = my_markers) +
  theme(axis.text.y = element_text(angle = 0, hjust = 1, vjust = 0.5)) +
  scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 11, name = "RdBu") %>% rev()) +
  scale_x_discrete() +
  scale_y_discrete(limits = rev) +
  labs(x = "", y = "") +
  theme(axis.text.x = element_text(angle = 30, hjust = 1, size = 8))
dev.off()

gc()
```

### Feature percentage individual cluster

```{r}
my_markers <- c("HCN1", "HCN4", "SCN5A", "CACNA1C", "CACNA1D", "CACNA1G", "CACNA2D1", "CACNA2D2", "CACNB2", "RYR2", "KCNA5", "KCNH7", "KCNJ2", "KCNJ3", "KCNJ4", "KCNMB2", "KCNN2", "KCNQ1", "KCNQ3", "KCNQ5", "CLCN3", "CLCN6", "CLIC1", "CLIC4", "CLIC5", "ANO4", "ANO5", "ANO6", "GRIA1", "GRID1", "GRID2", "GRIK1", "GRIK2", "GRIN2B", "P2RX1", "GABRB2", "GABRB3", "GABRE", "LRRC8B", "TRPC1", "TRPC3", "TRPM3", "TRPM4", "TRPM7", "PKD2", "GJA1", "GJA3", "GJA5", "GJC1", "AQP1")

nrow <- dim(CM)[2]

new_df <- data.frame(matrix(nrow = nrow, ncol = 0))
for (marker in my_markers) {
new_df <- cbind(new_df, PercentageFeatureSet(object = CM, features = marker))
}
colnames(new_df) <- my_markers

# take cells from a specific cluster
select_cluster <- "17"

# new data frame with cells only in specific cluster
new_df_spe <- new_df[rownames(new_df) %in% colnames(CM[, CM$clusters_subset_split == select_cluster]),]

count_greater_than_zero <- function(col) {
  as.integer(col > 0.00)
}
# Apply the function to each column of new_df_spe
result_df <- apply(new_df_spe, 2, count_greater_than_zero)
sum_of_ones <- colSums(result_df)
```


#### Feature percentage all clusters from object

```{r}
#parameters
nrow <- dim(CM)[2]
my_markers <- c("HCN1", "HCN4", "SCN5A", "CACNA1C", "CACNA1D", "CACNA1G", "CACNA2D1", "CACNA2D2", "CACNB2", "RYR2", "KCNA5", "KCNH7", "KCNJ2", "KCNJ3", "KCNJ4", "KCNMB2", "KCNN2", "KCNQ1", "KCNQ3", "KCNQ5", "CLCN3", "CLCN6", "CLIC1", "CLIC4", "CLIC5", "ANO4", "ANO5", "ANO6", "GRIA1", "GRID1", "GRID2", "GRIK1", "GRIK2", "GRIN2B", "P2RX1", "GABRB2", "GABRB3", "GABRE", "LRRC8B", "TRPC1", "TRPC3", "TRPM3", "TRPM4", "TRPM7", "PKD2", "GJA1", "GJA3", "GJA5", "GJC1", "AQP1") #ion channel

my_markers <- c("GNA12", "GNAI1", "GNAL", "GNAO1", "GNAQ", "GNAS", "GNB2", "GNB4", "GNB5", "GNG5", "GNG7", "GNG10", "GNG11", "GNG12", "RGS2", "RGS3", "RGS5", "RGS6", "RGS9", "RGS12") #GPRC1

my_markers <- c("ADRB1", "CHRM2", "ADORA1", "ADORA2A", "EDNRA", "HRH2", "HTR4", "GLP1R", "PTH1R", "VIPR2", "GABBR1", "GRM7", "FZD1", "FZD2", "FZD6", "FZD7", "FZD8", "ADGRA3", "ADGRB2", "ADGRB3", "ADGRD1", "ADGRG6", "ADGRL1", "ADGRL2", "ADGRL3", "ADGRV1", "LGR4", "GPR22", "GPR39", "GPR158", "GPR173", "GPRC5A", "GPRC5B") #GPRC2 ("PTHER3",)

#count function
count_greater_than_zero <- function(col) {
  as.integer(col > 0.00)
}

#create matrix with percentage of expression of each mentionned markers per cells
new_df <- data.frame(matrix(nrow = nrow, ncol = 0))
for (marker in my_markers) {
print(marker)
new_df <- cbind(new_df, PercentageFeatureSet(object = CM, features = marker))
}
colnames(new_df) <- my_markers

# Initialize an empty matrix to store the counts
marker_cluster_counts <- matrix(0, nrow = length(unique(CM$clusters_subset_split)), ncol = length(my_markers))
rownames(marker_cluster_counts) <- unique(CM$clusters_subset_split)
colnames(marker_cluster_counts) <- my_markers

# Loop through clusters
for (cluster in unique(CM$clusters_subset_split)) {
  # New data frame with cells only in the specific cluster
  new_df_spe <- new_df[rownames(new_df) %in% colnames(CM[, CM$clusters_subset_split == cluster]),]
  
  # Apply the count function to each column
  result_df <- apply(new_df_spe, 2, count_greater_than_zero)
  
  # Sum the counts for each marker
  sum_of_ones <- colSums(result_df)
  
  # Assign the counts to the corresponding row in the matrix
  marker_cluster_counts[as.character(cluster), ] <- sum_of_ones
}

#convert to dataframe
marker_cluster_counts <- as.data.frame(marker_cluster_counts)

#save as table
write.csv2(marker_cluster_counts, file = paste0(myfolder, "complementary/cells_marker_cluster_CM.csv"))
write.csv2(marker_cluster_counts, file = paste0(myfolder, "complementary/cells_marker_cluster_GPRC1_CM.csv"))
write.csv2(marker_cluster_counts, file = paste0(myfolder, "complementary/cells_marker_cluster_GPRC2_CM.csv"))


## Total number of cells per cluster
table1 <- as.data.frame(table(CM$clusters_subset_split))
colnames(table1) <- c("cluster", "number of cells")
write.csv2(table1, file = paste0(myfolder, "complementary/total_cells_CM.csv"))

```





## CM Dotplot - Gene enrichment - list 1

```{r}
my_markers <- c("ANO4", "ANO5", "ANO6", "AQP1", "CACNA1C", "CACNA1D", "CACNA1E", "CACNA1G", "CACNA2D1", "CACNA2D2", "CACNAB2", "CLCN3", "CLCN5", "CLCN6", "CLIC1", "CLIC4", "CLIC5", "GABRA4", "GABRB2", "GABRB3", "GABRE", "GABRR2", "GJA1", "GJA3", "GJA5", "GJC1", "GRIA1", "GRIA3", "GRID1", "GRID2", "GRIK1", "GRIK2", "GRIN2B", "GRIP1", "HCN1", "HCN4", "KCNA4", "KCNA5", "KCND2", "KCNH7", "KCNJ2", "KCNJ3", "KCNJ4", "KCNJ5", "KCNK13", "KCNN2", "KCNQ1", "KCNQ3", "KCNQ5", "KCNS3", "LRRC1", "LRRC23", "LRRC39", "LRRC4C", "LRRC8B", "LRRFIP1", "LRRFIP2", "LRRK2", "LRRN2", "LRRTM3", "LRRTM4", "P2RX1", "PKD2", "RYR2", "SCN5A", "TRPC1", "TRPC3", "TRPC4", "TRPC4AP", "TRPC6", "TRPM3", "TRPM4", "TRPM7", "TRPS1")

CM_subset <- CM[, CM$clusters_subset_split %in% c("6", "1", "10", "8-1", "8-3", "17", "8-2", "18", "4-2",  "4-1", "3", "14",  "9",  "12", "19-2",  "19-1", "7",  "2", "5")]

CM_subset$clusters_subset_split <- factor(CM_subset$clusters_subset_split, levels = c("6", "1", "10", "8-1", "8-3", "17", "8-2", "18", "4-2",  "4-1", "3", "14",  "9",  "12", "19-2",  "19-1", "7",  "2", "5"))

CM_subset <- SetIdent(CM_subset, value = "clusters_subset_split")

pdf(paste0(myfolder, "complementary/DGE_CM_list1.pdf"), width = 7, height = 12)
DotPlot(CM_subset, features = my_markers) +
  coord_flip() +
  theme(axis.text.y = element_text(angle = 0, hjust = 1, vjust = 0.5)) +
  scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 11, name = "RdBu") %>% rev()) +
  scale_x_discrete(limits = rev) +
  scale_y_discrete() +
  labs(x = "", y = "") + 
  theme(axis.text.x = element_text(angle = 30, hjust = 1, size = 8))
dev.off()

gc()
```


## CM Dotplot - Gene enrichment - list 2

```{r}
my_markers <- c("ADGRA3", "ADGRB2", "ADGRB3", "ADGRD1", "ADGRG6", "ADGRL1", "ADGRL2", "ADGRL3", "ADGRV1", "ADORA1", "ADORA2A", "ADRB1", "CHRM2", "EDNRA", "FZD1", "FZD2", "FZD6", "FZD7", "FZD8", "GABBR1", "GPR137C", "GPR158", "GPR173", "GPR22", "GPR39", "GPR89A", "GPR89B", "GPRC5A", "GPRC5B", "GRM7", "HRH2", "LGR4", "PFH1R", "PTGER3", "VIPR2")

CM_subset <- CM[, CM$clusters_subset_split %in% c("6", "1", "10", "8-1", "8-3", "17", "8-2", "18", "4-2",  "4-1", "3", "14",  "9",  "12", "19-2",  "19-1", "7",  "2", "5")]

CM_subset$clusters_subset_split <- factor(CM_subset$clusters_subset_split, levels = c("6", "1", "10", "8-1", "8-3", "17", "8-2", "18", "4-2",  "4-1", "3", "14",  "9",  "12", "19-2",  "19-1", "7",  "2", "5"))

CM_subset <- SetIdent(CM_subset, value = "clusters_subset_split")

pdf(paste0(myfolder, "complementary/DGE_CM_list2.pdf"), width = 7, height = 7)
DotPlot(CM_subset, features = my_markers) +
  coord_flip() +
  theme(axis.text.y = element_text(angle = 0, hjust = 1, vjust = 0.5)) +
  scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 11, name = "RdBu") %>% rev()) +
  scale_x_discrete(limits = rev) +
  scale_y_discrete() +
  labs(x = "", y = "") + 
  theme(axis.text.x = element_text(angle = 30, hjust = 1, size = 8))
dev.off()

gc()
```


## CM Dotplot - Gene enrichment - list 3

```{r}
my_markers <- c("GNA12", "GNA13", "GNAI1", "GNAL", "GNAO1", "GNAQ", "GNAS", "GNB2", "GNB4", "GNB5", "GNG10", "GNG11", "GNG12", "GNG2", "GNG4", "GNG5", "GNG7", "RGS12", "RGS2", "RGS3", "RGS5", "RGS6", "RGS9")

CM_subset <- CM[, CM$clusters_subset_split %in% c("6", "1", "10", "8-1", "8-3", "17", "8-2", "18", "4-2",  "4-1", "3", "14",  "9",  "12", "19-2",  "19-1", "7",  "2", "5")]

CM_subset$clusters_subset_split <- factor(CM_subset$clusters_subset_split, levels = c("6", "1", "10", "8-1", "8-3", "17", "8-2", "18", "4-2",  "4-1", "3", "14",  "9",  "12", "19-2",  "19-1", "7",  "2", "5"))

CM_subset <- SetIdent(CM_subset, value = "clusters_subset_split")

pdf(paste0(myfolder, "complementary/DGE_CM_list3.pdf"), width = 7, height = 5)
DotPlot(CM_subset, features = my_markers) +
  coord_flip() +
  theme(axis.text.y = element_text(angle = 0, hjust = 1, vjust = 0.5)) +
  scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 11, name = "RdBu") %>% rev()) +
  scale_x_discrete(limits = rev) +
  scale_y_discrete() +
  labs(x = "", y = "") + 
  theme(axis.text.x = element_text(angle = 30, hjust = 1, size = 8))
dev.off()

gc()
```


## CM Dotplot - MY

```{r}
my_markers <- c("MYH7", "MYL2", "MYH6", "MYL4")

## remove: 11, 13, 15, 16, 20, 22
CM_subset <- CM[, CM$clusters_subset_split %in% c("1" ,  "2" ,   "3",    "4-1",    "4-2", "5",    "6",    "7",    "8-1",  "8-2","8-3",   "9",    "10",     "12",    "14",     "17",   "18",   "19-1", "19-2",  "21")]

CM_subset$clusters_subset_split <- factor(CM_subset$clusters_subset_split, levels = c("1" ,  "2" ,   "3",    "4-1",    "4-2", "5",    "6",    "7",    "8-1",  "8-2","8-3",   "9",    "10",     "12",    "14",     "17",   "18",   "19-1", "19-2",  "21"))
CM_subset <- SetIdent(CM_subset, value = "clusters_subset_split")

pdf(paste0(myfolder, "complementary/DGE_CM_MY.pdf"), width = 4, height = 5)
DotPlot(CM_subset, features = my_markers) +
  theme(axis.text.y = element_text(angle = 0, hjust = 1, vjust = 0.5)) +
  scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 11, name = "RdBu") %>% rev()) +
  scale_x_discrete() +
  scale_y_discrete(limits = rev) +
  labs(x = "", y = "") + 
  theme(axis.text.x = element_text(angle = 30, hjust = 1, size = 8))
dev.off()

rm(p)
gc()
```


### FeaturePlot

```{r}
## remove: 11, 13, 15, 16, 20, 21, 22
CM_subset <- CM[, CM$clusters_subset_split %in% c("19-2", "19-1", "12", "9", "14", "4-2", "4-1", "18", "8-2", "8-1", "8-3", "17", "6", "1", "10", "3", "7", "2", "5")]

CM_subset$clusters_subset_split <- factor(CM_subset$clusters_subset_split, levels = c("19-2", "19-1", "12", "9", "14", "4-2", "4-1", "18", "8-2", "8-1", "8-3", "17", "6", "1", "10", "3", "7", "2", "5"))
CM_subset <- SetIdent(CM_subset, value = "clusters_subset_split")
```

```{r}
DimPlot(CM_subset, reduction = "umap_subset", raster = FALSE)
```

#### MYH6

```{r}
pdf(paste0(myfolder, "cardiomyocytes/FeaturePlot_MYH6.pdf"), width = 10, height = 10)
FeaturePlot(CM_subset, reduction = "umap_subset", features = "MYH6", raster = FALSE, order = T)
dev.off()
```

```{r}
pdf(paste0(myfolder, "cardiomyocytes/FeaturePlot_MYH6_not_ordered.pdf"), width = 10, height = 10)
FeaturePlot(CM_subset, reduction = "umap_subset", features = "MYH6", raster = FALSE, order = F)
dev.off()
```

#### MYH7

```{r}
pdf(paste0(myfolder, "cardiomyocytes/FeaturePlot_MYH7.pdf"), width = 10, height = 10)
FeaturePlot(CM_subset, reduction = "umap_subset", features = "MYH7", raster = FALSE, order = T)
dev.off()
```

```{r}
pdf(paste0(myfolder, "cardiomyocytes/FeaturePlot_MYH7_not_ordered.pdf"), width = 10, height = 10)
FeaturePlot(CM_subset, reduction = "umap_subset", features = "MYH7", raster = FALSE, order = F)
dev.off()
```

## CM Table subset / age

```{r}
CM_subset <- CM[, CM$clusters_subset_split %in% c("19-2", "19-1", "9", "14", "12", "4-1", "4-2", "6")]
CM_subset$clusters_subset_split <- factor(CM_subset$clusters_subset_split, levels = c("19-2", "19-1", "9", "14", "12", "4-1", "4-2", "6"))

### Add additional age groups to the metadata
df <- data.frame(age = c(5.5, 6, 7, 8, 8.5, 9, 10, 11.5, 12, 13.25, 14), age_group = c("5-6", "5-6", "7-8", "7-8", "7-8", "9-14", "9-14", "9-14", "9-14", "9-14", "9-14"))
age_to_age_group <- setNames(df$age_group, nm = df$age)


CM_subset$age_groups_refined <- factor(age_to_age_group[as.character(CM_subset$age)], levels = c("5-6", "7-8", "9-14"))
rm(df)
gc()

tb <- table(CM_subset$age_groups_refined, CM_subset$clusters_subset_split)
write.csv2(tb, paste0(myfolder, "cardiomyocytes/CM_subset_per_age.csv"))
```



## CM Dotplot - markers - 4 age

```{r}
my_markers <- c("CHRM1", "CHRM2", "CHRM3", "CHRM4", "CHRM5", "ADRA1", "ADRA2", "ADRB1", "ADRB2", "ADRB3", "HCN1", "HCN4", "SCN5A", "CACNA1C", "CACNA1D", "CACNA1G", "CACNA2D1", "CACNB2", "KCNA5", "KCNH7", "KCNJ2", "KCNJ3", "KCNJ4", "KCNN2", "KCNQ1", "KCNQ2", "KCNQ3", "KCNQ5", "KCNT2", "KCNIP4", "KCNMB2", "CLCN3", "CLCN6")

CM$groups <- paste0("cluster_", CM$clusters_subset_split, "_age_", CM$age_groups)

CM$groups <- factor(CM$groups, levels = c(  "cluster_1_age_5-6", "cluster_1_age_7-8", "cluster_1_age_9-11", "cluster_1_age_12-14",
                                              "cluster_2_age_5-6", "cluster_2_age_7-8", "cluster_2_age_9-11", "cluster_2_age_12-14",
                                              "cluster_3_age_5-6", "cluster_3_age_7-8", "cluster_3_age_9-11", "cluster_3_age_12-14",
                                              "cluster_4-1_age_5-6", "cluster_4-1_age_7-8", "cluster_4-1_age_9-11", "cluster_4-1_age_12-14",
                                              "cluster_4-2_age_5-6", "cluster_4-2_age_7-8", "cluster_4-2_age_9-11", "cluster_4-2_age_12-14",
                                              "cluster_5_age_5-6", "cluster_5_age_7-8", "cluster_5_age_9-11", "cluster_5_age_12-14",
                                              "cluster_6_age_5-6", "cluster_6_age_7-8", "cluster_6_age_9-11", "cluster_6_age_12-14",
                                              "cluster_7_age_5-6", "cluster_7_age_7-8", "cluster_7_age_9-11", "cluster_7_age_12-14",
                                              "cluster_8-1_age_5-6", "cluster_8-1_age_7-8", "cluster_8-1_age_9-11", "cluster_8-1_age_12-14",
                                              "cluster_8-2_age_5-6", "cluster_8-2_age_7-8", "cluster_8-2_age_9-11", "cluster_8-2_age_12-14",
                                              "cluster_8-3_age_5-6", "cluster_8-3_age_7-8", "cluster_8-3_age_9-11", "cluster_8-3_age_12-14",
                                              "cluster_9_age_5-6", "cluster_9_age_7-8", "cluster_9_age_9-11", "cluster_9_age_12-14",
                                              "cluster_10_age_5-6", "cluster_10_age_7-8", "cluster_10_age_9-11", "cluster_10_age_12-14",
                                              "cluster_11_age_5-6", "cluster_11_age_7-8", "cluster_11_age_9-11", "cluster_11_age_12-14",
                                              "cluster_12_age_5-6", "cluster_12_age_7-8", "cluster_12_age_9-11", "cluster_12_age_12-14",
                                              "cluster_13_age_5-6", "cluster_13_age_7-8", "cluster_13_age_9-11", "cluster_13_age_12-14",
                                              "cluster_14_age_5-6", "cluster_14_age_7-8", "cluster_14_age_9-11", "cluster_14_age_12-14",
                                              "cluster_15_age_5-6", "cluster_15_age_7-8", "cluster_15_age_9-11", "cluster_15_age_12-14",
                                              "cluster_16_age_5-6", "cluster_16_age_7-8", "cluster_16_age_9-11", "cluster_16_age_12-14",
                                              "cluster_17_age_5-6", "cluster_17_age_7-8", "cluster_17_age_9-11", "cluster_17_age_12-14",
                                              "cluster_18_age_5-6", "cluster_18_age_7-8", "cluster_18_age_9-11", "cluster_18_age_12-14",
                                              "cluster_19-1_age_5-6", "cluster_19-1_age_7-8", "cluster_19-1_age_9-11", "cluster_19-1_age_12-14",
                                              "cluster_19-2_age_5-6", "cluster_19-2_age_7-8", "cluster_19-2_age_9-11", "cluster_19-2_age_12-14",
                                              "cluster_20_age_5-6", "cluster_20_age_7-8", "cluster_20_age_9-11", "cluster_20_age_12-14",
                                              "cluster_21_age_5-6", "cluster_21_age_7-8", "cluster_21_age_9-11", "cluster_21_age_12-14",
                                              "cluster_22_age_5-6", "cluster_22_age_7-8", "cluster_22_age_9-11", "cluster_22_age_12-14"
                                              ))

table(CM$groups)
unique(CM$groups)

sample_size <- table(CM$groups)

# Randomly sample the cells that will be included when calculating DGE (same number of cells as selected above).
# Samples without replacement, so each cell only can be included once.
DGE_cells <- lapply(names(sample_size), function(x){
  set.seed(1)
  colnames(CM)[CM$groups == x]
  })
DGE_cells <- unlist(DGE_cells)

# Select all data that relates to the sampled DGE cells.
DGE_DATA <- CM[, DGE_cells]

#Find differential expressed genes based on groups defined by "clusters_subset".
DGE_DATA <- SetIdent(DGE_DATA, value = "groups")

pdf(paste0(myfolder, "complementary/DGE_CM_markers_4_age.pdf"), width = 25, height = 8)
par(mfrow=c(1,1), mar=c(3,6,2,5))
DotPlot(DGE_DATA, features = my_markers) +
  coord_flip() +
  theme(axis.text.y = element_text(angle = 0, hjust = 1, vjust = 0.5)) +
  scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 11, name = "RdBu") %>% rev()) +
  scale_x_discrete(limits = rev) +
  scale_y_discrete() +
  #labs(x = "", y = "") +
  theme(axis.text.x = element_text(angle = 90, hjust = 0.5))
dev.off()
```


## CM Dotplot - markers - 3 age

```{r}
my_markers <- c("CHRM1", "CHRM2", "CHRM3", "CHRM4", "CHRM5", "ADRA1", "ADRA2", "ADRB1", "ADRB2", "ADRB3", "HCN1", "HCN4", "SCN5A", "CACNA1C", "CACNA1D", "CACNA1G", "CACNA2D1", "CACNB2", "KCNA5", "KCNH7", "KCNJ2", "KCNJ3", "KCNJ4", "KCNN2", "KCNQ1", "KCNQ2", "KCNQ3", "KCNQ5", "KCNT2", "KCNIP4", "KCNMB2", "CLCN3", "CLCN6")

CM_subset <- CM[, CM$clusters_subset_split %in% c("6", "1", "10", "8-1", "8-3", "17", "8-2", "18", "4-2", "4-1", "3", "14", "9", "12", "21", "19-2", "19-1", "7", "2", "5")]

CM_subset$group_refined <- paste0("cluster_", CM_subset$clusters_subset_split, "_age_", CM_subset$age_group_refined)

CM_subset$group_refined <- factor(CM_subset$group_refined, levels = c(
                                              "cluster_6_age_5-6", "cluster_6_age_7-8", "cluster_6_age_9-14",
                                              "cluster_1_age_5-6", "cluster_1_age_7-8", "cluster_1_age_9-14",
                                              "cluster_10_age_5-6", "cluster_10_age_7-8", "cluster_10_age_9-14",
                                              "cluster_8-1_age_5-6", "cluster_8-1_age_7-8", "cluster_8-1_age_9-14",
                                              "cluster_8-3_age_5-6", "cluster_8-3_age_7-8", "cluster_8-3_age_9-14",
                                              "cluster_17_age_5-6", "cluster_17_age_7-8", "cluster_17_age_9-14",
                                              "cluster_8-2_age_5-6", "cluster_8-2_age_7-8", "cluster_8-2_age_9-14",
                                              "cluster_18_age_5-6", "cluster_18_age_7-8", "cluster_18_age_9-14",
                                              "cluster_4-2_age_5-6", "cluster_4-2_age_7-8", "cluster_4-2_age_9-14",
                                              "cluster_4-1_age_5-6", "cluster_4-1_age_7-8", "cluster_4-1_age_9-14",
                                              "cluster_3_age_5-6", "cluster_3_age_7-8", "cluster_3_age_9-14",
                                              "cluster_14_age_5-6", "cluster_14_age_7-8", "cluster_14_age_9-14",
                                              "cluster_9_age_5-6", "cluster_9_age_7-8", "cluster_9_age_9-14",
                                              "cluster_12_age_5-6", "cluster_12_age_7-8", "cluster_12_age_9-14",
                                              "cluster_21_age_5-6", "cluster_21_age_7-8", "cluster_21_age_9-14",
                                              "cluster_19-2_age_5-6", "cluster_19-2_age_7-8", "cluster_19-2_age_9-14",
                                              "cluster_19-1_age_5-6", "cluster_19-1_age_7-8", "cluster_19-1_age_9-14",
                                              "cluster_7_age_5-6", "cluster_7_age_7-8", "cluster_7_age_9-14",
                                              "cluster_2_age_5-6", "cluster_2_age_7-8", "cluster_2_age_9-14",
                                              "cluster_5_age_5-6", "cluster_5_age_7-8", "cluster_5_age_9-14"

                                              # "cluster_11_age_5-6", "cluster_11_age_7-8", "cluster_11_age_9-14",
                                              # "cluster_13_age_5-6", "cluster_13_age_7-8", "cluster_13_age_9-14",
                                              # "cluster_15_age_5-6", "cluster_15_age_7-8", "cluster_15_age_9-14",
                                              # "cluster_16_age_5-6", "cluster_16_age_7-8", "cluster_16_age_9-14",
                                              # "cluster_20_age_5-6", "cluster_20_age_7-8", "cluster_20_age_9-14"
                                                                                           ))

CM_subset <- SetIdent(CM_subset, value = "group_refined")

pdf(paste0(myfolder, "complementary/DGE_CM_markers_3_age.pdf"), width = 18, height = 8)
par(mfrow=c(1,1), mar=c(3,6,2,5))
DotPlot(CM_subset, features = my_markers) +
  coord_flip() +
  theme(axis.text.y = element_text(angle = 0, hjust = 1, vjust = 0.5)) +
  scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 11, name = "RdBu") %>% rev()) +
  scale_x_discrete(limits = rev) +
  scale_y_discrete() +
  #labs(x = "", y = "") +
  theme(axis.text.x = element_text(angle = 90, hjust = 0.5))
dev.off()
```


## CM Dotplot - markers - 3 age - CM_19_2, 19_1, 12, 9, 14

```{r}
my_markers <- c("HCN1", "HCN4", "CACNA1C", "CACNA1D", "CACNA1G", "CACNA2D2", "CACNB2", "KCNJ3", "KCNQ1", "KCNQ3", "KCNQ5", "KCNT2", "CLCN3", "ANO6", "GRID2", "GRIK1", "LRRC8B", "TRPC1", "TRPC3", "TRPM4", "GJA3", "GJC1")

CM_subset <- CM[, CM$clusters_subset_split %in% c("19-2", "19-1", "12", "9", "14")]

CM_subset$group_refined <- paste0("cluster_", CM_subset$clusters_subset_split, "_age_", CM_subset$age_group_refined)

CM_subset$group_refined <- factor(CM_subset$group_refined, levels = c(
                                              "cluster_19-2_age_5-6", "cluster_19-2_age_7-8", "cluster_19-2_age_9-14",
                                              "cluster_19-1_age_5-6", "cluster_19-1_age_7-8", "cluster_19-1_age_9-14",
                                              "cluster_12_age_5-6", "cluster_12_age_7-8", "cluster_12_age_9-14",
                                              "cluster_9_age_5-6", "cluster_9_age_7-8", "cluster_9_age_9-14",
                                              "cluster_14_age_5-6", "cluster_14_age_7-8", "cluster_14_age_9-14"

                                              # "cluster_11_age_5-6", "cluster_11_age_7-8", "cluster_11_age_9-14",
                                              # "cluster_13_age_5-6", "cluster_13_age_7-8", "cluster_13_age_9-14",
                                              # "cluster_15_age_5-6", "cluster_15_age_7-8", "cluster_15_age_9-14",
                                              # "cluster_16_age_5-6", "cluster_16_age_7-8", "cluster_16_age_9-14",
                                              # "cluster_20_age_5-6", "cluster_20_age_7-8", "cluster_20_age_9-14"
                                                                                           ))

CM_subset <- SetIdent(CM_subset, value = "group_refined")

pdf(paste0(myfolder, "complementary/DGE_CM_markers_3_age_CM_19-2_19-1_12_9_14.pdf"), width = 6.5, height = 6)
par(mfrow=c(1,1), mar=c(3,6,2,5))
DotPlot(CM_subset, features = my_markers) +
  coord_flip() +
  theme(axis.text.y = element_text(angle = 0, hjust = 1, vjust = 0.5)) +
  scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 11, name = "RdBu") %>% rev()) +
  scale_x_discrete(limits = rev) +
  scale_y_discrete() +
  #labs(x = "", y = "") +
  theme(axis.text.x = element_text(angle = 90, hjust = 0.5))
dev.off()
```


## CM Dotplot - markers - 3 age - CM_19_2

```{r}
my_markers <- c("HCN1", "HCN4", "CACNA1C", "CACNA1D", "CACNA1G", "CACNA2D2", "CACNB2", "KCNJ3", "KCNQ1", "KCNQ3", "KCNQ5", "KCNT2", "CLCN3", "ANO6", "GRID2", "GRIK1", "LRRC8B", "TRPC1", "TRPC3", "TRPM4", "GJA3", "GJC1")

CM_subset <- CM[, CM$clusters_subset_split %in% c("19-2")]

CM_subset$group_refined <- paste0("cluster_", CM_subset$clusters_subset_split, "_age_", CM_subset$age_group_refined)

CM_subset$group_refined <- factor(CM_subset$group_refined, levels = c(
                                              "cluster_19-2_age_5-6", "cluster_19-2_age_7-8", "cluster_19-2_age_9-14"
                                              # "cluster_19-1_age_5-6", "cluster_19-1_age_7-8", "cluster_19-1_age_9-14",
                                              # "cluster_12_age_5-6", "cluster_12_age_7-8", "cluster_12_age_9-14",
                                              # "cluster_9_age_5-6", "cluster_9_age_7-8", "cluster_9_age_9-14",
                                              # "cluster_14_age_5-6", "cluster_14_age_7-8", "cluster_14_age_9-14"

                                              # "cluster_11_age_5-6", "cluster_11_age_7-8", "cluster_11_age_9-14",
                                              # "cluster_13_age_5-6", "cluster_13_age_7-8", "cluster_13_age_9-14",
                                              # "cluster_15_age_5-6", "cluster_15_age_7-8", "cluster_15_age_9-14",
                                              # "cluster_16_age_5-6", "cluster_16_age_7-8", "cluster_16_age_9-14",
                                              # "cluster_20_age_5-6", "cluster_20_age_7-8", "cluster_20_age_9-14"
                                                                                           ))

CM_subset <- SetIdent(CM_subset, value = "group_refined")

pdf(paste0(myfolder, "complementary/DGE_CM_markers_3_age_CM_19-2.pdf"), width = 4, height = 7)
par(mfrow=c(1,1), mar=c(3,6,2,5))
DotPlot(CM_subset, features = my_markers) +
  coord_flip() +
  theme(axis.text.y = element_text(angle = 0, hjust = 1, vjust = 0.5)) +
  scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 11, name = "RdBu") %>% rev()) +
  scale_x_discrete(limits = rev) +
  scale_y_discrete() +
  #labs(x = "", y = "") +
  theme(axis.text.x = element_text(angle = 90, hjust = 0.5))
dev.off()
```


## CM Dotplot - markers - 3 age - CM_19_2, 19_1, 12, 9, 14, 4_2, 4_1, 18, 6

```{r}
my_markers <- c("ADRB1", "CHRM2")

CM_subset <- CM[, CM$clusters_subset_split %in% c("19-2", "19-1", "12", "9", "14", "4-2", "4-1", "18", "6")]

CM_subset$group_refined <- paste0("cluster_", CM_subset$clusters_subset_split, "_age_", CM_subset$age_group_refined)

CM_subset$group_refined <- factor(CM_subset$group_refined, levels = c(
                                              "cluster_19-2_age_5-6", "cluster_19-2_age_7-8", "cluster_19-2_age_9-14",
                                              "cluster_19-1_age_5-6", "cluster_19-1_age_7-8", "cluster_19-1_age_9-14",
                                              "cluster_12_age_5-6", "cluster_12_age_7-8", "cluster_12_age_9-14",
                                              "cluster_9_age_5-6", "cluster_9_age_7-8", "cluster_9_age_9-14",
                                              "cluster_14_age_5-6", "cluster_14_age_7-8", "cluster_14_age_9-14",
                                              "cluster_4-2_age_5-6", "cluster_4-2_age_7-8", "cluster_4-2_age_9-14",
                                              "cluster_4-1_age_5-6", "cluster_4-1_age_7-8", "cluster_4-1_age_9-14",
                                              "cluster_18_age_5-6", "cluster_18_age_7-8", "cluster_18_age_9-14",
                                              "cluster_6_age_5-6", "cluster_6_age_7-8", "cluster_6_age_9-14"

                                              # "cluster_11_age_5-6", "cluster_11_age_7-8", "cluster_11_age_9-14",
                                              # "cluster_13_age_5-6", "cluster_13_age_7-8", "cluster_13_age_9-14",
                                              # "cluster_15_age_5-6", "cluster_15_age_7-8", "cluster_15_age_9-14",
                                              # "cluster_16_age_5-6", "cluster_16_age_7-8", "cluster_16_age_9-14",
                                              # "cluster_20_age_5-6", "cluster_20_age_7-8", "cluster_20_age_9-14"
                                                                                           ))

CM_subset <- SetIdent(CM_subset, value = "group_refined")

pdf(paste0(myfolder, "complementary/DGE_CM_markers_3_age_CM_19-2_19_1_12_9_14_4_2_4_1_18_6.pdf"), width = 8.5, height = 5)
par(mfrow=c(1,1), mar=c(3,6,2,5))
DotPlot(CM_subset, features = my_markers) +
  coord_flip() +
  theme(axis.text.y = element_text(angle = 0, hjust = 1, vjust = 0.5)) +
  scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 11, name = "RdBu") %>% rev()) +
  scale_x_discrete(limits = rev) +
  scale_y_discrete() +
  #labs(x = "", y = "") +
  theme(axis.text.x = element_text(angle = 90, hjust = 0.5))
dev.off()
```


## Violin plot nFeatures

### CM - split & all

```{r}
data_CM <- CM

data_CM$clusters_subset_split <- factor(data_CM$clusters_subset_split, levels = c("6",  "1",  "10", "8-1", "8-3", "17", "8-2", "18", "4-2", "4-1", "3", "14", "9", "12", "21", "19-2", "19-1", "7", "2", "5"))

# Create a single violin plot
pdf(file = paste0(myfolder, "complementary/Vln_feature_CM_split_all.pdf"),  width = 15, height = 4)
VlnPlot(CM, group.by = "clusters_subset_split", features = "nFeature_RNA", pt.size = 0) +
stat_summary(fun.y = median, geom='point', size = 25, colour = "black", shape = 95) +
theme(legend.position = 'none')
dev.off()
```

### CM - split & selected

```{r}
values_to_keep <- c("6", "1", "10", "8-1", "8-3", "17", "8-2", "18", "4-2",  "4-1", "3", "14",  "9",  "12",  "21", "19-2",  "19-1", "7",  "2", "5")

# Filter the data using the %in% operator
data_CM <- CM[, CM$clusters_subset_split %in% values_to_keep]

data_CM$clusters_subset_split <- factor(data_CM$clusters_subset_split, levels = c("6",  "1",  "10", "8-1", "8-3", "17", "8-2", "18", "4-2", "4-1", "3", "14", "9", "12", "21", "19-2", "19-1", "7", "2", "5"))

# Create a single violin plot
pdf(file = paste0(myfolder, "complementary/Vln_feature_CM_split_ordered.pdf"),  width = 15, height = 4)
VlnPlot(data_CM, group.by = "clusters_subset_split", features = "nFeature_RNA", pt.size = 0) +
stat_summary(fun.y = median, geom='point', size = 25, colour = "black", shape = 95) +
theme(legend.position = 'none')
dev.off()
```


### CM - before split

```{r}
# Create a single violin plot
pdf(file = paste0(myfolder, "complementary/Vln_feature_CM_before_split.pdf"),  width = 15, height = 4)
VlnPlot(CM, group.by = "clusters_subset", features = "nFeature_RNA", pt.size = 0) +
stat_summary(fun.y = median, geom='point', size = 25, colour = "black", shape = 95) +
theme(legend.position = 'none')
dev.off()
```


### EN

```{r}
# Create a single violin plot
pdf(file = paste0(myfolder, "complementary/Vln_feature_EN.pdf"),  width = 15, height = 4)
VlnPlot(EN, group.by = "clusters_subset", features = "nFeature_RNA", pt.size = 0) +
stat_summary(fun.y = median, geom='point', size = 25, colour = "black", shape = 95) +
theme(legend.position = 'none')
dev.off()
```


### FB

```{r}
# Create a single violin plot
pdf(file = paste0(myfolder, "complementary/Vln_feature_FB.pdf"),  width = 15, height = 4)
VlnPlot(FB, group.by = "clusters_subset", features = "nFeature_RNA", pt.size = 0) +
stat_summary(fun.y = median, geom='point', size = 25, colour = "black", shape = 95) +
theme(legend.position = 'none')
dev.off()
```


### Data

```{r}
# Create a single violin plot
pdf(file = paste0(myfolder, "complementary/Vln_feature_data.pdf"),  width = 15, height = 4)
VlnPlot(data, group.by = "high_level_clusters_removed", features = "nFeature_RNA", pt.size = 0) +
stat_summary(fun.y = median, geom='point', size = 25, colour = "black", shape = 95) +
theme(legend.position = 'none')
dev.off()
```


### Data - all clusters (clusters_louvain_oi)

```{r}
# Order the clusters
# data$clusters_louvain_oi <- factor(data$clusters_louvain_oi, levels = c(""))

# Create a single violin plot
pdf(file = paste0(myfolder, "complementary/Vln_feature_data_all_clusters.pdf"),  width = 15, height = 4)
VlnPlot(data, group.by = "clusters_louvain_oi", features = "nFeature_RNA", pt.size = 0) +
stat_summary(fun.y = median, geom='point', size = 25, colour = "black", shape = 95) +
theme(legend.position = 'none')
dev.off()
```


## Single Gene violin plots

### HBA1 and HBE1 across all HL clusters (without exclusions, unordered)

```{r}
pdf(file = paste0(myfolder, "complementary/Vln_feature_HL_HBA1.pdf"),  width = 15, height = 4)
VlnPlot(data, group.by = "clusters_louvain_oi", features = "HBA1", pt.size = 0) +
theme(legend.position = 'none')
dev.off()

pdf(file = paste0(myfolder, "complementary/Vln_feature_HL_HBE1.pdf"),  width = 15, height = 4)
VlnPlot(data, group.by = "clusters_louvain_oi", features = "HBE1", pt.size = 0) +
theme(legend.position = 'none')
dev.off()

pdf(file = paste0(myfolder, "complementary/Vln_feature_HL_HBE1.pdf"),  width = 15, height = 4)
VlnPlot(data, group.by = "clusters_louvain_oi", features = "PDE4C", pt.size = 0) +
theme(legend.position = 'none')
dev.off()

pdf(file = paste0(myfolder, "complementary/Vln_feature_HL_HBE1.pdf"),  width = 15, height = 4)
VlnPlot(data, group.by = "clusters_louvain_oi", features = "FABP4", pt.size = 0) +
theme(legend.position = 'none')
dev.off()

pdf(file = paste0(myfolder, "complementary/Vln_feature_HL_HBE1.pdf"),  width = 15, height = 4)
VlnPlot(data, group.by = "clusters_louvain_oi", features = "RGCC", pt.size = 0) +
theme(legend.position = 'none')
dev.off()
```


### HBA1, MYH6 and PDE4C across all CM clusters (without exclusions, unordered)

```{r}
pdf(file = paste0(myfolder, "complementary/Vln_feature_CM_HBA1.pdf"),  width = 15, height = 4)
VlnPlot(CM, group.by = "clusters_subset", features = "HBA1", pt.size = 0) +
theme(legend.position = 'none')
dev.off()

pdf(file = paste0(myfolder, "complementary/Vln_feature_CM_HBE1.pdf"),  width = 15, height = 4)
VlnPlot(CM, group.by = "clusters_subset", features = "HBE1", pt.size = 0) +
theme(legend.position = 'none')
dev.off()

pdf(file = paste0(myfolder, "complementary/Vln_feature_CM_MYH6.pdf"),  width = 15, height = 4)
VlnPlot(CM, group.by = "clusters_subset_split", features = "MYH6", pt.size = 0) +
theme(legend.position = 'none')
dev.off()

pdf(file = paste0(myfolder, "complementary/Vln_feature_CM_PDE4C.pdf"),  width = 15, height = 4)
VlnPlot(CM, group.by = "clusters_subset_split", features = "PDE4C", pt.size = 0) +
theme(legend.position = 'none')
dev.off()

pdf(file = paste0(myfolder, "complementary/Vln_feature_CM_SEMA4A.pdf"),  width = 15, height = 4)
VlnPlot(CM, group.by = "clusters_subset_split", features = "SEMA4A", pt.size = 0) +
theme(legend.position = 'none')
dev.off()

pdf(file = paste0(myfolder, "complementary/Vln_feature_CM_DDR2.pdf"),  width = 15, height = 4)
VlnPlot(CM, group.by = "clusters_subset", features = "DDR2", pt.size = 0) +
theme(legend.position = 'none')
dev.off()
```


### CDH5 and PDE4C across all EC clusters (without exclusions, unordered)

```{r}
pdf(file = paste0(myfolder, "complementary/Vln_feature_EN_HBA1.pdf"),  width = 15, height = 4)
VlnPlot(EN, group.by = "clusters_subset", features = "HBA1", pt.size = 0) +
theme(legend.position = 'none')
dev.off()

pdf(file = paste0(myfolder, "complementary/Vln_feature_EN_HBE1.pdf"),  width = 15, height = 4)
VlnPlot(EN, group.by = "clusters_subset", features = "HBE1", pt.size = 0) +
theme(legend.position = 'none')
dev.off()

pdf(file = paste0(myfolder, "complementary/Vln_feature_EN_CDH5.pdf"),  width = 15, height = 4)
VlnPlot(EN, group.by = "clusters_subset", features = "CDH5", pt.size = 0) +
theme(legend.position = 'none')
dev.off()

pdf(file = paste0(myfolder, "complementary/Vln_feature_EN_PDE4C.pdf"),  width = 15, height = 4)
VlnPlot(EN, group.by = "clusters_subset", features = "PDE4C", pt.size = 0) +
theme(legend.position = 'none')
dev.off()
```


### HBA1 and PDE4C across all FB clusters (again without exclusions, unordered)

```{r}
pdf(file = paste0(myfolder, "complementary/Vln_feature_FB_HBA1.pdf"),  width = 15, height = 4)
VlnPlot(FB, group.by = "clusters_subset", features = "HBA1", pt.size = 0) +
theme(legend.position = 'none')
dev.off()

pdf(file = paste0(myfolder, "complementary/Vln_feature_FB_HBE1.pdf"),  width = 15, height = 4)
VlnPlot(FB, group.by = "clusters_subset", features = "HBE1", pt.size = 0) +
theme(legend.position = 'none')
dev.off()

pdf(file = paste0(myfolder, "complementary/Vln_feature_FB_PDE4C.pdf"),  width = 15, height = 4)
VlnPlot(FB, group.by = "clusters_subset", features = "PDE4C", pt.size = 0) +
theme(legend.position = 'none')
dev.off()
```


## Correlation of features in CM-21 

```{r}
CM_21 <- CM[,CM$clusters_subset == 21]

FeatureScatter(object = CM_21, feature1 = "DDR2", feature2 = "SEMA4A")

FeatureScatter(object = CM_21, feature1 = "MYH6", feature2 = "SEMA4A")

FeatureScatter(object = CM_21, feature1 = "MYH6", feature2 = "PDE4C")

FeatureScatter(object = CM_21, feature1 = "nCount_RNA", feature2 = "PDE4C")

FeatureScatter(object = CM_21, feature1 = "nFeature_RNA", feature2 = "SEMA4A", group.by = "age_groups")

FeatureScatter(object = CM_21, feature1 = "nFeature_RNA", feature2 = "DDR2")

FeatureScatter(object = CM_21, feature1 = "nFeature_RNA", feature2 = "TIMP2")

```

```{r}
CM_21
```


```{r}
EN_14 <- EN[,EN$clusters_subset == 14]


FeatureScatter(object = EN_14, feature1 = "PDE4C", feature2 = "FABP4")
FeatureScatter(object = EN_14, feature1 = "nFeature_RNA", feature2 = "PDE4C")
FeatureScatter(object = EN_14, feature1 = "nFeature_RNA", feature2 = "PDE4C")

```


```{r}
IN_15 <- IN[,IN$clusters_subset == 15]

FeatureScatter(object = IN_15, feature1 = "nFeature_RNA", feature2 = "PDE4C", group.by = "age_groups")

```


================================================================================
================================================================================

## DE analysis multi clusters

```{r}
data_curated <- data[, data$high_level_clusters %in% c("1", "2", "3", "4_35", "5", "6", "7", "8", "9", "10", "11", "12", "13", "14", "15", "16", "17", "18_1", "18_2", "20", "21", "22", "23", "24", "25", "26", "28", "29", "30", "31", "32", "33", "34")]

data_curated <- FindVariableFeatures(data_curated, nfeatures = 5000)
var.features <- data_curated@assays$RNA@var.features

sampled_cells <- data_curated[[]] %>% 
  tibble::rownames_to_column(var = "barcode") %>% 
  group_by(high_level_clusters)
rm(data_curated)

annotation <- sampled_cells %>% 
  select(barcode, high_level_clusters, age_groups) %>% 
  data.frame(row.names = 1)
colnames(annotation) <- c("bio_celltype", "age_groups")
```

```{r}
# Load annotations prepared earlier and filter
# CM
cardiomyocyte_clusters_keep <- c("1", "2", "3", "4-1", "4-2", "5", "6", "7", "8-1", "8-2", "8-3", "9", "10", "12", "14", "17", "18", "19-1", "19-2")
cardiomyocyte_metadata <- readRDS(paste0(myfolder, "rds_objects/cardiomyocyte_metadata.rds")) %>% 
  mutate(clusters_subset = paste0("CM_", as.character(clusters_subset))) %>%  # rename cardiomyocyte clusters by adding a prefix CM_
  mutate_if(is.factor, as.character) %>% 
  filter(clusters_subset_split %in% cardiomyocyte_clusters_keep)

# EN
endothelial_clusters_keep <- c("1", "2", "3", "5", "6", "7", "10", "11", "12", "13", "14", "15", "16", "17", "18", "19", "20", "21")
endothelial_metadata <- readRDS(paste0(myfolder, "rds_objects/endothelial_metadata.rds")) %>% 
  filter(clusters_subset %in% endothelial_clusters_keep) %>% 
  mutate(clusters_subset = paste0("EN_", as.character(clusters_subset))) # rename endothelial clusters by adding a prefix EN_

# FB
fibroblasts_clusters_keep <- c("1", "2", "3", "4", "5", "6", "7", "8", "9", "11", "12", "13", "14", "15", "17", "18", "19", "20")
fibroblasts_metadata <- readRDS(paste0(myfolder, "rds_objects/fibroblasts_metadata.rds")) %>% 
  filter(clusters_subset %in% fibroblasts_clusters_keep) %>% 
  mutate(clusters_subset = paste0("FB_", as.character(clusters_subset))) # rename fibroblasts clusters by adding a prefix FB_

# IN
innervation_clusters_keep <- c("1", "3", "4", "7", "8", "11", "12", "13", "17", "19")
innervation_metadata <- readRDS(paste0(myfolder, "rds_objects/innervation_metadata.rds")) %>% 
  filter(clusters_subset %in% innervation_clusters_keep) %>% 
  mutate(clusters_subset = paste0("IN_", as.character(clusters_subset))) # rename innervation clusters by adding a prefix IN_

# Merge deep level clusters from the three different subsets
deep_clusters <- cardiomyocyte_metadata %>% 
  select(clusters_subset, clusters_subset_split) %>% # Only select clustering results
  bind_rows(endothelial_metadata %>% select(clusters_subset)) %>% # Add endothelial clusters
  bind_rows(fibroblasts_metadata %>% select(clusters_subset)) %>% # Add fibroblast clusters
  bind_rows(innervation_metadata %>% select(clusters_subset)) %>% # Add innervation clusters
  mutate(clusters_subset_split = case_when(is.na(clusters_subset_split) ~ clusters_subset, # rename cardiomyocyte clusters by adding a prefix CM_
                                           TRUE ~ paste0("CM_", clusters_subset_split)))

# Define clusters to keep
keep_clusters <- c(paste0("HL_", c("6", "8", "10", "15", "18_1", "20", "22", "29", "34")), 
                   paste0("CM_", cardiomyocyte_clusters_keep),
                   paste0("EN_", endothelial_clusters_keep),
                   paste0("FB_", fibroblasts_clusters_keep),
                   paste0("IN_", innervation_clusters_keep))


# Create a new annotations tibble
annotation_split <- annotation 
annotation_split <- annotation_split %>% 
  rownames_to_column(var = "ID") %>% 
  left_join(y = deep_clusters %>% rownames_to_column(var = "ID"), by = "ID") %>% 
  mutate(clusters_subset_split = case_when(is.na(clusters_subset_split) ~ paste0("HL_", bio_celltype),
                                           TRUE ~ clusters_subset_split)) %>% 
  filter(clusters_subset_split %in% keep_clusters) # keep selected clusters

# Convert clusters_subset_split to a factor and set order of clusters
annotation_split <- annotation_split %>% 
  mutate(clusters_subset_split = factor(clusters_subset_split, keep_clusters))

# Check overlap (sanity check)
all(annotation$ID %in% colnames(data))

rm(annotation, cardiomyocyte_metadata, deep_clusters, endothelial_metadata, fibroblasts_metadata, sampled_cells, cardiomyocyte_clusters_keep, endothelial_clusters_keep, fibroblasts_clusters_keep, keep_clusters)
gc()
```

```{r}
data <- data[, colnames(data) %in% annotation_split$ID]
data$clusters_subset_split <- as.character(annotation_split$clusters_subset_split)
```


#### HL_22 FB_15

```{r}
clusters_to_analyze <- c("HL_22", "FB_15")

my_niche <- annotation_split[annotation_split$clusters_subset_split %in% clusters_to_analyze,]
data_sub <- data[, colnames(data) %in% my_niche$ID]

table(data_sub$clusters_subset_split)
```

```{r}
data_sub <- SetIdent(data_sub, value = "clusters_subset_split")

# The logfc threshold is set to 0.1 as compared to the default 0.25, which makes the detection of weaker signals possible (but it takes more computational power).
detable <- FindAllMarkers(data_sub,
                          only.pos = TRUE,
                          max.cells.per.ident = 300,
                          logfc.threshold = .1,
                          assay = "RNA",
                          min.pct = 0.05)

# Save the detable as a csv file.
write.csv2(detable, paste0(myfolder, "complementary/DEG_HL_22_FB_15.csv"))

# Only keep DEG that have a p-value below 0.05.
detable <- detable[detable$p_val < 0.05, ]

# Adding two extra columns giving the difference (and log difference) between pct1 and pct2.
# pct1 is the percentage of cells in the specific cluster that express this gene.
# pct2 is the percentage of cells in all other clusters that express this gene.
detable$pct.diff <- detable$pct.1 - detable$pct.2
detable$log.pct.diff <- log2((detable$pct.1*99+1) / (detable$pct.2*99+1))

# Pick the 40 genes with the lowest p value for each cluster, then 20 highest based on pct.diff, then 5 highest based on avg_log2FC.
detable %>%
  group_by(cluster) %>%
  top_n(-40, p_val) %>%
  top_n(20, pct.diff) %>%
  top_n(10, avg_log2FC) ->
  top10

pdf(paste0(myfolder, "complementary/DEG_HL_22_FB_15.pdf"), width = 4.5, height = 8)
par(mfrow=c(1,1), mar=c(3,6,2,5))
DotPlot(data_sub, features = top10 %>% pull(gene) %>% unique()) +
  coord_flip() +
  theme(axis.text.y = element_text(angle = 0, hjust = 1, vjust = 0.5)) +
  scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 11, name = "RdBu") %>% rev()) +
  scale_x_discrete(limits = rev) +
  scale_y_discrete() +
  labs(x = "", y = "") +
  theme(axis.text.x = element_text(angle = 30, hjust = 0.5))
dev.off()
```


#### HL_20, FB_7 and FB_3

```{r}
clusters_to_analyze <- c("HL_20", "FB_7", "FB_3")

my_niche <- annotation_split[annotation_split$clusters_subset_split %in% clusters_to_analyze,]
data_sub <- data[, colnames(data) %in% my_niche$ID]

table(data_sub$clusters_subset_split)
```

```{r}
data_sub <- SetIdent(data_sub, value = "clusters_subset_split")

# The logfc threshold is set to 0.1 as compared to the default 0.25, which makes the detection of weaker signals possible (but it takes more computational power).
detable <- FindAllMarkers(data_sub,
                          only.pos = TRUE,
                          max.cells.per.ident = 300,
                          logfc.threshold = .1,
                          assay = "RNA",
                          min.pct = 0.05)

# Save the detable as a csv file.
write.csv2(detable, paste0(myfolder, "complementary/DEG_HL_20_FB_7_FB_3.csv"))

# Only keep DEG that have a p-value below 0.05.
detable <- detable[detable$p_val < 0.05, ]

# Adding two extra columns giving the difference (and log difference) between pct1 and pct2.
# pct1 is the percentage of cells in the specific cluster that express this gene.
# pct2 is the percentage of cells in all other clusters that express this gene.
detable$pct.diff <- detable$pct.1 - detable$pct.2
detable$log.pct.diff <- log2((detable$pct.1*99+1) / (detable$pct.2*99+1))

# Pick the 40 genes with the lowest p value for each cluster, then 20 highest based on pct.diff, then 5 highest based on avg_log2FC.
detable %>%
  group_by(cluster) %>%
  top_n(-40, p_val) %>%
  top_n(20, pct.diff) %>%
  top_n(10, avg_log2FC) ->
  top10

pdf(paste0(myfolder, "complementary/DEG_HL_20_FB_7_FB_3.pdf"), width = 4.5, height = 8)
par(mfrow=c(1,1), mar=c(3,6,2,5))
DotPlot(data_sub, features = top10 %>% pull(gene) %>% unique()) +
  coord_flip() +
  theme(axis.text.y = element_text(angle = 0, hjust = 1, vjust = 0.5)) +
  scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 11, name = "RdBu") %>% rev()) +
  scale_x_discrete(limits = rev) +
  scale_y_discrete() +
  labs(x = "", y = "") +
  theme(axis.text.x = element_text(angle = 30, hjust = 0.5))
dev.off()
```


================================================================================
================================================================================
================================================================================
================================================================================
================================================================================









